--- 
title: "Bioinformatics for Evolution"
author: "Maria Fernanda Torres Jimenez"
date: "`r Sys.Date()`"
site: bookdown::bookdown_site
documentclass: book
bibliography: [book.bib, packages.bib]
# url: your book url like https://bookdown.org/bioinfo4evol/bookdown
# cover-image: path to the social sharing image like images/cover.jpg
description: |
  This is the material for the M.Sc. "Bioinformatics for Evolution" course at Vilnius University - 2023 Autumn. 
# This is a minimal example of using the bookdown package to write a book.
# The HTML output format for this example is bookdown::bs4_book,
# set in the _output.yml file.
biblio-style: apalike
csl: chicago-fullnote-bibliography.csl
---

# Sequencing technologies and file formats {#seqtech}

**Brief description:**
This notebook will walk you through the different sequence file formats generated by different sequencing technologies. You will learn to identify the metadata and quality information encoded in the files, perform data manipulations, and convert files between formats. We will also go over the SAM/BAM file format, CIGAR strings, information filters, and statistics.

1. Prepare the data
Before we start, we need to download some of the files from the Sequence Read Archive ([SRA](https://www.ncbi.nlm.nih.gov/sra)) using [sra-toolkit](https://hpc.nih.gov/apps/sratoolkit.html) commands. We will first install the tool kit and then get the read data, then, we will run the download jobs in the background using the command `nohup` ([no hang up](https://en.wikipedia.org/wiki/Nohup). `nohup` allows you to continue working on the command line prompt 

```bash
cd $HOME

# sra-toolkit for Ubuntu subsystems and Linux
wget https://ftp-trace.ncbi.nlm.nih.gov/sra/sdk/3.0.6/sratoolkit.3.0.6-ubuntu64.tar.gz

# untar and uncompress
tar xvzf sratoolkit.3.0.6-ubuntu64.tar.gz

# set a variable pointing to the command that we will use
cd sratoolkit.3.0.6-ubuntu64/bin
FASTDUMP="$PWD"/fastq-dump

# test the variable
$FASTDUMP --help

# go back to the lectures' folder
# edit the path to reflect the true path for you
cd /mnt/c/Users/username/Documents/folder/

# download samples. This command will keep the prompt busy and you won't be able to work
$FASTDUMP --split-3 DRR353013

# a useful command to run jobs in the "background" is nohup.
# we are also redirecting any stdout message to a file, in case we need to troubleshoot things
nohup $FASTDUMP --split-3 DRR353013 $2> fastdump.out &

# check the job is running using ps
ps

```


1. Text-based sequence formats
1.1. Sequence formats
1.1.1 Fasta
1.1.2. Fasta+Qual
1.1.3. Fastq
1.1.3.1. The PHRED score
1.1.3.2. Single sample file
1.1.3.3. Pooled sample file and demultiplexing
1.1.4. qseq
1.1.5. Sequence Read Format (SRF)

1.2. Annotation formats
1.2.1. GFF general
1.2.2. GTF
1.2.3. GFF3

2. Compressed or binary formats
2.1. The SAM/BAM file
2.2. The CRAM file
2.3. BCL Illumina format

3. Chromatogram based sequence formats
3.1. Capillary Sanger chromatograms

4. Hierarchical formats
4.1. Hierarchical Data Format version 5 (HDF5)
4.2. Fast5

5. Legacy formats
5.1. 

6. Summary by sequencing technology.
6.1. Sanger and capillar sequencing
6.2. Roche 545
6.3. SOLiD
IonTorrent
Helicos
Illumina
PAcBio
ONT

---

1. Text-based sequence formats
1.1. Sequence formats
1.1.1 Fasta
1.1.2. Fasta+Qual
1.1.3. Fastq
1.1.3.1. The PHRED score
1.1.3.2. Single sample file
1.1.3.3. Pooled sample file and demultiplexing
1.1.4. qseq
1.1.5. Sequence Read Format (SRF)

1.2. Annotation formats
1.2.1. GFF general
1.2.2. GTF
1.2.3. GFF3

2. Compressed or binary formats
2.1. The SAM/BAM file
2.2. The CRAM file
2.3. BCL Illumina format

3. Chromatogram based sequence formats
3.1. Capillary Sanger chromatograms

5. Hierarchical formats
5.1. Hierarchical Data Format version 5 (HDF5)
5.2. Fast5

5. Legacy formats
6. Summary by sequencing technology.
6.1. Sanger and capillar sequencing
6.2. Roche 545
6.3. SOLiD
IonTorrent
Helicos
Illumina
PAcBio
ONT
