{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7c073ce9",
   "metadata": {},
   "source": [
    "# Introduction to the command line\n",
    "\n",
    "**Brief description:** This notebook will walk you through the\n",
    "installation of Linux subsystems and Ubuntu OS and the steps for setting\n",
    "up a working environment. You will learn basic Linux commands from\n",
    "examples focused on analysing biological data, e.g. sequences and their\n",
    "metadata. This is an introductory tutorial for people that is not\n",
    "familiar with the command line.\n",
    "\n",
    "1.  [**The Linux subsystem (Windows only)**](#install-linux)  \n",
    "1.1. Installing the Linux subsystem  \n",
    "\n",
    "2.  [**Conda environments**](#conda)  \n",
    "2.1. Installing Conda  \n",
    "2.2. Creating and deleting environments  \n",
    "2.3. Creating environments from files  \n",
    "2.4. Exporting environments  \n",
    "\n",
    "3. [**Basic Linux commands**](#commands)  \n",
    "3.1. Folder hierarchy  \n",
    "3.2. Create, inspect, and delete files  \n",
    "3.3. Matching, replacing, and counting patterns  \n",
    "3.4. Variables, loops, and conditionals  \n",
    "3.5. Compress and decompress files  \n",
    "3.6. Connecting with other machines  \n",
    "\n",
    "4. [**AWK**](#awk)  \n",
    "4.1. Basic filtering  \n",
    "4.2. Operations  \n",
    "\n",
    "5. [**Summary**](#summary) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e028b43",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "\n",
    "# [**1. The Linux subsystem (Windows only)**](#install-linux) \n",
    "\n",
    "## 1.1. Installing the Linux subsystem\n",
    "\n",
    "Follow these steps if you work on a Windows machine (likely Windows 11 but it works for Windows 10 too). *If you have MacOS or Ubuntu/Linux already running in your machine, skip this section.*\n",
    "\n",
    "\n",
    "Right click on the Windows icon in the task bar, click search, search for 'turn windows features on or off'. Open the application.  \n",
    "\n",
    "Look for the options 'windows subsystem for linux' and 'virtual machine platform' and tick the boxes.  \n",
    "\n",
    "Hit OK and restart your machine.  \n",
    "\n",
    "Go again to the windows icon and search for bash. Open the application. If you see an error and a URL, open the link on a web browser and follow the instructions.\n",
    "\n",
    "Check **Lecture 2** slides for instructions."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f18ecfe3",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# [**2. Conda environments**](#conda)  \n",
    "\n",
    "This section takes you through installing Conda and managing environments. Documentation about Conda is available\n",
    "[here](https://conda.io/projects/conda/en/latest/user-guide/getting-started.html).  \n",
    "\n",
    "## 2.1. Installing Conda  \n",
    "\n",
    "Download the Linux version of Conda [here](https://www.anaconda.com/download). Scroll down the website and you will find it. It is important that you get the Linux distribution as you will be installing it in a **Linux subsystem (even within windows)**.\n",
    "\n",
    "Use the `wget` command to retrieve the file from conda\\'s website.\n",
    "\n",
    "The code below will not work easily if you are working under a VU VPN. If that is the case, try the manual download."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9849a6fb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GNU Wget 1.20.3, a non-interactive network retriever.\n",
      "Usage: wget [OPTION]... [URL]...\n",
      "\n",
      "Mandatory arguments to long options are mandatory for short options too.\n",
      "\n",
      "Startup:\n",
      "  -V,  --version                   display the version of Wget and exit\n",
      "  -h,  --help                      print this help\n",
      "  -b,  --background                go to background after startup\n",
      "  -e,  --execute=COMMAND           execute a `.wgetrc'-style command\n",
      "\n",
      "Logging and input file:\n",
      "  -o,  --output-file=FILE          log messages to FILE\n",
      "  -a,  --append-output=FILE        append messages to FILE\n",
      "  -d,  --debug                     print lots of debugging information\n",
      "  -q,  --quiet                     quiet (no output)\n",
      "  -v,  --verbose                   be verbose (this is the default)\n",
      "  -nv, --no-verbose                turn off verboseness, without being quiet\n",
      "       --report-speed=TYPE         output bandwidth as TYPE.  TYPE can be bits\n",
      "  -i,  --input-file=FILE           download URLs found in local or external FILE\n",
      "  -F,  --force-html                treat input file as HTML\n",
      "  -B,  --base=URL                  resolves HTML input-file links (-i -F)\n",
      "                                     relative to URL\n",
      "       --config=FILE               specify config file to use\n",
      "       --no-config                 do not read any config file\n",
      "       --rejected-log=FILE         log reasons for URL rejection to FILE\n",
      "\n",
      "Download:\n",
      "  -t,  --tries=NUMBER              set number of retries to NUMBER (0 unlimits)\n",
      "       --retry-connrefused         retry even if connection is refused\n",
      "       --retry-on-http-error=ERRORS    comma-separated list of HTTP errors to retry\n",
      "  -O,  --output-document=FILE      write documents to FILE\n",
      "  -nc, --no-clobber                skip downloads that would download to\n",
      "                                     existing files (overwriting them)\n",
      "       --no-netrc                  don't try to obtain credentials from .netrc\n",
      "  -c,  --continue                  resume getting a partially-downloaded file\n",
      "       --start-pos=OFFSET          start downloading from zero-based position OFFSET\n",
      "       --progress=TYPE             select progress gauge type\n",
      "       --show-progress             display the progress bar in any verbosity mode\n",
      "  -N,  --timestamping              don't re-retrieve files unless newer than\n",
      "                                     local\n",
      "       --no-if-modified-since      don't use conditional if-modified-since get\n",
      "                                     requests in timestamping mode\n",
      "       --no-use-server-timestamps  don't set the local file's timestamp by\n",
      "                                     the one on the server\n",
      "  -S,  --server-response           print server response\n",
      "       --spider                    don't download anything\n",
      "  -T,  --timeout=SECONDS           set all timeout values to SECONDS\n",
      "       --dns-timeout=SECS          set the DNS lookup timeout to SECS\n",
      "       --connect-timeout=SECS      set the connect timeout to SECS\n",
      "       --read-timeout=SECS         set the read timeout to SECS\n",
      "  -w,  --wait=SECONDS              wait SECONDS between retrievals\n",
      "       --waitretry=SECONDS         wait 1..SECONDS between retries of a retrieval\n",
      "       --random-wait               wait from 0.5*WAIT...1.5*WAIT secs between retrievals\n",
      "       --no-proxy                  explicitly turn off proxy\n",
      "  -Q,  --quota=NUMBER              set retrieval quota to NUMBER\n",
      "       --bind-address=ADDRESS      bind to ADDRESS (hostname or IP) on local host\n",
      "       --limit-rate=RATE           limit download rate to RATE\n",
      "       --no-dns-cache              disable caching DNS lookups\n",
      "       --restrict-file-names=OS    restrict chars in file names to ones OS allows\n",
      "       --ignore-case               ignore case when matching files/directories\n",
      "  -4,  --inet4-only                connect only to IPv4 addresses\n",
      "  -6,  --inet6-only                connect only to IPv6 addresses\n",
      "       --prefer-family=FAMILY      connect first to addresses of specified family,\n",
      "                                     one of IPv6, IPv4, or none\n",
      "       --user=USER                 set both ftp and http user to USER\n",
      "       --password=PASS             set both ftp and http password to PASS\n",
      "       --ask-password              prompt for passwords\n",
      "       --use-askpass=COMMAND       specify credential handler for requesting \n",
      "                                     username and password.  If no COMMAND is \n",
      "                                     specified the WGET_ASKPASS or the SSH_ASKPASS \n",
      "                                     environment variable is used.\n",
      "       --no-iri                    turn off IRI support\n",
      "       --local-encoding=ENC        use ENC as the local encoding for IRIs\n",
      "       --remote-encoding=ENC       use ENC as the default remote encoding\n",
      "       --unlink                    remove file before clobber\n",
      "       --xattr                     turn on storage of metadata in extended file attributes\n",
      "\n",
      "Directories:\n",
      "  -nd, --no-directories            don't create directories\n",
      "  -x,  --force-directories         force creation of directories\n",
      "  -nH, --no-host-directories       don't create host directories\n",
      "       --protocol-directories      use protocol name in directories\n",
      "  -P,  --directory-prefix=PREFIX   save files to PREFIX/..\n",
      "       --cut-dirs=NUMBER           ignore NUMBER remote directory components\n",
      "\n",
      "HTTP options:\n",
      "       --http-user=USER            set http user to USER\n",
      "       --http-password=PASS        set http password to PASS\n",
      "       --no-cache                  disallow server-cached data\n",
      "       --default-page=NAME         change the default page name (normally\n",
      "                                     this is 'index.html'.)\n",
      "  -E,  --adjust-extension          save HTML/CSS documents with proper extensions\n",
      "       --ignore-length             ignore 'Content-Length' header field\n",
      "       --header=STRING             insert STRING among the headers\n",
      "       --compression=TYPE          choose compression, one of auto, gzip and none. (default: none)\n",
      "       --max-redirect              maximum redirections allowed per page\n",
      "       --proxy-user=USER           set USER as proxy username\n",
      "       --proxy-password=PASS       set PASS as proxy password\n",
      "       --referer=URL               include 'Referer: URL' header in HTTP request\n",
      "       --save-headers              save the HTTP headers to file\n",
      "  -U,  --user-agent=AGENT          identify as AGENT instead of Wget/VERSION\n",
      "       --no-http-keep-alive        disable HTTP keep-alive (persistent connections)\n",
      "       --no-cookies                don't use cookies\n",
      "       --load-cookies=FILE         load cookies from FILE before session\n",
      "       --save-cookies=FILE         save cookies to FILE after session\n",
      "       --keep-session-cookies      load and save session (non-permanent) cookies\n",
      "       --post-data=STRING          use the POST method; send STRING as the data\n",
      "       --post-file=FILE            use the POST method; send contents of FILE\n",
      "       --method=HTTPMethod         use method \"HTTPMethod\" in the request\n",
      "       --body-data=STRING          send STRING as data. --method MUST be set\n",
      "       --body-file=FILE            send contents of FILE. --method MUST be set\n",
      "       --content-disposition       honor the Content-Disposition header when\n",
      "                                     choosing local file names (EXPERIMENTAL)\n",
      "       --content-on-error          output the received content on server errors\n",
      "       --auth-no-challenge         send Basic HTTP authentication information\n",
      "                                     without first waiting for the server's\n",
      "                                     challenge\n",
      "\n",
      "HTTPS (SSL/TLS) options:\n",
      "       --secure-protocol=PR        choose secure protocol, one of auto, SSLv2,\n",
      "                                     SSLv3, TLSv1, TLSv1_1, TLSv1_2 and PFS\n",
      "       --https-only                only follow secure HTTPS links\n",
      "       --no-check-certificate      don't validate the server's certificate\n",
      "       --certificate=FILE          client certificate file\n",
      "       --certificate-type=TYPE     client certificate type, PEM or DER\n",
      "       --private-key=FILE          private key file\n",
      "       --private-key-type=TYPE     private key type, PEM or DER\n",
      "       --ca-certificate=FILE       file with the bundle of CAs\n",
      "       --ca-directory=DIR          directory where hash list of CAs is stored\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       --crl-file=FILE             file with bundle of CRLs\n",
      "       --pinnedpubkey=FILE/HASHES  Public key (PEM/DER) file, or any number\n",
      "                                   of base64 encoded sha256 hashes preceded by\n",
      "                                   'sha256//' and separated by ';', to verify\n",
      "                                   peer against\n",
      "       --random-file=FILE          file with random data for seeding the SSL PRNG\n",
      "\n",
      "       --ciphers=STR           Set the priority string (GnuTLS) or cipher list string (OpenSSL) directly.\n",
      "                                   Use with care. This option overrides --secure-protocol.\n",
      "                                   The format and syntax of this string depend on the specific SSL/TLS engine.\n",
      "HSTS options:\n",
      "       --no-hsts                   disable HSTS\n",
      "       --hsts-file                 path of HSTS database (will override default)\n",
      "\n",
      "FTP options:\n",
      "       --ftp-user=USER             set ftp user to USER\n",
      "       --ftp-password=PASS         set ftp password to PASS\n",
      "       --no-remove-listing         don't remove '.listing' files\n",
      "       --no-glob                   turn off FTP file name globbing\n",
      "       --no-passive-ftp            disable the \"passive\" transfer mode\n",
      "       --preserve-permissions      preserve remote file permissions\n",
      "       --retr-symlinks             when recursing, get linked-to files (not dir)\n",
      "\n",
      "FTPS options:\n",
      "       --ftps-implicit                 use implicit FTPS (default port is 990)\n",
      "       --ftps-resume-ssl               resume the SSL/TLS session started in the control connection when\n",
      "                                         opening a data connection\n",
      "       --ftps-clear-data-connection    cipher the control channel only; all the data will be in plaintext\n",
      "       --ftps-fallback-to-ftp          fall back to FTP if FTPS is not supported in the target server\n",
      "WARC options:\n",
      "       --warc-file=FILENAME        save request/response data to a .warc.gz file\n",
      "       --warc-header=STRING        insert STRING into the warcinfo record\n",
      "       --warc-max-size=NUMBER      set maximum size of WARC files to NUMBER\n",
      "       --warc-cdx                  write CDX index files\n",
      "       --warc-dedup=FILENAME       do not store records listed in this CDX file\n",
      "       --no-warc-compression       do not compress WARC files with GZIP\n",
      "       --no-warc-digests           do not calculate SHA1 digests\n",
      "       --no-warc-keep-log          do not store the log file in a WARC record\n",
      "       --warc-tempdir=DIRECTORY    location for temporary files created by the\n",
      "                                     WARC writer\n",
      "\n",
      "Recursive download:\n",
      "  -r,  --recursive                 specify recursive download\n",
      "  -l,  --level=NUMBER              maximum recursion depth (inf or 0 for infinite)\n",
      "       --delete-after              delete files locally after downloading them\n",
      "  -k,  --convert-links             make links in downloaded HTML or CSS point to\n",
      "                                     local files\n",
      "       --convert-file-only         convert the file part of the URLs only (usually known as the basename)\n",
      "       --backups=N                 before writing file X, rotate up to N backup files\n",
      "  -K,  --backup-converted          before converting file X, back up as X.orig\n",
      "  -m,  --mirror                    shortcut for -N -r -l inf --no-remove-listing\n",
      "  -p,  --page-requisites           get all images, etc. needed to display HTML page\n",
      "       --strict-comments           turn on strict (SGML) handling of HTML comments\n",
      "\n",
      "Recursive accept/reject:\n",
      "  -A,  --accept=LIST               comma-separated list of accepted extensions\n",
      "  -R,  --reject=LIST               comma-separated list of rejected extensions\n",
      "       --accept-regex=REGEX        regex matching accepted URLs\n",
      "       --reject-regex=REGEX        regex matching rejected URLs\n",
      "       --regex-type=TYPE           regex type (posix)\n",
      "  -D,  --domains=LIST              comma-separated list of accepted domains\n",
      "       --exclude-domains=LIST      comma-separated list of rejected domains\n",
      "       --follow-ftp                follow FTP links from HTML documents\n",
      "       --follow-tags=LIST          comma-separated list of followed HTML tags\n",
      "       --ignore-tags=LIST          comma-separated list of ignored HTML tags\n",
      "  -H,  --span-hosts                go to foreign hosts when recursive\n",
      "  -L,  --relative                  follow relative links only\n",
      "  -I,  --include-directories=LIST  list of allowed directories\n",
      "       --trust-server-names        use the name specified by the redirection\n",
      "                                     URL's last component\n",
      "  -X,  --exclude-directories=LIST  list of excluded directories\n",
      "  -np, --no-parent                 don't ascend to the parent directory\n",
      "\n",
      "Email bug reports, questions, discussions to <bug-wget@gnu.org>\n",
      "and/or open issues at https://savannah.gnu.org/bugs/?func=additem&group=wget.\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "wget --help # check the manual if you want"
   ]
  },
  {
   "cell_type": "raw",
   "id": "8d15006e",
   "metadata": {},
   "source": [
    "%%bash\n",
    "wget https://repo.anaconda.com/archive/Anaconda3-2023.07-0-Linux-x86_64.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff394966",
   "metadata": {},
   "source": [
    "Now, open the command line and check where in the folders' hierarchy you are. You can do that with the `pwd` command: **print current\n",
    "directory**. You can explore how to use it by adding the `--help` argument."
   ]
  },
  {
   "cell_type": "raw",
   "id": "9b02a867",
   "metadata": {},
   "source": [
    "%%bash\n",
    "pwd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "beb95ada",
   "metadata": {},
   "source": [
    "You are likely to be in a folder called **home**, pretty much your starting folder. But in the case of linux  subsystems within windows, that is bash's starting folder and not your main disk's home folder.  \n",
    "\n",
    "You need to **change  directory** to your **Downloads** folder to find the conda installer.  \n",
    "\n",
    "Then, **list** files to check if the installer is in that folder. The asterisk is used as a [wildcard] (https://tldp.org/LDP/GNU-Linux-Tools-Summary/html/x11655.htm).  \n",
    "\n",
    "To run the bash `*sh` script use the `bash` command. Follow the instructions, either pressing ENTER, space for  scrolling down the page, or typing \"yes\".  \n",
    "\n",
    "**Note:** always replace my username with your username e.g. mftorr should be replaced with your username."
   ]
  },
  {
   "cell_type": "raw",
   "id": "935dd8ba",
   "metadata": {},
   "source": [
    "# run it in the command line!\n",
    "\n",
    "cd /mnt/c/Users/mftorr/Downloads/\n",
    "ls Ana*sh\n",
    "bash Anaconda3-2023.07-0-Linux-x86_64.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f6f0a41",
   "metadata": {},
   "source": [
    "Once you are done, close the bash command line interface and open a new one."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "220c0417",
   "metadata": {},
   "source": [
    "## 2.2. Creating and deleting environments\n",
    "\n",
    "Conda allows you to create discrete, enclosed environments with specific versions of programs to ensure a smooth interaction. If you are only worried about one particular program, conda will look for the versions of other programs and dependencies (programs that other programs require for working correctly) that works best and avoids conflict.  \n",
    "\n",
    "Every time we run an analysis associated to a publication, we can export the environment to a file that other researchers can use to recreate the environment used for running said analysis. Better yet, if you need to re-install conda, you can quickly use that file to re-install your environments. Conda environments help increase the reproducibility and transparency of your research.  \n",
    "\n",
    "Most of this section is taken from the [conda](https://conda.io/projects/conda/en/latest/user-guide/getting-started.html) documentation, which you can check in more depth if you need to.  \n",
    "\n",
    "Let's create an environment from scratch. we ask `conda` to `create` a new environment with `-n bioevo` name. When prompted, reply 'y' to whether you want to install the packages needed (for now none). Then, activate the environment using `conda activate`.  \n",
    "\n",
    "Once active, you should see it in the left side of the prompt. **Anything that you install whilst an environment is active will only be available for that environment.**"
   ]
  },
  {
   "cell_type": "raw",
   "id": "628bf15d",
   "metadata": {},
   "source": [
    "%%bash\n",
    "# or copy and run in the command line\n",
    "\n",
    "conda create -n bioevo\n",
    "conda activate bioevo\n",
    "conda deactivate # deactivates the environmnet and takes you back to conda base\n",
    "\n",
    "# anything after a hastag symbol is ignored, not executed, and considered a comment."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cbe58d3a",
   "metadata": {},
   "source": [
    "If we want to install a particular library through conda, e.g. [VCFtools](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3137218/), you can check for the correct channel (locations where the packages are stored) within conda's repository website, e.g. [here](https://anaconda.org/bioconda/vcftools).  \n",
    "\n",
    "The commands `conda config --add channels` indicate conda that it needs to include the channles bioconda and conda-forge, locations where most (but not all) libraries for bioinformatics are sitting."
   ]
  },
  {
   "cell_type": "raw",
   "id": "ace4d1e9",
   "metadata": {},
   "source": [
    "%%bash\n",
    "# or copy and run in the command line\n",
    "\n",
    "conda config --add channels bioconda\n",
    "conda config --add channels conda-forge\n",
    "conda install -c bioconda vcftools"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2965d7b3",
   "metadata": {},
   "source": [
    "You can check the list of environments installed within conda and their folder location using the command `conda env list`.  \n",
    "\n",
    "Once you see the folder location, you can remove the environment by simply deleting the entire folder using `rm` and the arguments `-r`. The command will ask you if are sure you want to delete the entire folder.  \n",
    "\n",
    "**BE CAREFUL!** That command will remove folders, files, data, everything you have in there. Don't drink and `rm`. Alternatively (if you remember the name of your environments), you can use `conda remove`"
   ]
  },
  {
   "cell_type": "raw",
   "id": "ea293663",
   "metadata": {},
   "source": [
    "%%bash\n",
    "# or copy and run in the command line\n",
    "\n",
    "conda env list\n",
    "#rm -r /path/to/environment/folder\n",
    "\n",
    "conda remove --name bioevo --all"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b69b3265",
   "metadata": {},
   "source": [
    "## 2.3. Creating environments from files\n",
    "\n",
    "For reproducibility, sometimes it is desirable to install an environment from a file that someone else has exported. \n",
    "The process is as simple as using `conda env create`\n",
    "\n",
    "\n",
    "You can confirm that all packages and their dependencies were intalled by listing every component within the environment with `conda list` and the name of the environment. The name of a single (or more, separated by spaces) package can be added at the end, to specifically check its installation and version."
   ]
  },
  {
   "cell_type": "raw",
   "id": "6e572fc4",
   "metadata": {},
   "source": [
    "%%bash\n",
    "\n",
    "conda env create -f bioevo_environment.yml"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6161bb90",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# packages in environment at /home/mft/anaconda3/envs/bioevo:\n",
      "#\n",
      "# Name                    Version                   Build  Channel\n",
      "_libgcc_mutex             0.1                 conda_forge    conda-forge\n",
      "_openmp_mutex             4.5                       2_gnu    conda-forge\n",
      "alsa-lib                  1.2.8                h166bdaf_0    conda-forge\n",
      "anyio                     3.7.1              pyhd8ed1ab_0    conda-forge\n",
      "argon2-cffi               21.3.0             pyhd8ed1ab_0    conda-forge\n",
      "argon2-cffi-bindings      21.2.0          py311hd4cff14_3    conda-forge\n",
      "asttokens                 2.2.1              pyhd8ed1ab_0    conda-forge\n",
      "attr                      2.5.1                h166bdaf_1    conda-forge\n",
      "attrs                     23.1.0             pyh71513ae_1    conda-forge\n",
      "backcall                  0.2.0              pyh9f0ad1d_0    conda-forge\n",
      "backports                 1.0                pyhd8ed1ab_3    conda-forge\n",
      "backports.functools_lru_cache 1.6.5              pyhd8ed1ab_0    conda-forge\n",
      "bcftools                  1.17                 h3cc50cf_1    bioconda\n",
      "beautifulsoup4            4.12.2             pyha770c72_0    conda-forge\n",
      "blast                     2.14.0               h7d5a4b4_1    bioconda\n",
      "bleach                    6.0.0              pyhd8ed1ab_0    conda-forge\n",
      "brotli                    1.0.9                h166bdaf_9    conda-forge\n",
      "brotli-bin                1.0.9                h166bdaf_9    conda-forge\n",
      "bwa                       0.7.17              he4a0461_11    bioconda\n",
      "bzip2                     1.0.8                h7f98852_4    conda-forge\n",
      "c-ares                    1.19.1               hd590300_0    conda-forge\n",
      "ca-certificates           2023.5.7             hbcca054_0    conda-forge\n",
      "cairo                     1.16.0            hbbf8b49_1016    conda-forge\n",
      "certifi                   2023.5.7           pyhd8ed1ab_0    conda-forge\n",
      "cffi                      1.15.1          py311h409f033_3    conda-forge\n",
      "comm                      0.1.3              pyhd8ed1ab_0    conda-forge\n",
      "contourpy                 1.1.0           py311h9547e67_0    conda-forge\n",
      "curl                      8.1.2                h409715c_0    conda-forge\n",
      "cycler                    0.11.0             pyhd8ed1ab_0    conda-forge\n",
      "dbus                      1.13.6               h5008d03_3    conda-forge\n",
      "debugpy                   1.6.7           py311hcafe171_0    conda-forge\n",
      "decorator                 5.1.1              pyhd8ed1ab_0    conda-forge\n",
      "defusedxml                0.7.1              pyhd8ed1ab_0    conda-forge\n",
      "entrez-direct             16.2                 he881be0_1    bioconda\n",
      "entrypoints               0.4                pyhd8ed1ab_0    conda-forge\n",
      "exceptiongroup            1.1.2              pyhd8ed1ab_0    conda-forge\n",
      "executing                 1.2.0              pyhd8ed1ab_0    conda-forge\n",
      "expat                     2.5.0                hcb278e6_1    conda-forge\n",
      "flit-core                 3.9.0              pyhd8ed1ab_0    conda-forge\n",
      "font-ttf-dejavu-sans-mono 2.37                 hab24e00_0    conda-forge\n",
      "font-ttf-inconsolata      3.000                h77eed37_0    conda-forge\n",
      "font-ttf-source-code-pro  2.038                h77eed37_0    conda-forge\n",
      "font-ttf-ubuntu           0.83                 hab24e00_0    conda-forge\n",
      "fontconfig                2.14.2               h14ed4e7_0    conda-forge\n",
      "fonts-conda-ecosystem     1                             0    conda-forge\n",
      "fonts-conda-forge         1                             0    conda-forge\n",
      "fonttools                 4.41.0          py311h459d7ec_0    conda-forge\n",
      "freetype                  2.12.1               hca18f0e_1    conda-forge\n",
      "gettext                   0.21.1               h27087fc_0    conda-forge\n",
      "glib                      2.76.4               hfc55251_0    conda-forge\n",
      "glib-tools                2.76.4               hfc55251_0    conda-forge\n",
      "gmp                       6.2.1                h58526e2_0    conda-forge\n",
      "graphite2                 1.3.13            h58526e2_1001    conda-forge\n",
      "gsl                       2.7                  he838d99_0    conda-forge\n",
      "gst-plugins-base          1.22.3               he621ea3_0  \n",
      "gstreamer                 1.22.3               h5eee18b_0  \n",
      "harfbuzz                  7.3.0                hdb3a94d_0    conda-forge\n",
      "htslib                    1.17                 h81da01d_2    bioconda\n",
      "icu                       72.1                 hcb278e6_0    conda-forge\n",
      "idna                      3.4                pyhd8ed1ab_0    conda-forge\n",
      "importlib-metadata        6.8.0              pyha770c72_0    conda-forge\n",
      "importlib_metadata        6.8.0                hd8ed1ab_0    conda-forge\n",
      "importlib_resources       6.0.0              pyhd8ed1ab_1    conda-forge\n",
      "ipykernel                 6.24.0             pyh71e2992_0    conda-forge\n",
      "ipython                   8.14.0             pyh41d4057_0    conda-forge\n",
      "ipython_genutils          0.2.0                      py_1    conda-forge\n",
      "ipywidgets                8.0.7              pyhd8ed1ab_0    conda-forge\n",
      "jedi                      0.18.2             pyhd8ed1ab_0    conda-forge\n",
      "jinja2                    3.1.2              pyhd8ed1ab_1    conda-forge\n",
      "jsonschema                4.18.2             pyhd8ed1ab_0    conda-forge\n",
      "jsonschema-specifications 2023.6.1           pyhd8ed1ab_0    conda-forge\n",
      "jupyter                   1.0.0           py311h38be061_8    conda-forge\n",
      "jupyter_client            8.3.0              pyhd8ed1ab_0    conda-forge\n",
      "jupyter_console           6.6.3              pyhd8ed1ab_0    conda-forge\n",
      "jupyter_core              5.3.1           py311h38be061_0    conda-forge\n",
      "jupyter_events            0.6.3              pyhd8ed1ab_0    conda-forge\n",
      "jupyter_server            2.7.0              pyhd8ed1ab_0    conda-forge\n",
      "jupyter_server_terminals  0.4.4              pyhd8ed1ab_1    conda-forge\n",
      "jupyterlab_pygments       0.2.2              pyhd8ed1ab_0    conda-forge\n",
      "jupyterlab_widgets        3.0.8              pyhd8ed1ab_0    conda-forge\n",
      "keyutils                  1.6.1                h166bdaf_0    conda-forge\n",
      "kiwisolver                1.4.4           py311h4dd048b_1    conda-forge\n",
      "krb5                      1.20.1               h81ceb04_0    conda-forge\n",
      "lame                      3.100             h166bdaf_1003    conda-forge\n",
      "lcms2                     2.15                 haa2dc70_1    conda-forge\n",
      "ld_impl_linux-64          2.40                 h41732ed_0    conda-forge\n",
      "lerc                      4.0.0                h27087fc_0    conda-forge\n",
      "libblas                   3.9.0           17_linux64_openblas    conda-forge\n",
      "libbrotlicommon           1.0.9                h166bdaf_9    conda-forge\n",
      "libbrotlidec              1.0.9                h166bdaf_9    conda-forge\n",
      "libbrotlienc              1.0.9                h166bdaf_9    conda-forge\n",
      "libcap                    2.67                 he9d0100_0    conda-forge\n",
      "libcblas                  3.9.0           17_linux64_openblas    conda-forge\n",
      "libclang                  16.0.6          default_h1cdf331_0    conda-forge\n",
      "libclang13                16.0.6          default_h4d60ac6_0    conda-forge\n",
      "libcups                   2.3.3                h36d4200_3    conda-forge\n",
      "libcurl                   8.1.2                h409715c_0    conda-forge\n",
      "libdeflate                1.18                 h0b41bf4_0    conda-forge\n",
      "libedit                   3.1.20191231         he28a2e2_2    conda-forge\n",
      "libev                     4.33                 h516909a_1    conda-forge\n",
      "libevent                  2.1.12               hf998b51_1    conda-forge\n",
      "libexpat                  2.5.0                hcb278e6_1    conda-forge\n",
      "libffi                    3.4.2                h7f98852_5    conda-forge\n",
      "libflac                   1.4.3                h59595ed_0    conda-forge\n",
      "libgcc-ng                 13.1.0               he5830b7_0    conda-forge\n",
      "libgcrypt                 1.10.1               h166bdaf_0    conda-forge\n",
      "libgfortran-ng            13.1.0               h69a702a_0    conda-forge\n",
      "libgfortran5              13.1.0               h15d22d2_0    conda-forge\n",
      "libglib                   2.76.4               hebfc3b9_0    conda-forge\n",
      "libgomp                   13.1.0               he5830b7_0    conda-forge\n",
      "libgpg-error              1.47                 h71f35ed_0    conda-forge\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "libiconv                  1.17                 h166bdaf_0    conda-forge\n",
      "libidn2                   2.3.4                h166bdaf_0    conda-forge\n",
      "libjpeg-turbo             2.1.5.1              h0b41bf4_0    conda-forge\n",
      "liblapack                 3.9.0           17_linux64_openblas    conda-forge\n",
      "libllvm16                 16.0.6               h5cf9203_0    conda-forge\n",
      "libnghttp2                1.52.0               h61bc06f_0    conda-forge\n",
      "libnsl                    2.0.0                h7f98852_0    conda-forge\n",
      "libogg                    1.3.5                h27cfd23_1  \n",
      "libopenblas               0.3.23          pthreads_h80387f5_0    conda-forge\n",
      "libopus                   1.3.1                h7f98852_1    conda-forge\n",
      "libpng                    1.6.39               h753d276_0    conda-forge\n",
      "libpq                     15.3                 hbcd7760_1    conda-forge\n",
      "libsndfile                1.2.0                hb75c966_0    conda-forge\n",
      "libsodium                 1.0.18               h36c2ea0_1    conda-forge\n",
      "libsqlite                 3.42.0               h2797004_0    conda-forge\n",
      "libssh2                   1.11.0               h0841786_0    conda-forge\n",
      "libstdcxx-ng              13.1.0               hfd8a6a1_0    conda-forge\n",
      "libsystemd0               253                  h8c4010b_1    conda-forge\n",
      "libtiff                   4.5.1                h8b53f26_0    conda-forge\n",
      "libunistring              0.9.10               h7f98852_0    conda-forge\n",
      "libuuid                   2.38.1               h0b41bf4_0    conda-forge\n",
      "libvorbis                 1.3.7                h9c3ff4c_0    conda-forge\n",
      "libwebp-base              1.3.1                hd590300_0    conda-forge\n",
      "libxcb                    1.15                 h0b41bf4_0    conda-forge\n",
      "libxkbcommon              1.5.0                h5d7e998_3    conda-forge\n",
      "libxml2                   2.11.4               h0d562d8_0    conda-forge\n",
      "libzlib                   1.2.13               hd590300_5    conda-forge\n",
      "lz4-c                     1.9.4                hcb278e6_0    conda-forge\n",
      "markupsafe                2.1.3           py311h459d7ec_0    conda-forge\n",
      "matplotlib                3.7.2           py311h38be061_0    conda-forge\n",
      "matplotlib-base           3.7.2           py311h54ef318_0    conda-forge\n",
      "matplotlib-inline         0.1.6              pyhd8ed1ab_0    conda-forge\n",
      "mistune                   3.0.0              pyhd8ed1ab_0    conda-forge\n",
      "mpg123                    1.31.3               hcb278e6_0    conda-forge\n",
      "munkres                   1.1.4              pyh9f0ad1d_0    conda-forge\n",
      "mysql-common              8.0.33               hf1915f5_1    conda-forge\n",
      "mysql-libs                8.0.33               hca2cd23_1    conda-forge\n",
      "nbclassic                 1.0.0              pyhb4ecaf3_1    conda-forge\n",
      "nbclient                  0.8.0              pyhd8ed1ab_0    conda-forge\n",
      "nbconvert                 7.6.0              pyhd8ed1ab_0    conda-forge\n",
      "nbconvert-core            7.6.0              pyhd8ed1ab_0    conda-forge\n",
      "nbconvert-pandoc          7.6.0              pyhd8ed1ab_0    conda-forge\n",
      "nbformat                  5.9.1              pyhd8ed1ab_0    conda-forge\n",
      "ncurses                   6.4                  hcb278e6_0    conda-forge\n",
      "nest-asyncio              1.5.6              pyhd8ed1ab_0    conda-forge\n",
      "notebook                  6.5.4              pyha770c72_0    conda-forge\n",
      "notebook-shim             0.2.3              pyhd8ed1ab_0    conda-forge\n",
      "nspr                      4.35                 h27087fc_0    conda-forge\n",
      "nss                       3.89                 he45b914_0    conda-forge\n",
      "numpy                     1.25.1          py311h64a7726_0    conda-forge\n",
      "openjpeg                  2.5.0                hfec8fc6_2    conda-forge\n",
      "openssl                   3.1.1                hd590300_1    conda-forge\n",
      "overrides                 7.3.1              pyhd8ed1ab_0    conda-forge\n",
      "packaging                 23.1               pyhd8ed1ab_0    conda-forge\n",
      "pandas                    2.0.3           py311h320fe9a_1    conda-forge\n",
      "pandoc                    3.1.3                h32600fe_0    conda-forge\n",
      "pandocfilters             1.5.0              pyhd8ed1ab_0    conda-forge\n",
      "parso                     0.8.3              pyhd8ed1ab_0    conda-forge\n",
      "pcre                      8.45                 h9c3ff4c_0    conda-forge\n",
      "pcre2                     10.40                hc3806b6_0    conda-forge\n",
      "perl                      5.32.1          3_hd590300_perl5    conda-forge\n",
      "perl-archive-tar          2.40            pl5321hdfd78af_0    bioconda\n",
      "perl-carp                 1.38            pl5321hdfd78af_4    bioconda\n",
      "perl-common-sense         3.75            pl5321hdfd78af_0    bioconda\n",
      "perl-compress-raw-bzip2   2.201           pl5321h87f3376_1    bioconda\n",
      "perl-compress-raw-zlib    2.105           pl5321h87f3376_0    bioconda\n",
      "perl-encode               3.19            pl5321hec16e2b_1    bioconda\n",
      "perl-exporter             5.72            pl5321hdfd78af_2    bioconda\n",
      "perl-exporter-tiny        1.002002        pl5321hdfd78af_0    bioconda\n",
      "perl-extutils-makemaker   7.70            pl5321hd8ed1ab_0    conda-forge\n",
      "perl-io-compress          2.201           pl5321hdbdd923_2    bioconda\n",
      "perl-io-zlib              1.14            pl5321hdfd78af_0    bioconda\n",
      "perl-json                 4.10            pl5321hdfd78af_0    bioconda\n",
      "perl-json-xs              2.34            pl5321h4ac6f70_6    bioconda\n",
      "perl-list-moreutils       0.430           pl5321hdfd78af_0    bioconda\n",
      "perl-list-moreutils-xs    0.430           pl5321h031d066_2    bioconda\n",
      "perl-parent               0.236           pl5321hdfd78af_2    bioconda\n",
      "perl-pathtools            3.75            pl5321hec16e2b_3    bioconda\n",
      "perl-scalar-list-utils    1.62            pl5321hec16e2b_1    bioconda\n",
      "perl-types-serialiser     1.01            pl5321hdfd78af_0    bioconda\n",
      "pexpect                   4.8.0              pyh1a96a4e_2    conda-forge\n",
      "pickleshare               0.7.5                   py_1003    conda-forge\n",
      "pillow                    10.0.0          py311h0b84326_0    conda-forge\n",
      "pip                       23.1.2             pyhd8ed1ab_0    conda-forge\n",
      "pixman                    0.40.0               h36c2ea0_0    conda-forge\n",
      "pkgutil-resolve-name      1.3.10             pyhd8ed1ab_0    conda-forge\n",
      "platformdirs              3.8.1              pyhd8ed1ab_0    conda-forge\n",
      "ply                       3.11                       py_1    conda-forge\n",
      "prometheus_client         0.17.1             pyhd8ed1ab_0    conda-forge\n",
      "prompt-toolkit            3.0.39             pyha770c72_0    conda-forge\n",
      "prompt_toolkit            3.0.39               hd8ed1ab_0    conda-forge\n",
      "psutil                    5.9.5           py311h2582759_0    conda-forge\n",
      "pthread-stubs             0.4               h36c2ea0_1001    conda-forge\n",
      "ptyprocess                0.7.0              pyhd3deb0d_0    conda-forge\n",
      "pulseaudio-client         16.1                 hb77b528_4    conda-forge\n",
      "pure_eval                 0.2.2              pyhd8ed1ab_0    conda-forge\n",
      "pycparser                 2.21               pyhd8ed1ab_0    conda-forge\n",
      "pygments                  2.15.1             pyhd8ed1ab_0    conda-forge\n",
      "pyparsing                 3.0.9              pyhd8ed1ab_0    conda-forge\n",
      "pyqt                      5.15.7          py311ha74522f_3    conda-forge\n",
      "pyqt5-sip                 12.11.0         py311hcafe171_3    conda-forge\n",
      "python                    3.11.4          hab00c5b_0_cpython    conda-forge\n",
      "python-dateutil           2.8.2              pyhd8ed1ab_0    conda-forge\n",
      "python-fastjsonschema     2.17.1             pyhd8ed1ab_0    conda-forge\n",
      "python-json-logger        2.0.7              pyhd8ed1ab_0    conda-forge\n",
      "python-tzdata             2023.3             pyhd8ed1ab_0    conda-forge\n",
      "python_abi                3.11                    3_cp311    conda-forge\n",
      "pytz                      2023.3             pyhd8ed1ab_0    conda-forge\n",
      "pyyaml                    6.0             py311hd4cff14_5    conda-forge\n",
      "pyzmq                     25.1.0          py311h75c88c4_0    conda-forge\n",
      "qt-main                   5.15.8              h01ceb2d_12    conda-forge\n",
      "qtconsole                 5.4.3              pyhd8ed1ab_0    conda-forge\n",
      "qtconsole-base            5.4.3              pyha770c72_0    conda-forge\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "qtpy                      2.3.1              pyhd8ed1ab_0    conda-forge\n",
      "readline                  8.2                  h8228510_1    conda-forge\n",
      "referencing               0.29.1             pyhd8ed1ab_0    conda-forge\n",
      "rfc3339-validator         0.1.4              pyhd8ed1ab_0    conda-forge\n",
      "rfc3986-validator         0.1.1              pyh9f0ad1d_0    conda-forge\n",
      "rpds-py                   0.8.10          py311h46250e7_0    conda-forge\n",
      "samtools                  1.17                 hd87286a_1    bioconda\n",
      "send2trash                1.8.2              pyh41d4057_0    conda-forge\n",
      "setuptools                68.0.0             pyhd8ed1ab_0    conda-forge\n",
      "sip                       6.7.9           py311hb755f60_0    conda-forge\n",
      "six                       1.16.0             pyh6c4a22f_0    conda-forge\n",
      "sniffio                   1.3.0              pyhd8ed1ab_0    conda-forge\n",
      "soupsieve                 2.3.2.post1        pyhd8ed1ab_0    conda-forge\n",
      "stack_data                0.6.2              pyhd8ed1ab_0    conda-forge\n",
      "terminado                 0.17.1             pyh41d4057_0    conda-forge\n",
      "tinycss2                  1.2.1              pyhd8ed1ab_0    conda-forge\n",
      "tk                        8.6.12               h27826a3_0    conda-forge\n",
      "toml                      0.10.2             pyhd8ed1ab_0    conda-forge\n",
      "tomli                     2.0.1              pyhd8ed1ab_0    conda-forge\n",
      "tornado                   6.3.2           py311h459d7ec_0    conda-forge\n",
      "traitlets                 5.9.0              pyhd8ed1ab_0    conda-forge\n",
      "typing-extensions         4.7.1                hd8ed1ab_0    conda-forge\n",
      "typing_extensions         4.7.1              pyha770c72_0    conda-forge\n",
      "typing_utils              0.1.0              pyhd8ed1ab_0    conda-forge\n",
      "tzdata                    2023c                h71feb2d_0    conda-forge\n",
      "wcwidth                   0.2.6              pyhd8ed1ab_0    conda-forge\n",
      "webencodings              0.5.1                      py_1    conda-forge\n",
      "websocket-client          1.6.1              pyhd8ed1ab_0    conda-forge\n",
      "wget                      1.20.3               ha35d2d1_1    conda-forge\n",
      "wheel                     0.40.0             pyhd8ed1ab_0    conda-forge\n",
      "widgetsnbextension        4.0.8              pyhd8ed1ab_0    conda-forge\n",
      "xcb-util                  0.4.0                hd590300_1    conda-forge\n",
      "xcb-util-image            0.4.0                h8ee46fc_1    conda-forge\n",
      "xcb-util-keysyms          0.4.0                h8ee46fc_1    conda-forge\n",
      "xcb-util-renderutil       0.3.9                hd590300_1    conda-forge\n",
      "xcb-util-wm               0.4.1                h8ee46fc_1    conda-forge\n",
      "xkeyboard-config          2.39                 hd590300_0    conda-forge\n",
      "xorg-kbproto              1.0.7             h7f98852_1002    conda-forge\n",
      "xorg-libice               1.1.1                hd590300_0    conda-forge\n",
      "xorg-libsm                1.2.4                h7391055_0    conda-forge\n",
      "xorg-libx11               1.8.6                h8ee46fc_0    conda-forge\n",
      "xorg-libxau               1.0.11               hd590300_0    conda-forge\n",
      "xorg-libxdmcp             1.1.3                h7f98852_0    conda-forge\n",
      "xorg-libxext              1.3.4                h0b41bf4_2    conda-forge\n",
      "xorg-libxrender           0.9.11               hd590300_0    conda-forge\n",
      "xorg-renderproto          0.11.1            h7f98852_1002    conda-forge\n",
      "xorg-xextproto            7.3.0             h0b41bf4_1003    conda-forge\n",
      "xorg-xf86vidmodeproto     2.3.1             h7f98852_1002    conda-forge\n",
      "xorg-xproto               7.0.31            h7f98852_1007    conda-forge\n",
      "xz                        5.2.6                h166bdaf_0    conda-forge\n",
      "yaml                      0.2.5                h7f98852_2    conda-forge\n",
      "zeromq                    4.3.4                h9c3ff4c_1    conda-forge\n",
      "zipp                      3.16.0             pyhd8ed1ab_1    conda-forge\n",
      "zlib                      1.2.13               hd590300_5    conda-forge\n",
      "zstd                      1.5.2                hfc55251_7    conda-forge\n",
      "# packages in environment at /home/mft/anaconda3/envs/bioevo:\n",
      "#\n",
      "# Name                    Version                   Build  Channel\n",
      "blast                     2.14.0               h7d5a4b4_1    bioconda\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "# check for packages\n",
    "conda list -n bioevo\n",
    "conda list -n bioevo blast"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f67b1f1d",
   "metadata": {},
   "source": [
    "## 2.4. Exporting environments\n",
    "\n",
    "You can export and share your own environment. To do so, make sure the environment is active before using `conda env export`.  \n",
    "\n",
    "Here, the `>` symbol re-directs the output to a new file called *bioevo_environment.yml*. Every time you run this command and re-direct the output to the file, the old file will be replaced by the new one (if both have the same name). This behaviour is typical of redirecting with `>` and we will cover that later."
   ]
  },
  {
   "cell_type": "raw",
   "id": "345242b1",
   "metadata": {},
   "source": [
    "%%bash\n",
    "\n",
    "conda activate bioevo # in case your environment is not active\n",
    "conda env export > bioevo_environmentv2.yml"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ec4eff8",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# [**3.  Basic Linux commands**](#commands)  \n",
    "\n",
    "We have set up an enviroment within which we can work, now the aim is to use biological data to learn basic Linux commands and how to use them within the context of biological data analysis.  \n",
    "\n",
    "First, we will gather data from [NCBI](https://www.ncbi.nlm.nih.gov/), one of the most widely used databases \"[of online resources for biological information and data](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6323993/)\".  \n",
    "\n",
    "We are also interested in exploring the data available for Lithuanian spiders on NCBI, using as a reference a species list published by [**Biteniekyt and Rlys (2011)**](https://lmaleidykla.lt/ojs/index.php/biologija/article/view/1926/828).  \n",
    "\n",
    "Our aim is to understand which spider species have data available, which don't, the type of data available and its quality. You are likely to use these commands often, following a similar data processing that will help you have an initial idea about data availability before you embark into your own projects."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf5638cf",
   "metadata": {},
   "source": [
    "## 3.1. Folder hierarchy\n",
    "\n",
    "Files are stored within folders (or directories, is the same) and folders are nested within other folders, following a hierarchy similar to a tree structure. Using the command line interface, we can navigate through folders by following the branches of the tree, up or down to different levels of the hierarchy.  \n",
    "\n",
    "We can express the **path** from one folder to another in two ways: absolute and relative. **Absolute paths** explicitely indicate thefolder names along the hierarchy path, from the current location to thedestination folder or file **Relative** paths express the steps alongthe hierarchy path to reach another destination and assume the user knowits current position *relative* to the rest of the structure.\n",
    "\n",
    "For example, assuming we arein the current folder:\n",
    "\n",
    "`C/Users/mftorr/Documents/RandomFolder/`\n",
    "\n",
    "Then the absolute and relative paths to reach `C/Users/mftorr/Documents/biol4evol/Lectures/Lecture0` are:  \n",
    "\n",
    "**Absolute:** `C/Users/mftorr/Documents/biol4evol/Lectures/Lecture0`  \n",
    "\n",
    "\n",
    "**Relative:** `../biol4evol/Lectures/Lecture0`  \n",
    "\n",
    "Thus, to move around folders, the first step is to know your **current working directory**, easily done with `pwd`.\n",
    "\n",
    "It is also helpful to understand which files are inside your current folder by **listing** its items. The command `ls` comes in handy, particularly when specific argumens are used. For example, `ls -l` will list all items within a folder, printing to screen the [permissions of a file](https://en.wikipedia.org/wiki/File-system_permissions) (first column from left to right), size of the file (fifth column) and last-modified date. To access the same information in a \"human readable\" format, you can use `ls -lh`. If you want to explore more options for listing (or in general for other commands), you can use `ls --help`.\n",
    "\n",
    "**Advanced:** try using `ls` and Linux wildcards to filter which files to list or move.\n",
    "\n",
    "You can list the contents of a folder if you know the path. Once you confirm that a particular folder is the one you want to move to, you can **change directory** to that path using `cd`. You can also copy (leaves a copy behind) or move (does not, or replaces one version with another) files across folders. Finally, you can create new folders with `mkdir` and the name of the new folder."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "15265ad5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/c/Users/mftor/Documents/bioinfo4evol/practicals/lecture2\n",
      "LT_spider_list.txt\n",
      "Lecture2_command_line.ipynb\n",
      "Zelote_coordinates.csv\n",
      "Zelote_coordinates_sel.csv\n",
      "bioevo_environment.yml\n",
      "count_sequences.sh\n",
      "genera.txt\n",
      "genera_filtered.txt\n",
      "sequence1.gb\n",
      "sequence2.gb\n",
      "species.txt\n",
      "spiders.fasta\n",
      "spiders.gb\n",
      "total 37M\n",
      "-rwxrwxrwx 1 mft mft  29K Jul 17 15:42 LT_spider_list.txt\n",
      "-rwxrwxrwx 1 mft mft  55K Sep 13 13:04 Lecture2_command_line.ipynb\n",
      "-rwxrwxrwx 1 mft mft  23M Jul 26 10:21 Zelote_coordinates.csv\n",
      "-rwxrwxrwx 1 mft mft 5.1M Jul 26 14:03 Zelote_coordinates_sel.csv\n",
      "-rwxrwxrwx 1 mft mft 9.4K Jul 17 15:42 bioevo_environment.yml\n",
      "-rwxrwxrwx 1 mft mft  438 Aug  3 13:08 count_sequences.sh\n",
      "-rwxrwxrwx 1 mft mft 1.9K Jul 19 15:09 genera.txt\n",
      "-rwxrwxrwx 1 mft mft  104 Jul 19 15:14 genera_filtered.txt\n",
      "-rwxrwxrwx 1 mft mft  50K Jul 18 11:07 sequence1.gb\n",
      "-rwxrwxrwx 1 mft mft  18K Jul 18 09:47 sequence2.gb\n",
      "-rwxrwxrwx 1 mft mft  12K Jul 19 14:41 species.txt\n",
      "-rwxrwxrwx 1 mft mft 1.5M Jul 20 18:42 spiders.fasta\n",
      "-rwxrwxrwx 1 mft mft 7.2M Jul 20 18:43 spiders.gb\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "pwd # likely something like /home/user\n",
    "ls\n",
    "ls -lh"
   ]
  },
  {
   "cell_type": "raw",
   "id": "c324a164",
   "metadata": {},
   "source": [
    "%%bash\n",
    "# replace user with your username\n",
    "\n",
    "cd /mnt/c/Users/user/Documents/\n",
    "mkdir bioinfo4evol\n",
    "mkdir bioinfo4evol/practicals/lecture2/\n",
    "\n",
    "ls /mnt/c/Users/user/Documents/bioinfo4evol/* # lists all files and folders\n",
    "cd /mnt/c/Users/user/Documents/bioinfo4evol/practicals/lecture2 # change directory\n",
    "\n",
    "mkdir testfolder # makes a new directory/folder\n",
    "\n",
    "cp sequence1.gb testfolder/sequence1.gb # copy leaves a copy behind\n",
    "mv sequence2.gb testfolder # move deletes the copy at origin and creates it at destination"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2d51e2d",
   "metadata": {},
   "source": [
    "**What happened to file1 and file2?**"
   ]
  },
  {
   "cell_type": "raw",
   "id": "6fca9d76",
   "metadata": {},
   "source": [
    "%%bash\n",
    "ls -lh\n",
    "ls -lh testfolder/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c0047809",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-success\">\n",
    "<b>TO DO:</b> Familiarise yourself with the folders and files in your system. No need to delete anything, just move around and get used to the structure within your machine.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "106b66b6",
   "metadata": {},
   "source": [
    "## 3.2. Create, inspect, and delete files  \n",
    "\n",
    "Creating, inspecting, and deleting files is relatively easy. For creating a file that you are not writing to inmediately (just creating) `touch filename` is enough. That command will create (but not open) a file called *filename*"
   ]
  },
  {
   "cell_type": "raw",
   "id": "dcd7fa96",
   "metadata": {},
   "source": [
    "%%bash\n",
    "\n",
    "# replace \"name\" with your name without special characters\n",
    "touch answersL0_name.txt # creates the file"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c75eb37f",
   "metadata": {},
   "source": [
    "If you want to create and modify a file manually, you can do it using `nano filename`.\n",
    "\n",
    "[nano](https://en.wikipedia.org/wiki/GNU_nano) is a text editor for the command line interface and it is very useful for changing files in a non-systematic way. You can navigate through a file inside of nano using the keyboard arrows, the PfDn and PgUP keys, and the Home and End keys (and many more).\n",
    "\n",
    "You can write and modify text as usual, just keep in mind that clicking with a mouse is not an option. Once you are happy with the changes, simply follow the sequence for saving the changes in a file with the same name and exiting the editor: `ctrl+x` (Save modified buffer?) `y` (File Name to Write: filename.file) `enter`."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ba00b69",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-success\">\n",
    "<b>TO DO:</b> As an exercise, open the <i>answersL0_name.txt</i> in <i>nano</i>, write your name, save and close the file. <b>You will have to answer a few questions throughout this tutorial. For each question, open the <i>answersL0_name.txt</i> in <i>nano</i>, add the question and answers, save and close the file. You will submit your file by the end of the class.</b> </div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59a13a03",
   "metadata": {},
   "source": [
    "Other usefull commands for exploring files without modifiying them are `less`, `head`, and `tail`.\n",
    "\n",
    "`less` is the equivalent of opening the file \"in view mode\". You can navigate through the file with the keyboard arrows, PgUp and PgDn, and the spacebar, and use the kwy `q` to exit the view.\n",
    "\n",
    "`head` and `tail` show the first and last lines of a file, respectively. By default, both commands show 10 lines, but it is possible tomodify the number of lines with the `-n 50` argument to show, for example, 50 lines.\n",
    "\n",
    "For example, imagine a program throws you an error that says \"unrecognised character in line 928th of sequence1.gb\". How can you check that line? You can use `head -n 928` and **pipe** the output to `tail`. **\"Piping\"** (`|`) allows one process to communicate with a subsequent process, it simplifies and makes more efficient the analysis of large amounts of data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "4593c380",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                     /PCR_primers=\"fwd_seq: attcaaccaatcataaagatattgg, rev_seq:\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "head -n 928 sequence1.gb | tail -n 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7cdbb0d1",
   "metadata": {},
   "source": [
    "Two more symbols that are very important and omnipresent are `>` and `>>` for creating and appending to files, respectively.\n",
    "\n",
    "For example, `head -n 10 sequence1.gb > sequence1_n10.txt` creates a file called `sequence1_n10.txt` that contains the first 10 lines in the `sequence1.gb` file. If the file `sequence1_n10.txt` already exists, then **the command will replace the old file with the new one** conserving no information from the old file.\n",
    "\n",
    "If, instead of creating a new `sequence1_n10.txt` file we want to **append** additional information to it, we can use `>>`. For example, `head -n 10 sequence2.gb >> sequence1_n10.txt` will append the first 10 lines from the file `sequence2.gb` to the already existing `sequence1_n10.txt` file."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fdd77ae5",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-success\">\n",
    "<b>TO DO:</b> What is the 500th line in <i>sequence1.gb</i>? And the 500th line in <i>sequence2.gb</i>? Append the two lines to your <i>answersL0_name.txt</i> file.</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d4f7ce2",
   "metadata": {},
   "source": [
    "We will now start working with data on Lithuanian spiders. We have a species list from [Biteniekyt and Rlys (2011)](https://lmaleidykla.lt/ojs/index.php/biologija/article/view/1926/828).\n",
    "\n",
    "The `LT_spider_list.txt` file contains the information as directly copy-pasted from the publication's PDF. Our aim is to transform the unorganised information in `LT_spider_list.txt` into a list of species with taxonomic IDs that we will use to gather published sequences available online.\n",
    "\n",
    "Have a look at the `LT_spider_list.txt` file using `less`, `head`, and `tail`."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71691451",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-success\">\n",
    "    <b>TO DO: </b>What is the 20th line in <i>LT_spider_list.txt</i>? Append the line to your <i>answersL0_name.txt</i> file</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "901db06a",
   "metadata": {},
   "source": [
    "## 3.3. Matching, replacing, and counting patterns  \n",
    "\n",
    "You might have noticed a pattern from looking at the file. First you find a line starting with \"FAMILY\" for the spider family and then subsequent lines for every species in that family.\n",
    "\n",
    "We can use `grep` to print the spider families represented in Lithuania. `grep` is a command to search for **matching patterns** in a file. `grep` has very useful arguments that will help you extract key information from very large files, for example: `-v`, `-A`, `-B`. Familiarise yourself with `grep` and its arguments checking `--help`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "30e45c60",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage: grep [OPTION]... PATTERNS [FILE]...\n",
      "Search for PATTERNS in each FILE.\n",
      "Example: grep -i 'hello world' menu.h main.c\n",
      "PATTERNS can contain multiple patterns separated by newlines.\n",
      "\n",
      "Pattern selection and interpretation:\n",
      "  -E, --extended-regexp     PATTERNS are extended regular expressions\n",
      "  -F, --fixed-strings       PATTERNS are strings\n",
      "  -G, --basic-regexp        PATTERNS are basic regular expressions\n",
      "  -P, --perl-regexp         PATTERNS are Perl regular expressions\n",
      "  -e, --regexp=PATTERNS     use PATTERNS for matching\n",
      "  -f, --file=FILE           take PATTERNS from FILE\n",
      "  -i, --ignore-case         ignore case distinctions in patterns and data\n",
      "      --no-ignore-case      do not ignore case distinctions (default)\n",
      "  -w, --word-regexp         match only whole words\n",
      "  -x, --line-regexp         match only whole lines\n",
      "  -z, --null-data           a data line ends in 0 byte, not newline\n",
      "\n",
      "Miscellaneous:\n",
      "  -s, --no-messages         suppress error messages\n",
      "  -v, --invert-match        select non-matching lines\n",
      "  -V, --version             display version information and exit\n",
      "      --help                display this help text and exit\n",
      "\n",
      "Output control:\n",
      "  -m, --max-count=NUM       stop after NUM selected lines\n",
      "  -b, --byte-offset         print the byte offset with output lines\n",
      "  -n, --line-number         print line number with output lines\n",
      "      --line-buffered       flush output on every line\n",
      "  -H, --with-filename       print file name with output lines\n",
      "  -h, --no-filename         suppress the file name prefix on output\n",
      "      --label=LABEL         use LABEL as the standard input file name prefix\n",
      "  -o, --only-matching       show only nonempty parts of lines that match\n",
      "  -q, --quiet, --silent     suppress all normal output\n",
      "      --binary-files=TYPE   assume that binary files are TYPE;\n",
      "                            TYPE is 'binary', 'text', or 'without-match'\n",
      "  -a, --text                equivalent to --binary-files=text\n",
      "  -I                        equivalent to --binary-files=without-match\n",
      "  -d, --directories=ACTION  how to handle directories;\n",
      "                            ACTION is 'read', 'recurse', or 'skip'\n",
      "  -D, --devices=ACTION      how to handle devices, FIFOs and sockets;\n",
      "                            ACTION is 'read' or 'skip'\n",
      "  -r, --recursive           like --directories=recurse\n",
      "  -R, --dereference-recursive  likewise, but follow all symlinks\n",
      "      --include=GLOB        search only files that match GLOB (a file pattern)\n",
      "      --exclude=GLOB        skip files that match GLOB\n",
      "      --exclude-from=FILE   skip files that match any file pattern from FILE\n",
      "      --exclude-dir=GLOB    skip directories that match GLOB\n",
      "  -L, --files-without-match  print only names of FILEs with no selected lines\n",
      "  -l, --files-with-matches  print only names of FILEs with selected lines\n",
      "  -c, --count               print only a count of selected lines per FILE\n",
      "  -T, --initial-tab         make tabs line up (if needed)\n",
      "  -Z, --null                print 0 byte after FILE name\n",
      "\n",
      "Context control:\n",
      "  -B, --before-context=NUM  print NUM lines of leading context\n",
      "  -A, --after-context=NUM   print NUM lines of trailing context\n",
      "  -C, --context=NUM         print NUM lines of output context\n",
      "  -NUM                      same as --context=NUM\n",
      "      --color[=WHEN],\n",
      "      --colour[=WHEN]       use markers to highlight the matching strings;\n",
      "                            WHEN is 'always', 'never', or 'auto'\n",
      "  -U, --binary              do not strip CR characters at EOL (MSDOS/Windows)\n",
      "\n",
      "When FILE is '-', read standard input.  With no FILE, read '.' if\n",
      "recursive, '-' otherwise.  With fewer than two FILEs, assume -h.\n",
      "Exit status is 0 if any line (or file if -L) is selected, 1 otherwise;\n",
      "if any error occurs and -q is not given, the exit status is 2.\n",
      "\n",
      "Report bugs to: bug-grep@gnu.org\n",
      "GNU grep home page: <http://www.gnu.org/software/grep/>\n",
      "General help using GNU software: <https://www.gnu.org/gethelp/>\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "grep --help"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d625ca21",
   "metadata": {},
   "source": [
    "First, let's check the families in the file.\n",
    "\n",
    "Then, use `wc`, or the **word count** command for counting the lines that match \"FAMILY\".\n",
    "\n",
    "Finally, you can count the number of unique lines that match \"FAMILY\" (i.e. the lines with the exact same characters, visible or \"invisible\") using `uniq -c`.\n",
    "\n",
    "Beware, before checking for unique lines, you must `sort` the lines."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "665e9374",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FAMILY  PHOLCIDAE\n",
      "FAMILY  SEGESTRIIDAE\n",
      "FAMILY  DYSDERIDAE\n",
      "FAMILY  MIMETIDAE\n",
      "FAMILY  ERESIDAE\n",
      "FAMILY  ULOBORIDAE\n",
      "FAMILY  THERIDIIDAE\n",
      "FAMILY  LINYPHIIDAE\n",
      "FAMILY  TETRAGNATHIDAE\n",
      "FAMILY  ARANEIDAE\n",
      "FAMILY  LYCOSIDAE\n",
      "FAMILY  PISAURIDAE\n",
      "FAMILY  OXYOPIDAE\n",
      "FAMILY  ZORIDAE\n",
      "FAMILY  AGELENIDAE\n",
      "FAMILY  CYBAEIDAE\n",
      "FAMILY  HAHNIIDAE\n",
      "FAMILY  DICTYNIDAE\n",
      "FAMILY  AMAUROBIIDAE\n",
      "FAMILY  MITURGIDAE\n",
      "FAMILY  ANYPHAENIDAE\n",
      "FAMILY  LIOCRANIDAE\n",
      "FAMILY  CLUBIONIDAE\n",
      "FAMILY  CORINNIDAE\n",
      "FAMILY  GNAPHOSIDAE\n",
      "FAMILY  SPARASSIDAE\n",
      "FAMILY  PHILODROMIDAE\n",
      "FAMILY  THOMISIDAE\n",
      "FAMILY SALTICIDAE\n",
      "29\n",
      "      1 FAMILY SALTICIDAE\n",
      "      1 FAMILY  AGELENIDAE\n",
      "      1 FAMILY  AMAUROBIIDAE\n",
      "      1 FAMILY  ANYPHAENIDAE\n",
      "      1 FAMILY  ARANEIDAE\n",
      "      1 FAMILY  CLUBIONIDAE\n",
      "      1 FAMILY  CORINNIDAE\n",
      "      1 FAMILY  CYBAEIDAE\n",
      "      1 FAMILY  DICTYNIDAE\n",
      "      1 FAMILY  DYSDERIDAE\n",
      "      1 FAMILY  ERESIDAE\n",
      "      1 FAMILY  GNAPHOSIDAE\n",
      "      1 FAMILY  HAHNIIDAE\n",
      "      1 FAMILY  LINYPHIIDAE\n",
      "      1 FAMILY  LIOCRANIDAE\n",
      "      1 FAMILY  LYCOSIDAE\n",
      "      1 FAMILY  MIMETIDAE\n",
      "      1 FAMILY  MITURGIDAE\n",
      "      1 FAMILY  OXYOPIDAE\n",
      "      1 FAMILY  PHILODROMIDAE\n",
      "      1 FAMILY  PHOLCIDAE\n",
      "      1 FAMILY  PISAURIDAE\n",
      "      1 FAMILY  SEGESTRIIDAE\n",
      "      1 FAMILY  SPARASSIDAE\n",
      "      1 FAMILY  TETRAGNATHIDAE\n",
      "      1 FAMILY  THERIDIIDAE\n",
      "      1 FAMILY  THOMISIDAE\n",
      "      1 FAMILY  ULOBORIDAE\n",
      "      1 FAMILY  ZORIDAE\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "# grep 'pattern' file\n",
    "# use \" instead of ' if the pattern is a variable (more on that later)\n",
    "grep 'FAMILY' LT_spider_list.txt\n",
    "grep 'FAMILY' LT_spider_list.txt | wc -l\n",
    "grep 'FAMILY' LT_spider_list.txt | sort | uniq -c"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8f73f64",
   "metadata": {},
   "source": [
    "I encourage you to check all the options for `wc`, `sort`, and `uniq`. These are commands that you will be using very frequently."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69eab3c4",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-success\">\n",
    "    <b>TO DO: </b>How many spider families are in <i>LT_spider_list.txt</i>? Append the line to your <i>answersL0_name.txt</i> file. What do <i>wc -l</i> and <i>uniq -c</i> do? Manually add your answer to <i>answersL0_name.txt</i> using nano.</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "170a5bac",
   "metadata": {},
   "source": [
    "We can use grep to search **regular expressions** (known as [regex](https://en.wikipedia.org/wiki/Regular_expression)) to match patterns instead of exact words. For example, we can match a pattern that is made of two strings of alphabetic characters separated by a space character, each string can be of unlimited length as long as one alphabetic character comes after another."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "7c66db61",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "THE CHECKLIST OF SPECIES\n",
      "Pholcus phalangioides (Fuesslin, 1775). [5, 16].\n",
      "Segestria senoculata (Linnaeus, 1758). [34; 7, 17].\n",
      "Harpactea rubicunda (C. L. Koch, 1838). [34; 16, 18];\n",
      "Ero cambridgei Kulczynski, 1911. [20].\n",
      "Ero furcata (Villers, 1789). [3, 14, 16, 19].\n",
      "Eresus cinnaberinus (O. P.-Cambridge, 1872). [12, 21].\n",
      "Hyptiotes paradoxus C. L.Koch, 1834. [12].\n",
      "Crustulina guttata (Wider, 1834). [5, 9, 14].\n",
      "Cryptachaea riparia (Blackwall, 1834). [12].\n",
      "Enoplognatha ovata (Clerck, 1757). [34; 14, 16, 20];\n",
      "Enoplognatha thoracica (Hahn, 1833). [14, 27].\n",
      "Episinus angulatus (Blackwall, 1836). [5, 7].\n",
      "Episinus truncatus Latreille, 1809. [14].\n",
      "Euryopis flavomaculata (C. L. Koch, 1836). [56; 9,\n",
      "Keijia tincta (Walckenaer, 1802). [34; 14, 2325; 29\n",
      "Lasaeola prona (Menge, 1868). [5, 9] (as Dipoena).\n",
      "Lasaeola tristis (Hahn, 1833). [5, 12].\n",
      "Lessertia dentichelis (Simon, 1884). [12].\n",
      "Neottiura bimaculata (Linnaeus, 1767). [34; 14, 20, 23]\n",
      "Parasteatoda lunata (Clerck, 1757). [5, 16] (as Achaearanea).\n",
      "Parasteatoda simulans (Thorell, 1875). [5] (as Achaearanea).\n",
      "Pholcomma gibbum (Westring, 1851). [14, 27].\n",
      "Robertus arundineti (O. P.-Cambridge, 1871). [4, 6, 9].\n",
      "Robertus lividus (Blackwall, 1836). [4, 67; 911;\n",
      "Robertus lyrifer Holm, 1939. [9].\n",
      "Robertus neglectus (O. P.-Cambridge, 1871). [5, 14].\n",
      "Robertus scoticus Jackson, 1914. [5, 14, 16, 20].\n",
      "Robertus ungulatus Vogelsanger, 1944. [5, 14].\n",
      "Simitidion simile (C. L. Koch, 1836). [34; 2325] (as\n",
      "Steatoda albomaculata (De Geer, 1778). [4, 6].\n",
      "Steatoda bipunctata (Linnaeus, 1758). [34; 24] (as\n",
      "Steatoda castanea (Clerck, 1757). [34] (as Asagena);\n",
      "Steatoda grossa (C. L.Koch, 1838). [5].\n",
      "Steatoda phalerata (Panzer, 1801). [34; 26] (as Asagena).\n",
      "Theonoe minutissima (O. P.-Cambridge, 1879). [7, 9,\n",
      "Phyloneta impressa L. Koch, 1881. [34; 16, 19, 2326]\n",
      "Phyloneta sisyphia (Clerck, 1757). [4, 14] (as Theridion).\n",
      "Theridion mystaceum L.Koch, 1870. [9].\n",
      "Theridion pictum (Walckenaer, 1802). [34; 1920;\n",
      "Theridion pinastri L.Koch, 1872. [34; 23].\n",
      "Theridion varians Hahn, 1833. [34; 20, 2225].\n",
      "Abacoproeces saltum (L.Koch, 1872). [11].\n",
      "Agyneta cauta (O. P.-Cambridge, 1902). [7, 9, 11, 14\n",
      "Agyneta conigera (O. P.-Cambridge, 1863). [11, 1415].\n",
      "Agyneta decora (O. P.-Cambridge, 1871). [7, 9, 15, 31].\n",
      "Agyneta ramosa Jackson, 1912. [7].\n",
      "Agyneta subtilis (O. P.-Cambridge, 1863). [7, 11, 14].\n",
      "Anguliphantes angulipalpis (Westring, 1851). [14, 27]\n",
      "Aphileta misera (O. P.-Cambridge, 1882). [14, 31].\n",
      "Araeoncus humilis (Blackwall, 1841). [4, 6, 14].\n",
      "Bathyphantes approximatus (O. P.-Cambridge, 1871).\n",
      "Bathyphantes gracilis (Blackwall, 1841). [6, 9, 14, 31].\n",
      "Bathyphantes nigrinus (Westring, 1851). [14].\n",
      "Bathyphantes parvulus (Westring, 1851). [6, 27].\n",
      "Bathyphantes similis Kulcynski, 1894. [16].\n",
      "Bolyphantes alticeps (Sundevall, 1833). [34; 14, 24].\n",
      "Bolyphantes luteolus (Blackwall, 1833). [34; 24].\n",
      "Centromerita bicolor (Blackwall, 1833). [67; 10, 14].\n",
      "Centromerus arcanus (O. P.-Cambridge, 1873). [4, 7, 9,\n",
      "Centromerus brevivulvatus Dahl, 1912\n",
      "Centromerus incilium (L.Koch, 1881). [14, 27].\n",
      "Centromerus levitarsis (Simon, 1884). [4, 7, 9, 14, 18,\n",
      "Centromerus sylvaticus (Blackwall, 1841). [67; 911;\n",
      "Centromerus unidentatus Miller, 1958. [7, 31].\n",
      "Ceratinella brevipes (Westring, 1851). [27].\n",
      "Ceratinella brevis (Wider, 1834). [4, 7, 9, 14].\n",
      "Ceratinella scabrosa (O. P.-Cambridge, 1871). [4].\n",
      "Cnephalocotes obscurus (Blackwall, 1834). [9, 15, 31].\n",
      "Dicymbium nigrum (Blackwall, 1834). [4, 6, 1011].\n",
      "Dicymbium nigrum brevisetosum Locket, 1962. [14].\n",
      "Dicymbium tibiale (Blackwall, 1836). [4, 14].\n",
      "Diplocentria bidentata (Emerton, 1882). [7, 14].\n",
      "Diplocephalus connatus Bertkau, 1889. [10].\n",
      "Diplocephalus cristatus (Blackwall, 1833). [16].\n",
      "Diplocephalus dentatus Tullgren, 1955. [9].\n",
      "Diplocephalus latifrons (O. P.-Cambridge, 1863). [4, 14].\n",
      "Diplocephalus permixtus (O. P.-Cambridge, 1871). [27].\n",
      "Diplocephalus picinus (Blackwall, 1841). [6, 14].\n",
      "Diplostyla concolor (Wider, 1834). [6, 10, 14].\n",
      "Dismodicus bifrons (Blackwall, 1841). [14].\n",
      "Dismodicus elevatus (C. L.Koch, 1838). [34; 9, 2325].\n",
      "Drapetisca socialis (Sundevall, 1833). [34; 14, 23].\n",
      "Drepanotylus uncatus (O. P.-Cambridge, 1873). [6, 27,\n",
      "Erigone atra Blackwall, 1833. [34; 6, 10, 24, 31].\n",
      "Erigone dentipalpis (Wider, 1834). [34; 6, 10, 14, 24].\n",
      "Erigonella hiemalis (Blackwall, 1841). [14, 27].\n",
      "Erigonella ignobilis (O. P.-Cambridge, 1871). [6, 14].\n",
      "Erigone longipalpis (Sundevall, 1830). [34; 24].\n",
      "Floronia bucculenta (Clerck, 1757). [34; 7, 14, 24].\n",
      "Formiphantes lepthyphantiformis (Strand, 1907). [14]\n",
      "Glyphesis cottonae (La Touche, 1945). [11].\n",
      "Gnathonarium dentatum (Wider, 1834). [4, 6].\n",
      "Gonatium rubellum (Blackwall, 1841). [6, 27].\n",
      "Gonatium rubens (Blackwall, 1833). [4, 7, 9, 11, 14\n",
      "Gongylidiellum latebricola (O. P.-Cambridge, 1871). [4].\n",
      "Gongylidiellum murcidum Simon, 1884. [4, 9].\n",
      "Gongylidium rufipes (Linnaeus, 1758). [4].\n",
      "Helophora insignis (Blackwall, 1841). [4].\n",
      "Hylyphantes graminicola (Sundevall, 1830). [4, 16]; [3,\n",
      "Hypomma bituberculatum (Wider, 1834). [16].\n",
      "Hypomma cornutum (Blackwall, 1833). [34; 22].\n",
      "Hypselistes jacksoni (O. P.-Cambridge, 1902). [14, 31].\n",
      "Kaestneria dorsalis (Wider, 1834). [34; 16, 20, 2425].\n",
      "Kaestneria pullata (O. P.-Cambridge, 1863). [20].\n",
      "Lepthyphantes leprosus (Ohlert, 1865). [14].\n",
      "Linyphia triangularis (Clerck, 1757). [34; 9, 14, 23\n",
      "Lophomma punctatum (Blackwall, 1841). [9, 14, 27].\n",
      "Macrargus carpenteri (O. P.-Cambridge, 1894). [7, 9,\n",
      "Macrargus rufus (Wider, 1834). [7, 14, 16, 27].\n",
      "Mansuphantes mansuetus (Thorell, 1875). [7] (as Lepthyphanthes).\n",
      "Maro minutus O. P.-Cambridge, 1906. [9, 15, 31].\n",
      "Maso sundevalli (Westring, 1851). [4, 20].\n",
      "Megalepthyphantes nebulosus (Sundevall, 1830). [16];\n",
      "Mecynargus foveatus (Dahl, 1912). [10].\n",
      "Meioneta affinis (Kulczynski, 1898). [9, 11, 31].\n",
      "Meioneta fuscipalpa (C. L.Koch, 1836). [10].\n",
      "Meioneta innotabilis (O. P.-Cambridge, 1863). [3] (as\n",
      "Meioneta mossica Schikora, 1993. [7, 9, 31].\n",
      "Meioneta rurestris (C. L.Koch, 1836). [4, 67; 10].\n",
      "Metopobactrus prominulus (O. P.-Cambridge, 1872).\n",
      "Micrargus apertus (O. P.-Cambridge, 1871). [7, 9, 14\n",
      "Micrargus herbigradus (Blackwall, 1854). [67; 14, 27,\n",
      "Micrargus subaequalis (Westring, 1851). [6, 10].\n",
      "Minicia marginella (Wider, 1834). [6].\n",
      "Minyriolus pusillus (Wider, 1834). [4, 7, 11, 14, 16, 20].\n",
      "Microneta viaria (Blackwall, 1841). [34; 9, 14, 16, 23].\n",
      "Microlinyphia pusilla (Sundevall, 1830). [34; 10]; [23\n",
      "Mioxena blanda (Simon, 1884). [6].\n",
      "Neriene clathrata (Sundevall, 1830). [4, 14]; [3, 24] (as\n",
      "Neriene emphana (Walckenaer, 1842). [34; 14]; [23] (as\n",
      "Neriene marginata C. L. Koch, 1834. [34]; [23] (as\n",
      "Neriene montana (Clerck, 1757). [4, 14].\n",
      "Neriene peltata (Wider, 1834). [34; 14, 20]; [23] (as\n",
      "Neriene radiata (Walckenaer, 1842). [4, 7, 9, 14].\n",
      "Notioscopus sarcinatus (O. P.-Cambridge, 1872). [7, 14\n",
      "Oedothorax apicatus (Blackwall, 1850). [6, 10, 27].\n",
      "Oedothorax gibbosus (Blackwall, 1841). [4].\n",
      "Oedothorax retusus (Westring, 1851). [4].\n",
      "Oryphantes angulatus (O. P.-Cambridge, 1881). [67; 9,\n",
      "Palliduphantes insignis (O. P.-Cambridge, 1913). [6] (as\n",
      "Palliduphantes pallidus (O.P.-Cambridge, 1871). [10, 14]\n",
      "Pelecopsis elongata (Wider, 1834). [7].\n",
      "Pelecopsis parallela (Wider, 1834). [10].\n",
      "Pityohyphantes phrygianus (C. L. Koch, 1836). [4, 23];\n",
      "Pocadicnemis pumila (Blackwall, 1841). [67; 9, 11,\n",
      "Porrhomma microphthalmum (O. P.-Cambridge, 1871).\n",
      "Porrhomma oblitum (O. P.-Cambridge, 1871). [14, 16].\n",
      "Porrhomma pallidum Jackson, 1913. [7, 14].\n",
      "Porrhomma pygmaeum (Blackwall, 1834). [4, 14].\n",
      "Saaristoa abnormis (Blackwall, 1841). [7].\n",
      "Savignia frontata Blackwall, 1833. [4, 6, 9, 14].\n",
      "Silometopus reussi (Thorell, 1871). [27].\n",
      "Sintula cornigera (Blackwall, 1856). [9, 14, 31].\n",
      "Stemonyphantes lineatus (Linnaeus, 1758). [67; 910;\n",
      "Tallusia experta (O. P.-Cambridge, 1871). [67; 9, 14, 27,\n",
      "Tapinocyba biscissa (O. P.-Cambridge, 1872). [6, 18].\n",
      "Tapinocyba insecta (L.Koch, 1869). [4, 6, 9, 14].\n",
      "Tapinocyba pallens (O. P.-Cambridge, 1872). [4, 7, 9, 11,\n",
      "Tapinocyba praecox (O. P.-Cambridge, 1873). [34;\n",
      "Tapinopa longidens (Wider, 1834). [14].\n",
      "Taranucnus setosus (O. P.-Cambridge, 1863). [7, 9, 14,\n",
      "Thyreosthenius parasiticus (Westring, 1851). [14, 16].\n",
      "Tiso vagans (Blackwall, 1834). [14, 16, 27].\n",
      "Tmeticus affinis (Blackwall, 1855). [4].\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tenuiphantes alacris (Blackwall, 1853). [16]; [14] (as\n",
      "Tenuiphantes cristatus (Menge, 1866). [67; 9, 1415]\n",
      "Tenuiphantes flavipes (Blackwall, 1854). [34; 26] (as\n",
      "Tenuiphantes mengei (Kulczynski, 1887). [6, 9, 14, 27]\n",
      "Tenuiphantes tenebricola (Wider, 1834). [14, 20] (as Lepthyphanthes).\n",
      "Troxochrus scabriculus (Westring, 1851). [6, 11, 14].\n",
      "Walckenaeria acuminata Blackwall, 1833. [7].\n",
      "Walckenaeria alticeps (Denis, 1952). [7, 11, 1415;\n",
      "Walckenaeria antica (Wider, 1834). [6, 9, 11, 14, 27,\n",
      "Walckenaeria atrotibialis (O. P.-Cambridge, 1878). [7, 9,\n",
      "Walckenaeria capito Westring, 1861. [10].\n",
      "Walckenaeria cucullata (C. L. Koch, 1836). [34; 7, 11,\n",
      "Walckenaeria cuspidata Blackwall, 1833. [7, 9, 14].\n",
      "Walckenaeria dysderoides (Wider, 1834). [7, 27].\n",
      "Walckenaeria incisa (O. P.-Cambridge, 1871). [11].\n",
      "Walckenaeria karpinskii (O. P.-Cambridge, 1873). [9].\n",
      "Walckenaeria kochi (O. P.-Cambridge, 1872). [14, 27].\n",
      "Walckenaeria mitrata (Menge, 1868). [27].\n",
      "Walckeaneria nodosa O. P.-Cambridge, 1873. [7, 9, 31].\n",
      "Walckenaeria nudipalpis (Westring, 1851). [4, 67; 9,\n",
      "Walckenaeria obtusa Blackwall, 1836. [14].\n",
      "Walckenaeria unicornis O. P.-Cambridge, 1861. [6, 14].\n",
      "Walckenaeria vigilax (Blackwall, 1853). [6].\n",
      "Zornella cultrigera (L.Koch, 1879). [7].\n",
      "Metellina mengei (Blackwall, 1870). [3, 7]; [4] (as Meta)\n",
      "= reticulata mengei Wiehle, 1931. [32] (as Meta).\n",
      "Metellina merianae (Scopoli, 1763). [3, 16]; [14, 32] (as\n",
      "Metellina segmentata (Clerck, 1757). [3, 16]; [2324; 26,\n",
      "Pachygnatha clercki Sundevall, 1823. [34; 6, 910; 14,\n",
      "Pachygnatha degeeri Sundevall, 1830. [34; 67; 910;\n",
      "Pachygnatha listeri Sundevall, 1830. [34; 7, 9, 14, 24,\n",
      "Tetragnatha extensa (Linnaeus, 1758). [34; 16, 19, 23\n",
      "= T. solandrii Scopoli 1763. [19, 2425; 32].\n",
      "Tetragnatha montana Simon, 1874. [34; 16].\n",
      "The status of this species in Lithuania is not clear since\n",
      "all records of T. solandrii were wrongly assigned to T. montana in [3] and continued in [4]. The recent data [16] can\n",
      "not be verified.\n",
      "Tetragnatha nigrita Lendl, 1886. [34; 32].\n",
      "Tetragnatha obtusa C. L. Koch, 1837. [34; 19, 2324;\n",
      "Tetragnatha pinicola L. Koch, 1870. [34; 19, 2325;\n",
      "Tetragnatha striata L.Koch, 1862. [34; 32].\n",
      "Aculepeira ceropegia (Walckenaer, 1802). [3, 4], [32] (as\n",
      "Agalenatea redii (Scopoli, 1763). [4]; [3, 2325; 32] (as\n",
      "Araneus alsine (Walckenaer, 1802). [34; 24, 32].\n",
      "Araneus angulatus Clerck, 1757. [34; 7, 23, 32].\n",
      "Araneus diadematus Clerck, 1757. [34; 7, 20, 2324;\n",
      "Araneus marmoreus Clerck, 1757. [34; 20, 23].\n",
      "= raji betulae Sulzer, 1776. [32];\n",
      "= marmoreus pyramidatus Clerck, 1757. [3].\n",
      "Araneus quadratus Clerck, 1757. [34; 6, 20].\n",
      "Araneus sturmi (Hahn, 1831). [4, 2325; 32]; [3] (as\n",
      "Araneus triguttatus (Fabricius, 1793). [2325]; [34; 18]\n",
      "Araniella cucurbitina (Clerck, 1757). [34; 16]; [2425;\n",
      "Araniella displicata (Hentz, 1847). [34]; [26] (as Aranea).\n",
      "Argiope bruennichi (Scopoli, 1772). [13].\n",
      "Cercidia prominens (Westring, 1851). [34; 7, 9, 14, 19,\n",
      "Cyclosa conica (Pallas, 1772). [34; 2325, 32].\n",
      "Cyclosa oculata (Walckenaer, 1802). [5].\n",
      "Gibbaranea gibbosa (Walckenaer, 1802). [12].\n",
      "Hypsosinga pygmaea (Sundevall, 1831). [34; 14, 20];\n",
      "Hypsosinga sanquinea (C. L.Koch, 1844). [4, 9, 7].\n",
      "Larinioides cornutus (Clerck, 1757). [4, 16].\n",
      "Larinioides ixobolus (Thorell, 1873). [34]; [19, 24, 32]\n",
      "Larinioides patagiatus (Clerck, 1757). [34; 16, 20].\n",
      "Larinioides sclopetarius (Clerck, 1757). [34; 10, 18].\n",
      "Mangora acalypha (Walckenaer, 1802). [34; 7, 2325;\n",
      "Neoscona adianta (Walckenaer, 1802). [34]; [19, 23\n",
      "Nuctenea silvicultrix (C. L. Koch, 1835). [4, 20]; [3] (as\n",
      "Nuctenea umbratica (Clerck, 1757). [34; 7]; [19, 24] (as\n",
      "Singa albovittata Caporiacco, 1947. [4, 32]; [3] (as Hypsosinga).\n",
      "Singa hamata (Clerck, 1757). [34; 1920; 32]\n",
      "Singa nitidula C. L.Koch, 1844. [34; 32].\n",
      "Stroemiellus stroemi (Thorell, 1870). [3]; [19, 24, 32] (as\n",
      "Zygiella atrica (C. L. Koch, 1845). [34]; [32] (as Zilla)\n",
      "Acantholycoa lignaria (Clerck, 1757). [9, 31].\n",
      "Alopecosa aculeata (Clerck, 1757). [7, 11, 14, 27]; [34\n",
      "Alopecosa barbipes (Sundevall, 1833). [4]; [3435] (as\n",
      "Old records of T. barbipes (Sundevall, 1833) in [3435]\n",
      "were wrongly assigned to A. accentuata (Latreille,\n",
      "Alopecosa cuneata (Clerck, 1757). [34; 6, 10, 14]; [34\n",
      "Alopecosa cursor (Hahn, 1831). [34]; [3435] (as\n",
      "Alopecosa fabrilis (Clerck, 1757). [3]; [3435] (as Tarentula).\n",
      "Alopecosa inquilina (Clerck, 1757). [3435]; [34] (as\n",
      "Alopecosa mariae (Dahl, 1908). [34; 18]; [3435] (as\n",
      "Alopecosa pinetorum (Thorell, 1856). [34; 14, 31].\n",
      "Alopecosa pulverulenta (Clerck, 1757). [34; 67; 9, 11,\n",
      "Alopecosa trabalis (Clerck, 1757). [34]; [3435] (as\n",
      "Arctosa alpigena (Doleschall). [9, 31].\n",
      "Arctosa alpigena lamperti (Dahl, 1908). [9, 27, 31].\n",
      "Arctosa cinerea (Fabricius, 1777). [34; 3436].\n",
      "Arctosa leopardus (Sundevall, 1833). [34; 6, 3435].\n",
      "Arctosa perita (Latreille, 1799). [34; 3435].\n",
      "Arctosa stigmosa (Thorell, 1875). [34; 3435].\n",
      "Aulonia albimana (Walckenaer, 1805). [67; 9, 11, 15,\n",
      "Hygrolycosa rubrofasciata (Ohlert, 1865). [34; 9, 11,\n",
      "Pardosa agrestis (Westring, 1861). [34; 6, 10]; [19, 24,\n",
      "Pardosa agricola (Thorell, 1856). [34].\n",
      "Pardosa agricola fucicola (Dahl, 1908).\n",
      "= arenicola fucicola Dahl, 1908. [34] (as Lycosa).\n",
      "= agricola arenicola var. fucicola (Dahl, 1908). [3].\n",
      "Pardosa amentata (Clerck, 1757). [34; 6, 19, 24].\n",
      "Pardosa bifasciata (C. L.Koch, 1834). [5].\n",
      "Pardosa hyperborea (Thorell, 1872). [3, 4, 7, 15, 20,\n",
      "= hyperborea pusilla (Thorell, 1872). [3435] (as Lycosa).\n",
      "Pardosa lugubris (Walckenaer, 1802). [34; 67; 9, 11,\n",
      "Pardosa monticola (Clerck, 1757). [34]; [3435] (as\n",
      "Pardosa paludicola (Clerck, 1757). [34, 14]; [3435] (as\n",
      "Pardosa palustris (Linnaeus, 1758). [34; 6, 10, 14, 16].\n",
      "Pardosa prativaga (L. Koch, 1870). [34; 67; 910; 14,\n",
      "Pardosa pullata (Clerck, 1757). [34; 67; 911; 1416;\n",
      "Pardosa riparia (C. L.Koch, 1833). [34; 31].\n",
      "Pardosa schenkeli Lessert, 1904. [34].\n",
      "Pardosa sphagnicola (Dahl, 1908). [34; 6, 7, 9, 11, 14,\n",
      "= riparia sphagnicola Dahl, 1908. [34] (as Lycosa).\n",
      "Pirata hygrophilus Thorell, 1872. [34; 7, 9, 1011; 14,\n",
      "Pirata insularis Emerton, 1885. [5, 7, 9, 11, 1415; 31].\n",
      "Pirata latitans (Blackwall, 1841). [34; 14, 34].\n",
      "Pirata piraticus (Clerck, 1757). [3, 910; 19, 24, 31, 34\n",
      "Pirata piscatorius (Clerck, 1757). [34; 14, 1920; 31,\n",
      "Pirata tenuitarsis Simon, 1876. [56; 14].\n",
      "Pirata uliginosus (Thorell, 1856). [34; 7, 9, 11, 1415;\n",
      "Trochosa robusta (Simon, 1876). [34].\n",
      "Trochosa ruricola (De Geer, 1778). [34; 6, 911; 14, 16,\n",
      "Trochosa spinipalpis (F. O. P.-Cambridge, 1895). [34;\n",
      "Trochosa terricola Thorell, 1856. [34; 67; 11, 14, 16,\n",
      "Xerolycosa miniata (C. L.Koch, 1834). [6, 10, 27].\n",
      "Xerolycosa nemoralis (Westring, 1861). [34; 67; 14,\n",
      "Dolomedes fimbriatus (Clerck, 1757). [4, 7, 910; 14,\n",
      "Dolomedes plantarius (Clerck, 1757). [4, 12, 19, 2324];\n",
      "Pisaura mirabilis (Clerck, 1757). [34; 16, 20, 29, 34].\n",
      "Oxyopes ramosus (Martini & Goeze, 1778). [34; 16,\n",
      "Zora nemoralis (Blackwall, 1861). [34; 14, 3435].\n",
      "Zora silvestris Kulczynski, 1897. [7, 9, 31].\n",
      "Zora spinimana (Sundevall, 1833). [34; 67; 9, 11, 15,\n",
      "Agelena labyrinthica (Clerck, 1757). [34; 67; 23].\n",
      "Tegenaria agrestis (Walckenaer, 1802). [10].\n",
      "Tegenaria atrica C. L.Koch, 1843. [5, 10].\n",
      "Tegenaria domestica (Clerck, 1757). [34].\n",
      "= derhami Scopoli 1763. [24].\n",
      "Argyroneta aquatica (Clerck, 1757). [34; 19, 24], (Family Argyronetidae. [16]).\n",
      "Antistea elegans (Blackwall, 1841). [57; 9, 14].\n",
      "Cryphoeca silvicola (C. L.Koch, 1834). [5, 14, 20].\n",
      "Hahnia nava (Blackwall, 1841). [5, 10, 14].\n",
      "Hahnia pusilla C. L.Koch, 1841. [9, 11, 27].\n",
      "Hahnia ononidum Simon, 1875. [20].\n",
      "Archaeodictyna consecuta (O. P.-Cambridge, 1872). [4].\n",
      "Argenna patula (Simon, 1874). [4].\n",
      "Argenna subnigra (O. P.-Cambridge, 1861). [5].\n",
      "Cicurina cicur (Fabricius, 1793). [57; 910; 14].\n",
      "Dictyna arundinacea (Linnaeus, 1758). [34; 7, 20,\n",
      "Dictyna pusilla Thorell, 1856. [34; 23].\n",
      "Dictyna uncinata Thorell, 1856. [4, 16].\n",
      "Amaurobius fenestralis (Strm, 1768). [12].\n",
      "Cheiracanthium erraticum (Walckenaer, 1802). (Family\n",
      "Cheiracanthium montanum L.Koch, 1877. [18]. (Family\n",
      "Cheiracanthium punctorium (Villers, 1789). (Family\n",
      "Cheiracanthium virescens (Sundevall, 1833). (Family\n",
      "Anyphaena accentuata (Walckenaer, 1802). [34; 14, 17].\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Agraecina striata (Kulczynski, 1882). [8].\n",
      "Agroeca brunnea (Blackwall, 1833). [34; 69; 14, 17,\n",
      "Agroeca cuprea Menge, 1873. [3, 8, 17].\n",
      "Agroeca dentigera Kulczyski, 1913. [89; 15, 31].\n",
      "Agroeca lusatica (L.Koch, 1875). [8].\n",
      "Agroeca proxima (O. P.-Cambridge, 1871). [59; 11, 14,\n",
      "Scotina palliardi (L.Koch, 1881). [69, 11, 15, 31].\n",
      "Clubiona caerulescens L.Koch, 1867. [34; 1617]\n",
      "Clubiona comta C. L.Koch, 1839. [34; 1617].\n",
      "Clubiona diversa O. P.-Cambridge, 1862. [34; 6, 9,\n",
      "Clubiona frutetorum L.Koch, 1866. [34; 17].\n",
      "Clubiona germanica Thorell, 1871. [34; 17, 22].\n",
      "Clubiona juvenis Simon, 1878. [27].\n",
      "Clubiona lutescens Westring, 1851. [34; 1617; 20].\n",
      "Clubiona marmorata L.Koch, 1866. [34; 1718].\n",
      "Clubiona neglecta O. P.-Cambridge, 1862. [ 34; 6, 10,\n",
      "Clubiona norvegica Strand, 1900. [27].\n",
      "Clubiona pallidula (Clerck, 1757). [34; 6, 16, 22, 26].\n",
      "Clubiona phragmitis C. L. Koch, 1843. [34; 6, 1617;\n",
      "Clubiona reclusa O. P.-Cambridge, 1863. [34; 17].\n",
      "Clubiona similis L.Koch, 1867. [34; 10, 17].\n",
      "Clubiona stagnatilis Kulczynski, 1897. [34; 7, 16, 19].\n",
      "Clubiona subsultans Thorell, 1875. [4, 7, 14, 16].\n",
      "Clubiona trivialis C. L. Koch, 1843. [4, 1617; 2325;\n",
      "Phrurolithus festivus (C. L.Koch, 1835). [4, 910; 1415;\n",
      "Phrurolithus minimus C. L. Koch, 1839. [7, 9; 15, 31];\n",
      "Berlandina cinerea (Menge, 1868). [34; 8]; [17] (as\n",
      "Callilepis nocturna (Linnaeus, 1758). [12].\n",
      "Drassodes hypocrita (Simon, 1878). [8].\n",
      "Drassodes pubescens (Thorell, 1856). [34; 69; 15, 17,\n",
      "Drassodes villosus (Thorell, 1856). [8].\n",
      "Drassyllus lutetianus (L.Koch, 1866). [59; 14, 31].\n",
      "Gnaphosa bicolor (Hahn, 1831). [34; 8, 17\n",
      "Gnaphosa lugubris (C. L.Koch, 1839). [12].\n",
      "Gnaphosa microps Holm, 1939. [79; 31].\n",
      "Gnaphosa montana (L.Koch, 1866). [34; 8, 17].\n",
      "Gnaphosa muscorum (L.Koch, 1866). [5, 78].\n",
      "Gnaphosa nigerrima L.Koch, 1877. [5, 79; 14, 31].\n",
      "Haplodrassus cognatus (Westring, 1861). [34; 8, 14];\n",
      "Haplodrassus dalmatensis (L.Koch, 1866). [5, 8].\n",
      "Haplodrassus moderatus (Kulczynski, 1897). [89; 31].\n",
      "Haplodrassus signifer (C. L. Koch, 1839). [34; 69; 11,\n",
      "Haplodrassus silvestris (Blackwall, 1833). [5, 89, 31].\n",
      "Haplodrassus soerenseni (Strand, 1900). [5, 79; 14, 31].\n",
      "Haplodrassus umbratilis (L. Koch, 1866). [4, 14, 8]; [17]\n",
      "Micaria fulgens (Walckenaer, 1802). [4, 8]; [17]; (Family\n",
      "Micaria lenzi Bsenberg, 1899. [5, 8].\n",
      "Micaria pulicaria (Sundevall, 1831). [4; 6, 810; 17];\n",
      "Micaria silesiaca L.Koch, 1875. [56; 8].\n",
      "Micaria subopaca Westring, 1862. [4, 8]; (Family Micariidae. [3]).\n",
      "Scotophaeus blackwalli (Thorell, 1871).\n",
      "An old record as S. gotlandicus in [17] was wrongly assigned to S. quadripunctatus (Linnaeus, 1758) in [3]\n",
      "and was followed in [4, 8].\n",
      "Scotophaeus scutulatus (L.Koch, 1866). [34; 8, 1718].\n",
      "Sosticus loricatus (L. Koch, 1866). [34; 8]; [17] (as Scotophaeus).\n",
      "Zelotes aeneus (Simon, 1878). [6, 8, 10].\n",
      "Zelotes electus (C. L.Koch, 1839). [34; 6, 8, 10, 14]; (Family Drassidae. [26]).\n",
      "Zelotes exiguus (Mller & Schenkel, 1895). [8, 27].\n",
      "Zelotes clivicola (L.Koch, 1870). [711; 14, 27, 31].\n",
      "Zelotes latreillei (Simon, 1878). [34; 69; 14, 17, 31].\n",
      "Zelotes longipes (L.Koch, 1866). [5, 9, 8].\n",
      "Zelotes petrensis (C. L.Koch, 1839). [34; 78; 14, 17].\n",
      "Drassyllus praeficus (L.Koch, 1866). [56; 8].\n",
      "Drassylus pusillus (C. L. Koch, 1833). [69; 14, 31]; [4]\n",
      "Zelotes subterraneus (C. L.Koch, 1833). [34; 79; 17].\n",
      "Micromata virescens (Clerck, 1757). [11].\n",
      "Philodromus aureolus (Clerck, 1757). [34; 19, 2226].\n",
      "Philodromus cespitum (Walckenaer, 1802). [7, 16, 20].\n",
      "Philodromus collinus C. L.Koch, 1835. [5, 14].\n",
      "Philodromus emarginatus (Schrank, 1803). [34; 19,\n",
      "Philodromus fuscomarginatus (De Geer, 1778). [34; 24,\n",
      "Philodromus histrio (Latreille, 1819). [4, 6].\n",
      "Philodromus margaritatus (Clerck, 1757). [12].\n",
      "Philodromus poecilus (Thorell, 1872). [34; 2425].\n",
      "Thanatus arenarius L.Koch, 1872. [10].\n",
      "Thanatus formicinus (Clerck, 1757). [4].\n",
      "Thanatus striatus C. L.Koch, 1845. [6].\n",
      "Tibellus maritimus (Menge, 1875). [34; 19, 24].\n",
      "Tibellus oblongus (Walckenaer, 1802). [34; 1920; 23,\n",
      "Coriarachne depressa (C. L.Koch, 1837). [5].\n",
      "Diaea dorsata (Fabricius, 1777). [4, 14].\n",
      "Misumena vatia (Clerck, 1757). [4, 16, 20].\n",
      "Misumenops tricuspidatus (Fabricius, 1775). [34; 18],\n",
      "Ozyptila atomaria (Panzer, 1801). [5].\n",
      "Ozyptila brevipes (Hahn, 1826). [34; 14, 1920].\n",
      "Ozyptila praticola (C. L.Koch, 1837). [5, 14].\n",
      "Ozyptila scabricula (Westring, 1851). [27].\n",
      "Ozyptila simplex (O. P.-Cambridge, 1862). [4].\n",
      "Ozyptila trux (Blackwall, 1846). [4, 14].\n",
      "Thomisus onustus Walckenaer, 1805. [34; 23].\n",
      "Xysticus audax (Schrank, 1803). [34].\n",
      "Xysticus bifasciatus C. L.Koch, 1837. [4, 14, 15].\n",
      "Xysticus cristatus (Clerck, 1757). [34; 67; 910; 14,\n",
      "Xysticus erraticus (Blackwall, 1834). [34; 7, 23].\n",
      "Xysticus kochi Thorell, 1872. [10, 27].\n",
      "Xysticus lanio C. L.Koch, 1835. [34; 6, 19, 24].\n",
      "Xysticus lineatus (Westring, 1851). [9].\n",
      "Xysticus luctuosus (Blackwall, 1836). [5, 7].\n",
      "Xysticus obscurus Collett, 1877. [7].\n",
      "Xysticus sabulosus (Hahn, 1832). [4, 6].\n",
      "Xysticus ulmi (Hahn, 1831). [34; 67; 9, 14, 1920].\n",
      "Asianellus festivus (C. L.Koch, 1834). [12].\n",
      "Ballus chalybeius (Walckenaer, 1802). [12].\n",
      "Dendryphantes hastatus (Clerck, 1757). [34; 2325].\n",
      "Dendryphantes rudis (Sundevall, 1833). [5].\n",
      "Euophrys frontalis (Walckenaer, 1802). [910; 14,\n",
      "Evarcha arcuata (Clerck, 1757). [34; 9, 14, 16, 20, 23\n",
      "Evarcha falcata (Clerck, 1757). [34; 9, 14, 16, 20, 24\n",
      "Evarcha laetabunda (C. L.Koch, 1846). [4, 9].\n",
      "Heliophanus auratus C. L.Koch, 1835. [5, 16].\n",
      "Heliophanus dampfi Schenkel, 1923. [27].\n",
      "Heliophanus dubius C. L. Koch, 1835. [34; 9, 24\n",
      "Heliophanus flavipes (Hahn, 1832). [5, 10].\n",
      "Marpissa muscosa (Clerck, 1757). [34; 19, 24].\n",
      "Marpissa radiata (Grube, 1859). [34; 16, 19, 24].\n",
      "Neon reticulatus (Blackwall, 1853). [5, 7, 9, 11, 15].\n",
      "Neon robustus Lohmander, 1945. [14].\n",
      "Neon valentulus Falconer, 1912. [5, 7, 9, 14].\n",
      "Pelenes tripunctatus (Walckenaer, 1802). [27].\n",
      "Philaeus chrysops (Poda, 1761). [5].\n",
      "Phlegra fasciata (Hahn, 1826). [5, 10].\n",
      "Pseudeuophrys erratica (Walckenaer, 1826). [5] (as Euophrys).\n",
      "Salticus cingulatus (Panzer, 1797). [4, 16].\n",
      "Salticus scenicus (Clerck, 1757). [4, 16].\n",
      "Sibianor aurocinctus (Ohlert, 1865). [5] (as Bianor).\n",
      "Sitticus caricis (Westring, 1861). [27].\n",
      "Sitticus distinguendus (Simon, 1868). [27].\n",
      "Sitticus dzieduszyckii (L.Koch, 1870). [5].\n",
      "Sitticus floricola (C. L.Koch, 1837). [34].\n",
      "Sitticus rupicola (C. L.Koch, 1837). [6].\n",
      "Sitticus saltator (O. P.-Cambridge, 1868). [5] (as Attulus).\n",
      "Sitticus terebratus (Clerck, 1757). [34; 23].\n",
      "Synageles hilarulus (C. L.Koch, 1846). [5].\n",
      "Synageles venator (Lucas, 1836). [34; 2425; 30].\n",
      "Talavera aequipes (O.P.-Cambridge, 1871). [10]; [27] (as\n",
      "Talavera petrensis (C. L. Koch, 1837). [9, 10]; [67] (as\n",
      "Talavera parvistyla Logunov, Kronestedt, 2003.\n",
      "Yllenus arenarius Menge, 1868. [27].\n",
      "THE CHECKLIST OF SPECIES\n",
      "Pholcus phalangioides (Fuesslin, 1775). [5, 16].\n",
      "Segestria senoculata (Linnaeus, 1758). [34; 7, 17].\n",
      "Harpactea rubicunda (C. L. Koch, 1838). [34; 16, 18];\n",
      "[17] (as Harpactocrates).\n",
      "Ero cambridgei Kulczynski, 1911. [20].\n",
      "Ero furcata (Villers, 1789). [3, 14, 16, 19].\n",
      "Eresus cinnaberinus (O. P.-Cambridge, 1872). [12, 21].\n",
      "Hyptiotes paradoxus C. L.Koch, 1834. [12].\n",
      "Crustulina guttata (Wider, 1834). [5, 9, 14].\n",
      "Cryptachaea riparia (Blackwall, 1834). [12].\n",
      "Enoplognatha ovata (Clerck, 1757). [34; 14, 16, 20];\n",
      "[2223] (as Theridium);\n",
      "= redimitum Clerck, 1757. [2426] (as Theridium).\n",
      "Enoplognatha thoracica (Hahn, 1833). [14, 27].\n",
      "Episinus angulatus (Blackwall, 1836). [5, 7].\n",
      "Episinus truncatus Latreille, 1809. [14].\n",
      "Euryopis flavomaculata (C. L. Koch, 1836). [56; 9,\n",
      "Keijia tincta (Walckenaer, 1802). [34; 14, 2325; 29\n",
      "30] (as Theridion).\n",
      "Lasaeola prona (Menge, 1868). [5, 9] (as Dipoena).\n",
      "Lasaeola tristis (Hahn, 1833). [5, 12].\n",
      "Lessertia dentichelis (Simon, 1884). [12].\n",
      "Neottiura bimaculata (Linnaeus, 1767). [34; 14, 20, 23]\n",
      "(as Theridium).\n",
      "Parasteatoda lunata (Clerck, 1757). [5, 16] (as Achaearanea).\n",
      "Parasteatoda simulans (Thorell, 1875). [5] (as Achaearanea).\n",
      "Pholcomma gibbum (Westring, 1851). [14, 27].\n",
      "Robertus arundineti (O. P.-Cambridge, 1871). [4, 6, 9].\n",
      "Robertus lividus (Blackwall, 1836). [4, 67; 911;\n",
      "Robertus lyrifer Holm, 1939. [9].\n",
      "Robertus neglectus (O. P.-Cambridge, 1871). [5, 14].\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Robertus scoticus Jackson, 1914. [5, 14, 16, 20].\n",
      "Robertus ungulatus Vogelsanger, 1944. [5, 14].\n",
      "Simitidion simile (C. L. Koch, 1836). [34; 2325] (as\n",
      "Steatoda albomaculata (De Geer, 1778). [4, 6].\n",
      "Steatoda bipunctata (Linnaeus, 1758). [34; 24] (as\n",
      "Steatoda castanea (Clerck, 1757). [34] (as Asagena);\n",
      "[24] (as Teutana).\n",
      "Steatoda grossa (C. L.Koch, 1838). [5].\n",
      "Steatoda phalerata (Panzer, 1801). [34; 26] (as Asagena).\n",
      "Theonoe minutissima (O. P.-Cambridge, 1879). [7, 9,\n",
      "Phyloneta impressa L. Koch, 1881. [34; 16, 19, 2326]\n",
      "(as Theridion).\n",
      "Phyloneta sisyphia (Clerck, 1757). [4, 14] (as Theridion).\n",
      "= notatum Linnaeus, 1758. [3, 2325] (as Theridion).\n",
      "Theridion mystaceum L.Koch, 1870. [9].\n",
      "Theridion pictum (Walckenaer, 1802). [34; 1920;\n",
      "Theridion pinastri L.Koch, 1872. [34; 23].\n",
      "Theridion varians Hahn, 1833. [34; 20, 2225].\n",
      "Abacoproeces saltum (L.Koch, 1872). [11].\n",
      "Agyneta cauta (O. P.-Cambridge, 1902). [7, 9, 11, 14\n",
      "Agyneta conigera (O. P.-Cambridge, 1863). [11, 1415].\n",
      "Agyneta decora (O. P.-Cambridge, 1871). [7, 9, 15, 31].\n",
      "Agyneta ramosa Jackson, 1912. [7].\n",
      "Agyneta subtilis (O. P.-Cambridge, 1863). [7, 11, 14].\n",
      "Anguliphantes angulipalpis (Westring, 1851). [14, 27]\n",
      "(as Lepthyphanthes).\n",
      "Aphileta misera (O. P.-Cambridge, 1882). [14, 31].\n",
      "Araeoncus humilis (Blackwall, 1841). [4, 6, 14].\n",
      "Bathyphantes approximatus (O. P.-Cambridge, 1871).\n",
      "Bathyphantes gracilis (Blackwall, 1841). [6, 9, 14, 31].\n",
      "Bathyphantes nigrinus (Westring, 1851). [14].\n",
      "Bathyphantes parvulus (Westring, 1851). [6, 27].\n",
      "Bathyphantes similis Kulcynski, 1894. [16].\n",
      "Bolyphantes alticeps (Sundevall, 1833). [34; 14, 24].\n",
      "Bolyphantes luteolus (Blackwall, 1833). [34; 24].\n",
      "Centromerita bicolor (Blackwall, 1833). [67; 10, 14].\n",
      "Centromerus arcanus (O. P.-Cambridge, 1873). [4, 7, 9,\n",
      "Centromerus brevivulvatus Dahl, 1912\n",
      "= aequalis C. L.Koch, 1841 [67; 9, 14].\n",
      "Centromerus incilium (L.Koch, 1881). [14, 27].\n",
      "Centromerus levitarsis (Simon, 1884). [4, 7, 9, 14, 18,\n",
      "Centromerus sylvaticus (Blackwall, 1841). [67; 911;\n",
      "Centromerus unidentatus Miller, 1958. [7, 31].\n",
      "Ceratinella brevipes (Westring, 1851). [27].\n",
      "Ceratinella brevis (Wider, 1834). [4, 7, 9, 14].\n",
      "Ceratinella scabrosa (O. P.-Cambridge, 1871). [4].\n",
      "Cnephalocotes obscurus (Blackwall, 1834). [9, 15, 31].\n",
      "Dicymbium nigrum (Blackwall, 1834). [4, 6, 1011].\n",
      "Dicymbium nigrum brevisetosum Locket, 1962. [14].\n",
      "Dicymbium tibiale (Blackwall, 1836). [4, 14].\n",
      "Diplocentria bidentata (Emerton, 1882). [7, 14].\n",
      "Diplocephalus connatus Bertkau, 1889. [10].\n",
      "Diplocephalus cristatus (Blackwall, 1833). [16].\n",
      "Diplocephalus dentatus Tullgren, 1955. [9].\n",
      "Diplocephalus latifrons (O. P.-Cambridge, 1863). [4, 14].\n",
      "Diplocephalus permixtus (O. P.-Cambridge, 1871). [27].\n",
      "Diplocephalus picinus (Blackwall, 1841). [6, 14].\n",
      "Diplostyla concolor (Wider, 1834). [6, 10, 14].\n",
      "Dismodicus bifrons (Blackwall, 1841). [14].\n",
      "Dismodicus elevatus (C. L.Koch, 1838). [34; 9, 2325].\n",
      "Drapetisca socialis (Sundevall, 1833). [34; 14, 23].\n",
      "Drepanotylus uncatus (O. P.-Cambridge, 1873). [6, 27,\n",
      "Erigone atra Blackwall, 1833. [34; 6, 10, 24, 31].\n",
      "Erigone dentipalpis (Wider, 1834). [34; 6, 10, 14, 24].\n",
      "Erigonella hiemalis (Blackwall, 1841). [14, 27].\n",
      "Erigonella ignobilis (O. P.-Cambridge, 1871). [6, 14].\n",
      "Erigone longipalpis (Sundevall, 1830). [34; 24].\n",
      "Floronia bucculenta (Clerck, 1757). [34; 7, 14, 24].\n",
      "Formiphantes lepthyphantiformis (Strand, 1907). [14]\n",
      "(as Lepthyphanthes).\n",
      "Glyphesis cottonae (La Touche, 1945). [11].\n",
      "Gnathonarium dentatum (Wider, 1834). [4, 6].\n",
      "Gonatium rubellum (Blackwall, 1841). [6, 27].\n",
      "Gonatium rubens (Blackwall, 1833). [4, 7, 9, 11, 14\n",
      "Gongylidiellum latebricola (O. P.-Cambridge, 1871). [4].\n",
      "Gongylidiellum murcidum Simon, 1884. [4, 9].\n",
      "Gongylidium rufipes (Linnaeus, 1758). [4].\n",
      "Helophora insignis (Blackwall, 1841). [4].\n",
      "Hylyphantes graminicola (Sundevall, 1830). [4, 16]; [3,\n",
      "2223] (as Erigonidium); [2425] (as Erigone).\n",
      "Hypomma bituberculatum (Wider, 1834). [16].\n",
      "Hypomma cornutum (Blackwall, 1833). [34; 22].\n",
      "Hypselistes jacksoni (O. P.-Cambridge, 1902). [14, 31].\n",
      "Kaestneria dorsalis (Wider, 1834). [34; 16, 20, 2425].\n",
      "Kaestneria pullata (O. P.-Cambridge, 1863). [20].\n",
      "Lepthyphantes leprosus (Ohlert, 1865). [14].\n",
      "Linyphia triangularis (Clerck, 1757). [34; 9, 14, 23\n",
      "Lophomma punctatum (Blackwall, 1841). [9, 14, 27].\n",
      "Macrargus carpenteri (O. P.-Cambridge, 1894). [7, 9,\n",
      "Macrargus rufus (Wider, 1834). [7, 14, 16, 27].\n",
      "Mansuphantes mansuetus (Thorell, 1875). [7] (as Lepthyphanthes).\n",
      "Maro minutus O. P.-Cambridge, 1906. [9, 15, 31].\n",
      "Maso sundevalli (Westring, 1851). [4, 20].\n",
      "Megalepthyphantes nebulosus (Sundevall, 1830). [16];\n",
      "[34; 14, 22] (as Lepthyphanthes).\n",
      "Mecynargus foveatus (Dahl, 1912). [10].\n",
      "Meioneta affinis (Kulczynski, 1898). [9, 11, 31].\n",
      "= beata O. P.-Cambridge, 1906. [6, 14].\n",
      "Meioneta fuscipalpa (C. L.Koch, 1836). [10].\n",
      "Meioneta innotabilis (O. P.-Cambridge, 1863). [3] (as\n",
      "Agyneta), [4, 26] (as Micryphantes).\n",
      "Meioneta mossica Schikora, 1993. [7, 9, 31].\n",
      "Meioneta rurestris (C. L.Koch, 1836). [4, 67; 10].\n",
      "Metopobactrus prominulus (O. P.-Cambridge, 1872).\n",
      "Micrargus apertus (O. P.-Cambridge, 1871). [7, 9, 14\n",
      "Micrargus herbigradus (Blackwall, 1854). [67; 14, 27,\n",
      "Micrargus subaequalis (Westring, 1851). [6, 10].\n",
      "Minicia marginella (Wider, 1834). [6].\n",
      "Minyriolus pusillus (Wider, 1834). [4, 7, 11, 14, 16, 20].\n",
      "Microneta viaria (Blackwall, 1841). [34; 9, 14, 16, 23].\n",
      "Microlinyphia pusilla (Sundevall, 1830). [34; 10]; [23\n",
      "24; 26] (as Linyphia).\n",
      "Mioxena blanda (Simon, 1884). [6].\n",
      "Neriene clathrata (Sundevall, 1830). [4, 14]; [3, 24] (as\n",
      "Neriene emphana (Walckenaer, 1842). [34; 14]; [23] (as\n",
      "Neriene marginata C. L. Koch, 1834. [34]; [23] (as\n",
      "Neriene montana (Clerck, 1757). [4, 14].\n",
      "Neriene peltata (Wider, 1834). [34; 14, 20]; [23] (as\n",
      "Neriene radiata (Walckenaer, 1842). [4, 7, 9, 14].\n",
      "Notioscopus sarcinatus (O. P.-Cambridge, 1872). [7, 14\n",
      "Oedothorax apicatus (Blackwall, 1850). [6, 10, 27].\n",
      "Oedothorax gibbosus (Blackwall, 1841). [4].\n",
      "Oedothorax retusus (Westring, 1851). [4].\n",
      "Oryphantes angulatus (O. P.-Cambridge, 1881). [67; 9,\n",
      "1415; 31] (as Lepthyphantes).\n",
      "Palliduphantes insignis (O. P.-Cambridge, 1913). [6] (as\n",
      "Palliduphantes pallidus (O.P.-Cambridge, 1871). [10, 14]\n",
      "(as Lepthyphanthes).\n",
      "Pelecopsis elongata (Wider, 1834). [7].\n",
      "Pelecopsis parallela (Wider, 1834). [10].\n",
      "Pityohyphantes phrygianus (C. L. Koch, 1836). [4, 23];\n",
      "[3] (as Linyphia).\n",
      "Pocadicnemis pumila (Blackwall, 1841). [67; 9, 11,\n",
      "Porrhomma microphthalmum (O. P.-Cambridge, 1871).\n",
      "Porrhomma oblitum (O. P.-Cambridge, 1871). [14, 16].\n",
      "Porrhomma pallidum Jackson, 1913. [7, 14].\n",
      "Porrhomma pygmaeum (Blackwall, 1834). [4, 14].\n",
      "Saaristoa abnormis (Blackwall, 1841). [7].\n",
      "Savignia frontata Blackwall, 1833. [4, 6, 9, 14].\n",
      "Silometopus reussi (Thorell, 1871). [27].\n",
      "Sintula cornigera (Blackwall, 1856). [9, 14, 31].\n",
      "Stemonyphantes lineatus (Linnaeus, 1758). [67; 910;\n",
      "Tallusia experta (O. P.-Cambridge, 1871). [67; 9, 14, 27,\n",
      "Tapinocyba biscissa (O. P.-Cambridge, 1872). [6, 18].\n",
      "Tapinocyba insecta (L.Koch, 1869). [4, 6, 9, 14].\n",
      "Tapinocyba pallens (O. P.-Cambridge, 1872). [4, 7, 9, 11,\n",
      "Tapinocyba praecox (O. P.-Cambridge, 1873). [34;\n",
      "Tapinopa longidens (Wider, 1834). [14].\n",
      "Taranucnus setosus (O. P.-Cambridge, 1863). [7, 9, 14,\n",
      "Thyreosthenius parasiticus (Westring, 1851). [14, 16].\n",
      "Tiso vagans (Blackwall, 1834). [14, 16, 27].\n",
      "Tmeticus affinis (Blackwall, 1855). [4].\n",
      "Tenuiphantes alacris (Blackwall, 1853). [16]; [14] (as\n",
      "Tenuiphantes cristatus (Menge, 1866). [67; 9, 1415]\n",
      "(as Lepthyphanthes).\n",
      "Tenuiphantes flavipes (Blackwall, 1854). [34; 26] (as\n",
      "Tenuiphantes mengei (Kulczynski, 1887). [6, 9, 14, 27]\n",
      "(as Lepthyphanthes).\n",
      "Tenuiphantes tenebricola (Wider, 1834). [14, 20] (as Lepthyphanthes).\n",
      "Troxochrus scabriculus (Westring, 1851). [6, 11, 14].\n",
      "Walckenaeria acuminata Blackwall, 1833. [7].\n",
      "Walckenaeria alticeps (Denis, 1952). [7, 11, 1415;\n",
      "Walckenaeria antica (Wider, 1834). [6, 9, 11, 14, 27,\n",
      "Walckenaeria atrotibialis (O. P.-Cambridge, 1878). [7, 9,\n",
      "Walckenaeria capito Westring, 1861. [10].\n",
      "Walckenaeria cucullata (C. L. Koch, 1836). [34; 7, 11,\n",
      "14]; [23] (as Wideria).\n",
      "Walckenaeria cuspidata Blackwall, 1833. [7, 9, 14].\n",
      "Walckenaeria dysderoides (Wider, 1834). [7, 27].\n",
      "Walckenaeria incisa (O. P.-Cambridge, 1871). [11].\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Walckenaeria karpinskii (O. P.-Cambridge, 1873). [9].\n",
      "Walckenaeria kochi (O. P.-Cambridge, 1872). [14, 27].\n",
      "Walckenaeria mitrata (Menge, 1868). [27].\n",
      "Walckeaneria nodosa O. P.-Cambridge, 1873. [7, 9, 31].\n",
      "Walckenaeria nudipalpis (Westring, 1851). [4, 67; 9,\n",
      "26, 31]; [3] (as Trachynella).\n",
      "Walckenaeria obtusa Blackwall, 1836. [14].\n",
      "Walckenaeria unicornis O. P.-Cambridge, 1861. [6, 14].\n",
      "Walckenaeria vigilax (Blackwall, 1853). [6].\n",
      "Zornella cultrigera (L.Koch, 1879). [7].\n",
      "Metellina mengei (Blackwall, 1870). [3, 7]; [4] (as Meta)\n",
      "(Family Metidae).\n",
      "= reticulata mengei Wiehle, 1931. [32] (as Meta).\n",
      "= reticulata Wiehle, 1931. [26] (as Meta).\n",
      "Metellina merianae (Scopoli, 1763). [3, 16]; [14, 32] (as\n",
      "Meta); [4] (as Meta) (Family Metidae).\n",
      "Metellina segmentata (Clerck, 1757). [3, 16]; [2324; 26,\n",
      "32] (as Meta); [4] (as Meta) (Family Metidae).\n",
      "Pachygnatha clercki Sundevall, 1823. [34; 6, 910; 14,\n",
      "Pachygnatha degeeri Sundevall, 1830. [34; 67; 910;\n",
      "Pachygnatha listeri Sundevall, 1830. [34; 7, 9, 14, 24,\n",
      "Tetragnatha extensa (Linnaeus, 1758). [34; 16, 19, 23\n",
      "= T. solandrii Scopoli 1763. [19, 2425; 32].\n",
      "Tetragnatha montana Simon, 1874. [34; 16].\n",
      "The status of this species in Lithuania is not clear since\n",
      "all records of T. solandrii were wrongly assigned to T. montana in [3] and continued in [4]. The recent data [16] can\n",
      "not be verified.\n",
      "Tetragnatha nigrita Lendl, 1886. [34; 32].\n",
      "Tetragnatha obtusa C. L. Koch, 1837. [34; 19, 2324;\n",
      "Tetragnatha pinicola L. Koch, 1870. [34; 19, 2325;\n",
      "Tetragnatha striata L.Koch, 1862. [34; 32].\n",
      "Aculepeira ceropegia (Walckenaer, 1802). [3, 4], [32] (as\n",
      "Agalenatea redii (Scopoli, 1763). [4]; [3, 2325; 32] (as\n",
      "Araneus); [26] (as Aranea).\n",
      "Araneus alsine (Walckenaer, 1802). [34; 24, 32].\n",
      "Araneus angulatus Clerck, 1757. [34; 7, 23, 32].\n",
      "Araneus diadematus Clerck, 1757. [34; 7, 20, 2324;\n",
      "= diadema Linnaeus, 1758. [26].\n",
      "Araneus marmoreus Clerck, 1757. [34; 20, 23].\n",
      "= raji Scopoli, 1763. [2426; 32];\n",
      "= raji betulae Sulzer, 1776. [32];\n",
      "= marmoreus pyramidatus Clerck, 1757. [3].\n",
      "Araneus quadratus Clerck, 1757. [34; 6, 20].\n",
      "Araneus sturmi (Hahn, 1831). [4, 2325; 32]; [3] (as\n",
      "Araneus triguttatus (Fabricius, 1793). [2325]; [34; 18]\n",
      "(as Atea).\n",
      "Araniella cucurbitina (Clerck, 1757). [34; 16]; [2425;\n",
      "32] (as Araneus); [23, 26] (as Aranea) (Family Argiopidae).\n",
      "Araniella displicata (Hentz, 1847). [34]; [26] (as Aranea).\n",
      "Argiope bruennichi (Scopoli, 1772). [13].\n",
      "Cercidia prominens (Westring, 1851). [34; 7, 9, 14, 19,\n",
      "Cyclosa conica (Pallas, 1772). [34; 2325, 32].\n",
      "Cyclosa oculata (Walckenaer, 1802). [5].\n",
      "Gibbaranea gibbosa (Walckenaer, 1802). [12].\n",
      "Hypsosinga pygmaea (Sundevall, 1831). [34; 14, 20];\n",
      "[19, 24, 32] (as Singa).\n",
      "Hypsosinga sanquinea (C. L.Koch, 1844). [4, 9, 7].\n",
      "Larinioides cornutus (Clerck, 1757). [4, 16].\n",
      "= folium Fourcroy, 1785. [3]; [19, 2425; 32] (as\n",
      "Larinioides ixobolus (Thorell, 1873). [34]; [19, 24, 32]\n",
      "(as Araneus).\n",
      "Larinioides patagiatus (Clerck, 1757). [34; 16, 20].\n",
      "= ocellatus Clerck, 1757. [19, 2326] (as Araneus).\n",
      "= duometorum Fourcroy, 1785. [32] (as Araneus).\n",
      "Larinioides sclopetarius (Clerck, 1757). [34; 10, 18].\n",
      "= sericatus Clerck, 1757. [19, 24] (as Araneus).\n",
      "= undatus Olivier, 1789. [32] (as Araneus).\n",
      "Mangora acalypha (Walckenaer, 1802). [34; 7, 2325;\n",
      "Neoscona adianta (Walckenaer, 1802). [34]; [19, 23\n",
      "25] (as Araneus).\n",
      "Nuctenea silvicultrix (C. L. Koch, 1835). [4, 20]; [3] (as\n",
      "Cyphepeira); [19] (as Aranea).\n",
      "Nuctenea umbratica (Clerck, 1757). [34; 7]; [19, 24] (as\n",
      "= sexpunctatus Linnaeus, 1758 [32] (as Araneus).\n",
      "Singa albovittata Caporiacco, 1947. [4, 32]; [3] (as Hypsosinga).\n",
      "Singa hamata (Clerck, 1757). [34; 1920; 32]\n",
      "Singa nitidula C. L.Koch, 1844. [34; 32].\n",
      "Stroemiellus stroemi (Thorell, 1870). [3]; [19, 24, 32] (as\n",
      "Zilla); [4] (as Zilla) (Family Metidae).\n",
      "Zygiella atrica (C. L. Koch, 1845). [34]; [32] (as Zilla)\n",
      "(Family Metidae).\n",
      "Zygiella x-notata (Clerck, 1757). [34] (as Zilla) (Family\n",
      "= litterata Olivier, 1789. [2326] (as Zilla).\n",
      "Acantholycoa lignaria (Clerck, 1757). [9, 31].\n",
      "Alopecosa aculeata (Clerck, 1757). [7, 11, 14, 27]; [34\n",
      "35] (as Tarentula).\n",
      "Alopecosa barbipes (Sundevall, 1833). [4]; [3435] (as\n",
      "Old records of T. barbipes (Sundevall, 1833) in [3435]\n",
      "were wrongly assigned to A. accentuata (Latreille,\n",
      "Alopecosa cuneata (Clerck, 1757). [34; 6, 10, 14]; [34\n",
      "35] (as Tarentula).\n",
      "Alopecosa cursor (Hahn, 1831). [34]; [3435] (as\n",
      "Alopecosa fabrilis (Clerck, 1757). [3]; [3435] (as Tarentula).\n",
      "Alopecosa inquilina (Clerck, 1757). [3435]; [34] (as\n",
      "Alopecosa mariae (Dahl, 1908). [34; 18]; [3435] (as\n",
      "Alopecosa pinetorum (Thorell, 1856). [34; 14, 31].\n",
      "= fumigata Dahl, 1908. [26] (as Tarentula).\n",
      "Alopecosa pulverulenta (Clerck, 1757). [34; 67; 9, 11,\n",
      "1415; 31]; [3435] (as Tarentula).\n",
      "Alopecosa trabalis (Clerck, 1757). [34]; [3435] (as\n",
      "Arctosa alpigena (Doleschall). [9, 31].\n",
      "Arctosa alpigena lamperti (Dahl, 1908). [9, 27, 31].\n",
      "Arctosa cinerea (Fabricius, 1777). [34; 3436].\n",
      "Arctosa leopardus (Sundevall, 1833). [34; 6, 3435].\n",
      "Arctosa perita (Latreille, 1799). [34; 3435].\n",
      "Arctosa stigmosa (Thorell, 1875). [34; 3435].\n",
      "Aulonia albimana (Walckenaer, 1805). [67; 9, 11, 15,\n",
      "Hygrolycosa rubrofasciata (Ohlert, 1865). [34; 9, 11,\n",
      "Pardosa agrestis (Westring, 1861). [34; 6, 10]; [19, 24,\n",
      "3435] (as Lycosa).\n",
      "Pardosa agricola (Thorell, 1856). [34].\n",
      "= fluviatilis Blackwall, 1861. [3435] (as Lycosa).\n",
      "Pardosa agricola fucicola (Dahl, 1908).\n",
      "= arenicola fucicola Dahl, 1908. [34] (as Lycosa).\n",
      "= agricola arenicola var. fucicola (Dahl, 1908). [3].\n",
      "Pardosa amentata (Clerck, 1757). [34; 6, 19, 24].\n",
      "= saccata Linnaeus, 1758. [3435] (as Lycosa).\n",
      "Pardosa bifasciata (C. L.Koch, 1834). [5].\n",
      "Pardosa hyperborea (Thorell, 1872). [3, 4, 7, 15, 20,\n",
      "= hyperborea pusilla (Thorell, 1872). [3435] (as Lycosa).\n",
      "Pardosa lugubris (Walckenaer, 1802). [34; 67; 9, 11,\n",
      "= chellata Mller, 1764. [3435] (as Lycosa).\n",
      "Pardosa monticola (Clerck, 1757). [34]; [3435] (as\n",
      "Pardosa paludicola (Clerck, 1757). [34, 14]; [3435] (as\n",
      "Pardosa palustris (Linnaeus, 1758). [34; 6, 10, 14, 16].\n",
      "= tarsalis Thorell, 1856. [3435] (as Lycosa).\n",
      "Pardosa prativaga (L. Koch, 1870). [34; 67; 910; 14,\n",
      "= riparia O. P.-Cambridge, 1875. [19, 2324, 26, 3435]\n",
      "(as Lycosa).\n",
      "Pardosa pullata (Clerck, 1757). [34; 67; 911; 1416;\n",
      "20, 31]; [3435] (as Lycosa).\n",
      "Pardosa riparia (C. L.Koch, 1833). [34; 31].\n",
      "= cursoria (C. L.Koch, 1847). [3435] (as Lycosa).\n",
      "Pardosa schenkeli Lessert, 1904. [34].\n",
      "= calida Dahl, 1908. [3435] (as Lycosa).\n",
      "Pardosa sphagnicola (Dahl, 1908). [34; 6, 7, 9, 11, 14,\n",
      "= riparia sphagnicola Dahl, 1908. [34] (as Lycosa).\n",
      "Pirata hygrophilus Thorell, 1872. [34; 7, 9, 1011; 14,\n",
      "Pirata insularis Emerton, 1885. [5, 7, 9, 11, 1415; 31].\n",
      "Pirata latitans (Blackwall, 1841). [34; 14, 34].\n",
      "Pirata piraticus (Clerck, 1757). [3, 910; 19, 24, 31, 34\n",
      "Pirata piscatorius (Clerck, 1757). [34; 14, 1920; 31,\n",
      "Pirata tenuitarsis Simon, 1876. [56; 14].\n",
      "Pirata uliginosus (Thorell, 1856). [34; 7, 9, 11, 1415;\n",
      "Trochosa robusta (Simon, 1876). [34].\n",
      "Trochosa ruricola (De Geer, 1778). [34; 6, 911; 14, 16,\n",
      "Trochosa spinipalpis (F. O. P.-Cambridge, 1895). [34;\n",
      "Trochosa terricola Thorell, 1856. [34; 67; 11, 14, 16,\n",
      "Xerolycosa miniata (C. L.Koch, 1834). [6, 10, 27].\n",
      "Xerolycosa nemoralis (Westring, 1861). [34; 67; 14,\n",
      "Dolomedes fimbriatus (Clerck, 1757). [4, 7, 910; 14,\n",
      "16, 1921; 2324; 31, 3435]; (Family Dolomedidae.\n",
      "Dolomedes plantarius (Clerck, 1757). [4, 12, 19, 2324];\n",
      "(Family Dolomedidae. [3]).\n",
      "Pisaura mirabilis (Clerck, 1757). [34; 16, 20, 29, 34].\n",
      "Oxyopes ramosus (Martini & Goeze, 1778). [34; 16,\n",
      "Zora nemoralis (Blackwall, 1861). [34; 14, 3435].\n",
      "Zora silvestris Kulczynski, 1897. [7, 9, 31].\n",
      "Zora spinimana (Sundevall, 1833). [34; 67; 9, 11, 15,\n",
      "Agelena labyrinthica (Clerck, 1757). [34; 67; 23].\n",
      "Tegenaria agrestis (Walckenaer, 1802). [10].\n",
      "Tegenaria atrica C. L.Koch, 1843. [5, 10].\n",
      "Tegenaria domestica (Clerck, 1757). [34].\n",
      "= derhami Scopoli 1763. [24].\n",
      "Argyroneta aquatica (Clerck, 1757). [34; 19, 24], (Family Argyronetidae. [16]).\n",
      "Antistea elegans (Blackwall, 1841). [57; 9, 14].\n",
      "Cryphoeca silvicola (C. L.Koch, 1834). [5, 14, 20].\n",
      "Hahnia nava (Blackwall, 1841). [5, 10, 14].\n",
      "Hahnia pusilla C. L.Koch, 1841. [9, 11, 27].\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Hahnia ononidum Simon, 1875. [20].\n",
      "Archaeodictyna consecuta (O. P.-Cambridge, 1872). [4].\n",
      "= sedilloti Simon, 1875. [3, 26] (as Dictyna).\n",
      "Argenna patula (Simon, 1874). [4].\n",
      "Argenna subnigra (O. P.-Cambridge, 1861). [5].\n",
      "Cicurina cicur (Fabricius, 1793). [57; 910; 14].\n",
      "Dictyna arundinacea (Linnaeus, 1758). [34; 7, 20,\n",
      "Dictyna pusilla Thorell, 1856. [34; 23].\n",
      "Dictyna uncinata Thorell, 1856. [4, 16].\n",
      "Amaurobius fenestralis (Strm, 1768). [12].\n",
      "Cheiracanthium erraticum (Walckenaer, 1802). (Family\n",
      "Cheiracanthium montanum L.Koch, 1877. [18]. (Family\n",
      "Cheiracanthium punctorium (Villers, 1789). (Family\n",
      "Cheiracanthium virescens (Sundevall, 1833). (Family\n",
      "= lapidicolens Simon, 1878. [17].\n",
      "Anyphaena accentuata (Walckenaer, 1802). [34; 14, 17].\n",
      "Agraecina striata (Kulczynski, 1882). [8].\n",
      "Agroeca brunnea (Blackwall, 1833). [34; 69; 14, 17,\n",
      "Agroeca cuprea Menge, 1873. [3, 8, 17].\n",
      "= pullata Thorell, 1875 [4].\n",
      "Agroeca dentigera Kulczyski, 1913. [89; 15, 31].\n",
      "Agroeca lusatica (L.Koch, 1875). [8].\n",
      "Agroeca proxima (O. P.-Cambridge, 1871). [59; 11, 14,\n",
      "Scotina palliardi (L.Koch, 1881). [69, 11, 15, 31].\n",
      "Clubiona caerulescens L.Koch, 1867. [34; 1617]\n",
      "Clubiona comta C. L.Koch, 1839. [34; 1617].\n",
      "Clubiona diversa O. P.-Cambridge, 1862. [34; 6, 9,\n",
      "Clubiona frutetorum L.Koch, 1866. [34; 17].\n",
      "Clubiona germanica Thorell, 1871. [34; 17, 22].\n",
      "Clubiona juvenis Simon, 1878. [27].\n",
      "Clubiona lutescens Westring, 1851. [34; 1617; 20].\n",
      "Clubiona marmorata L.Koch, 1866. [34; 1718].\n",
      "Clubiona neglecta O. P.-Cambridge, 1862. [ 34; 6, 10,\n",
      "Clubiona norvegica Strand, 1900. [27].\n",
      "Clubiona pallidula (Clerck, 1757). [34; 6, 16, 22, 26].\n",
      "= C. holosericea Hahn, 1829 [17].\n",
      "Clubiona phragmitis C. L. Koch, 1843. [34; 6, 1617;\n",
      "Clubiona reclusa O. P.-Cambridge, 1863. [34; 17].\n",
      "Clubiona similis L.Koch, 1867. [34; 10, 17].\n",
      "Clubiona stagnatilis Kulczynski, 1897. [34; 7, 16, 19].\n",
      "Clubiona subsultans Thorell, 1875. [4, 7, 14, 16].\n",
      "= erratica C. L.Koch, 1843 [17, 19].\n",
      "Clubiona trivialis C. L. Koch, 1843. [4, 1617; 2325;\n",
      "Phrurolithus festivus (C. L.Koch, 1835). [4, 910; 1415;\n",
      "17, 31]; (Family Micariidae. [3]); (Family Liocranidae. [3, 6, 8].)\n",
      "Phrurolithus minimus C. L. Koch, 1839. [7, 9; 15, 31];\n",
      "(Family Liocranidae. [8].)\n",
      "Berlandina cinerea (Menge, 1868). [34; 8]; [17] (as\n",
      "Callilepis nocturna (Linnaeus, 1758). [12].\n",
      "Drassodes hypocrita (Simon, 1878). [8].\n",
      "Drassodes pubescens (Thorell, 1856). [34; 69; 15, 17,\n",
      "Drassodes villosus (Thorell, 1856). [8].\n",
      "Drassyllus lutetianus (L.Koch, 1866). [59; 14, 31].\n",
      "Gnaphosa bicolor (Hahn, 1831). [34; 8, 17\n",
      "Gnaphosa lugubris (C. L.Koch, 1839). [12].\n",
      "Gnaphosa microps Holm, 1939. [79; 31].\n",
      "Gnaphosa montana (L.Koch, 1866). [34; 8, 17].\n",
      "Gnaphosa muscorum (L.Koch, 1866). [5, 78].\n",
      "Gnaphosa nigerrima L.Koch, 1877. [5, 79; 14, 31].\n",
      "Haplodrassus cognatus (Westring, 1861). [34; 8, 14];\n",
      "[17] (as Drassodes).\n",
      "Haplodrassus dalmatensis (L.Koch, 1866). [5, 8].\n",
      "Haplodrassus moderatus (Kulczynski, 1897). [89; 31].\n",
      "Haplodrassus signifer (C. L. Koch, 1839). [34; 69; 11,\n",
      "14, 31]; [17] (as Drassodes).\n",
      "Haplodrassus silvestris (Blackwall, 1833). [5, 89, 31].\n",
      "Haplodrassus soerenseni (Strand, 1900). [5, 79; 14, 31].\n",
      "Haplodrassus umbratilis (L. Koch, 1866). [4, 14, 8]; [17]\n",
      "(as Drassodes).\n",
      "Micaria fulgens (Walckenaer, 1802). [4, 8]; [17]; (Family\n",
      "Micaria lenzi Bsenberg, 1899. [5, 8].\n",
      "Micaria pulicaria (Sundevall, 1831). [4; 6, 810; 17];\n",
      "(Family Micariidae. [3]).\n",
      "Micaria silesiaca L.Koch, 1875. [56; 8].\n",
      "Micaria subopaca Westring, 1862. [4, 8]; (Family Micariidae. [3]).\n",
      "= albostriata L.Koch, 1877. [17].\n",
      "Scotophaeus blackwalli (Thorell, 1871).\n",
      "= gotlandicus Thorell, 1871 [17].\n",
      "An old record as S. gotlandicus in [17] was wrongly assigned to S. quadripunctatus (Linnaeus, 1758) in [3]\n",
      "and was followed in [4, 8].\n",
      "Scotophaeus scutulatus (L.Koch, 1866). [34; 8, 1718].\n",
      "Sosticus loricatus (L. Koch, 1866). [34; 8]; [17] (as Scotophaeus).\n",
      "Zelotes aeneus (Simon, 1878). [6, 8, 10].\n",
      "Zelotes electus (C. L.Koch, 1839). [34; 6, 8, 10, 14]; (Family Drassidae. [26]).\n",
      "Zelotes exiguus (Mller & Schenkel, 1895). [8, 27].\n",
      "Zelotes clivicola (L.Koch, 1870). [711; 14, 27, 31].\n",
      "Zelotes latreillei (Simon, 1878). [34; 69; 14, 17, 31].\n",
      "Zelotes longipes (L.Koch, 1866). [5, 9, 8].\n",
      "Zelotes petrensis (C. L.Koch, 1839). [34; 78; 14, 17].\n",
      "Drassyllus praeficus (L.Koch, 1866). [56; 8].\n",
      "Drassylus pusillus (C. L. Koch, 1833). [69; 14, 31]; [4]\n",
      "(as Zelotes).\n",
      "Zelotes subterraneus (C. L.Koch, 1833). [34; 79; 17].\n",
      "Micromata virescens (Clerck, 1757). [11].\n",
      "Philodromus aureolus (Clerck, 1757). [34; 19, 2226].\n",
      "Philodromus cespitum (Walckenaer, 1802). [7, 16, 20].\n",
      "Philodromus collinus C. L.Koch, 1835. [5, 14].\n",
      "Philodromus emarginatus (Schrank, 1803). [34; 19,\n",
      "Philodromus fuscomarginatus (De Geer, 1778). [34; 24,\n",
      "Philodromus histrio (Latreille, 1819). [4, 6].\n",
      "Philodromus margaritatus (Clerck, 1757). [12].\n",
      "Philodromus poecilus (Thorell, 1872). [34; 2425].\n",
      "Thanatus arenarius L.Koch, 1872. [10].\n",
      "Thanatus formicinus (Clerck, 1757). [4].\n",
      "Thanatus striatus C. L.Koch, 1845. [6].\n",
      "Tibellus maritimus (Menge, 1875). [34; 19, 24].\n",
      "Tibellus oblongus (Walckenaer, 1802). [34; 1920; 23,\n",
      "Coriarachne depressa (C. L.Koch, 1837). [5].\n",
      "Diaea dorsata (Fabricius, 1777). [4, 14].\n",
      "Misumena vatia (Clerck, 1757). [4, 16, 20].\n",
      "Misumenops tricuspidatus (Fabricius, 1775). [34; 18],\n",
      "[26] (as Misumena).\n",
      "Ozyptila atomaria (Panzer, 1801). [5].\n",
      "Ozyptila brevipes (Hahn, 1826). [34; 14, 1920].\n",
      "Ozyptila praticola (C. L.Koch, 1837). [5, 14].\n",
      "Ozyptila scabricula (Westring, 1851). [27].\n",
      "Ozyptila simplex (O. P.-Cambridge, 1862). [4].\n",
      "Ozyptila trux (Blackwall, 1846). [4, 14].\n",
      "Thomisus onustus Walckenaer, 1805. [34; 23].\n",
      "Xysticus audax (Schrank, 1803). [34].\n",
      "= pini Hahn, 1831. [19, 23].\n",
      "Xysticus bifasciatus C. L.Koch, 1837. [4, 14, 15].\n",
      "Xysticus cristatus (Clerck, 1757). [34; 67; 910; 14,\n",
      "= viaticus Linnaeus, 1758. [2325].\n",
      "Xysticus erraticus (Blackwall, 1834). [34; 7, 23].\n",
      "Xysticus kochi Thorell, 1872. [10, 27].\n",
      "Xysticus lanio C. L.Koch, 1835. [34; 6, 19, 24].\n",
      "Xysticus lineatus (Westring, 1851). [9].\n",
      "Xysticus luctuosus (Blackwall, 1836). [5, 7].\n",
      "Xysticus obscurus Collett, 1877. [7].\n",
      "Xysticus sabulosus (Hahn, 1832). [4, 6].\n",
      "Xysticus ulmi (Hahn, 1831). [34; 67; 9, 14, 1920].\n",
      "FAMILY SALTICIDAE\n",
      "Aellurillus v-insignitus (Clerck, 1757). [6, 27].\n",
      "Asianellus festivus (C. L.Koch, 1834). [12].\n",
      "Ballus chalybeius (Walckenaer, 1802). [12].\n",
      "Dendryphantes hastatus (Clerck, 1757). [34; 2325].\n",
      "Dendryphantes rudis (Sundevall, 1833). [5].\n",
      "Euophrys frontalis (Walckenaer, 1802). [910; 14,\n",
      "Evarcha arcuata (Clerck, 1757). [34; 9, 14, 16, 20, 23\n",
      "= marcgravii Scopoli, 1763. [26].\n",
      "Evarcha falcata (Clerck, 1757). [34; 9, 14, 16, 20, 24\n",
      "Evarcha laetabunda (C. L.Koch, 1846). [4, 9].\n",
      "Heliophanus auratus C. L.Koch, 1835. [5, 16].\n",
      "Heliophanus dampfi Schenkel, 1923. [27].\n",
      "Heliophanus dubius C. L. Koch, 1835. [34; 9, 24\n",
      "Heliophanus flavipes (Hahn, 1832). [5, 10].\n",
      "Marpissa muscosa (Clerck, 1757). [34; 19, 24].\n",
      "Marpissa radiata (Grube, 1859). [34; 16, 19, 24].\n",
      "Neon reticulatus (Blackwall, 1853). [5, 7, 9, 11, 15].\n",
      "Neon robustus Lohmander, 1945. [14].\n",
      "Neon valentulus Falconer, 1912. [5, 7, 9, 14].\n",
      "Pelenes tripunctatus (Walckenaer, 1802). [27].\n",
      "Philaeus chrysops (Poda, 1761). [5].\n",
      "Phlegra fasciata (Hahn, 1826). [5, 10].\n",
      "Pseudeuophrys erratica (Walckenaer, 1826). [5] (as Euophrys).\n",
      "Salticus cingulatus (Panzer, 1797). [4, 16].\n",
      "Salticus scenicus (Clerck, 1757). [4, 16].\n",
      "Sibianor aurocinctus (Ohlert, 1865). [5] (as Bianor).\n",
      "Sitticus caricis (Westring, 1861). [27].\n",
      "Sitticus distinguendus (Simon, 1868). [27].\n",
      "Sitticus dzieduszyckii (L.Koch, 1870). [5].\n",
      "Sitticus floricola (C. L.Koch, 1837). [34].\n",
      "= littoralis Hahn, 1836. [19, 24].\n",
      "Sitticus rupicola (C. L.Koch, 1837). [6].\n",
      "Sitticus saltator (O. P.-Cambridge, 1868). [5] (as Attulus).\n",
      "Sitticus terebratus (Clerck, 1757). [34; 23].\n",
      "Synageles hilarulus (C. L.Koch, 1846). [5].\n",
      "Synageles venator (Lucas, 1836). [34; 2425; 30].\n",
      "Talavera aequipes (O.P.-Cambridge, 1871). [10]; [27] (as\n",
      "Talavera petrensis (C. L. Koch, 1837). [9, 10]; [67] (as\n",
      "Talavera parvistyla Logunov, Kronestedt, 2003.\n",
      "= westringi Miller, 1971. [7, 9] (as Euophrys).\n",
      "Yllenus arenarius Menge, 1868. [27].\n",
      "THE CHECKLIST\n",
      "OF SPECIES\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pholcus phalangioides\n",
      "Segestria senoculata\n",
      "Harpactea rubicunda\n",
      "as Harpactocrates\n",
      "Ero cambridgei\n",
      "Ero furcata\n",
      "Eresus cinnaberinus\n",
      "Hyptiotes paradoxus\n",
      "Crustulina guttata\n",
      "Cryptachaea riparia\n",
      "Enoplognatha ovata\n",
      "as Theridium\n",
      "redimitum Clerck\n",
      "as Theridium\n",
      "Enoplognatha thoracica\n",
      "Episinus angulatus\n",
      "Episinus truncatus\n",
      "Euryopis flavomaculata\n",
      "Keijia tincta\n",
      "as Theridion\n",
      "Lasaeola prona\n",
      "as Dipoena\n",
      "Lasaeola tristis\n",
      "Lessertia dentichelis\n",
      "Neottiura bimaculata\n",
      "as Theridium\n",
      "Parasteatoda lunata\n",
      "as Achaearanea\n",
      "Parasteatoda simulans\n",
      "as Achaearanea\n",
      "Pholcomma gibbum\n",
      "Robertus arundineti\n",
      "Robertus lividus\n",
      "Robertus lyrifer\n",
      "Robertus neglectus\n",
      "Robertus scoticus\n",
      "Robertus ungulatus\n",
      "Simitidion simile\n",
      "Steatoda albomaculata\n",
      "De Geer\n",
      "Steatoda bipunctata\n",
      "Steatoda castanea\n",
      "as Asagena\n",
      "as Teutana\n",
      "Steatoda grossa\n",
      "Steatoda phalerata\n",
      "as Asagena\n",
      "Theonoe minutissima\n",
      "Phyloneta impressa\n",
      "as Theridion\n",
      "Phyloneta sisyphia\n",
      "as Theridion\n",
      "notatum Linnaeus\n",
      "as Theridion\n",
      "Theridion mystaceum\n",
      "Theridion pictum\n",
      "Theridion pinastri\n",
      "Theridion varians\n",
      "Abacoproeces saltum\n",
      "Agyneta cauta\n",
      "Agyneta conigera\n",
      "Agyneta decora\n",
      "Agyneta ramosa\n",
      "Agyneta subtilis\n",
      "Anguliphantes angulipalpis\n",
      "as Lepthyphanthes\n",
      "Aphileta misera\n",
      "Araeoncus humilis\n",
      "Bathyphantes approximatus\n",
      "Bathyphantes gracilis\n",
      "Bathyphantes nigrinus\n",
      "Bathyphantes parvulus\n",
      "Bathyphantes similis\n",
      "Bolyphantes alticeps\n",
      "Bolyphantes luteolus\n",
      "Centromerita bicolor\n",
      "Centromerus arcanus\n",
      "Centromerus brevivulvatus\n",
      "aequalis C\n",
      "Centromerus incilium\n",
      "Centromerus levitarsis\n",
      "Centromerus sylvaticus\n",
      "Centromerus unidentatus\n",
      "Ceratinella brevipes\n",
      "Ceratinella brevis\n",
      "Ceratinella scabrosa\n",
      "Cnephalocotes obscurus\n",
      "Dicymbium nigrum\n",
      "Dicymbium nigrum\n",
      "brevisetosum Locket\n",
      "Dicymbium tibiale\n",
      "Diplocentria bidentata\n",
      "Diplocephalus connatus\n",
      "Diplocephalus cristatus\n",
      "Diplocephalus dentatus\n",
      "Diplocephalus latifrons\n",
      "Diplocephalus permixtus\n",
      "Diplocephalus picinus\n",
      "Diplostyla concolor\n",
      "Dismodicus bifrons\n",
      "Dismodicus elevatus\n",
      "Drapetisca socialis\n",
      "Drepanotylus uncatus\n",
      "Erigone atra\n",
      "Erigone dentipalpis\n",
      "Erigonella hiemalis\n",
      "Erigonella ignobilis\n",
      "Erigone longipalpis\n",
      "Floronia bucculenta\n",
      "Formiphantes lepthyphantiformis\n",
      "as Lepthyphanthes\n",
      "Glyphesis cottonae\n",
      "La Touche\n",
      "Gnathonarium dentatum\n",
      "Gonatium rubellum\n",
      "Gonatium rubens\n",
      "Gongylidiellum latebricola\n",
      "Gongylidiellum murcidum\n",
      "Gongylidium rufipes\n",
      "Helophora insignis\n",
      "Hylyphantes graminicola\n",
      "as Erigonidium\n",
      "as Erigone\n",
      "Hypomma bituberculatum\n",
      "Hypomma cornutum\n",
      "Hypselistes jacksoni\n",
      "Kaestneria dorsalis\n",
      "Kaestneria pullata\n",
      "Lepthyphantes leprosus\n",
      "Linyphia triangularis\n",
      "Lophomma punctatum\n",
      "Macrargus carpenteri\n",
      "Macrargus rufus\n",
      "Mansuphantes mansuetus\n",
      "as Lepthyphanthes\n",
      "Maro minutus\n",
      "Maso sundevalli\n",
      "Megalepthyphantes nebulosus\n",
      "as Lepthyphanthes\n",
      "Mecynargus foveatus\n",
      "Meioneta affinis\n",
      "beata O\n",
      "Meioneta fuscipalpa\n",
      "Meioneta innotabilis\n",
      "as Micryphantes\n",
      "Meioneta mossica\n",
      "Meioneta rurestris\n",
      "Metopobactrus prominulus\n",
      "Micrargus apertus\n",
      "Micrargus herbigradus\n",
      "Micrargus subaequalis\n",
      "Minicia marginella\n",
      "Minyriolus pusillus\n",
      "Microneta viaria\n",
      "Microlinyphia pusilla\n",
      "as Linyphia\n",
      "Mioxena blanda\n",
      "Neriene clathrata\n",
      "Neriene emphana\n",
      "Neriene marginata\n",
      "Neriene montana\n",
      "Neriene peltata\n",
      "Neriene radiata\n",
      "Notioscopus sarcinatus\n",
      "Oedothorax apicatus\n",
      "Oedothorax gibbosus\n",
      "Oedothorax retusus\n",
      "Oryphantes angulatus\n",
      "as Lepthyphantes\n",
      "Palliduphantes insignis\n",
      "Palliduphantes pallidus\n",
      "as Lepthyphanthes\n",
      "Pelecopsis elongata\n",
      "Pelecopsis parallela\n",
      "Pityohyphantes phrygianus\n",
      "as Linyphia\n",
      "Pocadicnemis pumila\n",
      "Porrhomma microphthalmum\n",
      "Porrhomma oblitum\n",
      "Porrhomma pallidum\n",
      "Porrhomma pygmaeum\n",
      "Saaristoa abnormis\n",
      "Savignia frontata\n",
      "Silometopus reussi\n",
      "Sintula cornigera\n",
      "Stemonyphantes lineatus\n",
      "Tallusia experta\n",
      "Tapinocyba biscissa\n",
      "Tapinocyba insecta\n",
      "Tapinocyba pallens\n",
      "Tapinocyba praecox\n",
      "Tapinopa longidens\n",
      "Taranucnus setosus\n",
      "Thyreosthenius parasiticus\n",
      "Tiso vagans\n",
      "Tmeticus affinis\n",
      "Tenuiphantes alacris\n",
      "Tenuiphantes cristatus\n",
      "as Lepthyphanthes\n",
      "Tenuiphantes flavipes\n",
      "Tenuiphantes mengei\n",
      "as Lepthyphanthes\n",
      "Tenuiphantes tenebricola\n",
      "as Lepthyphanthes\n",
      "Troxochrus scabriculus\n",
      "Walckenaeria acuminata\n",
      "Walckenaeria alticeps\n",
      "Walckenaeria antica\n",
      "Walckenaeria atrotibialis\n",
      "Walckenaeria capito\n",
      "Walckenaeria cucullata\n",
      "as Wideria\n",
      "Walckenaeria cuspidata\n",
      "Walckenaeria dysderoides\n",
      "Walckenaeria incisa\n",
      "Walckenaeria karpinskii\n",
      "Walckenaeria kochi\n",
      "Walckenaeria mitrata\n",
      "Walckeaneria nodosa\n",
      "Walckenaeria nudipalpis\n",
      "as Trachynella\n",
      "Walckenaeria obtusa\n",
      "Walckenaeria unicornis\n",
      "Walckenaeria vigilax\n",
      "Zornella cultrigera\n",
      "Metellina mengei\n",
      "as Meta\n",
      "Family Metidae\n",
      "reticulata mengei\n",
      "as Meta\n",
      "reticulata Wiehle\n",
      "as Meta\n",
      "Metellina merianae\n",
      "as Meta\n",
      "Family Metidae\n",
      "Metellina segmentata\n",
      "as Meta\n",
      "as Meta\n",
      "Family Metidae\n",
      "Pachygnatha clercki\n",
      "Pachygnatha degeeri\n",
      "Pachygnatha listeri\n",
      "Tetragnatha extensa\n",
      "solandrii Scopoli\n",
      "Tetragnatha montana\n",
      "The status\n",
      "of this\n",
      "species in\n",
      "Lithuania is\n",
      "not clear\n",
      "all records\n",
      "of T\n",
      "solandrii were\n",
      "wrongly assigned\n",
      "to T\n",
      "montana in\n",
      "and continued\n",
      "The recent\n",
      "not be\n",
      "Tetragnatha nigrita\n",
      "Tetragnatha obtusa\n",
      "Tetragnatha pinicola\n",
      "Tetragnatha striata\n",
      "Aculepeira ceropegia\n",
      "Agalenatea redii\n",
      "as Aranea\n",
      "Araneus alsine\n",
      "Araneus angulatus\n",
      "Araneus diadematus\n",
      "diadema Linnaeus\n",
      "Araneus marmoreus\n",
      "raji Scopoli\n",
      "raji betulae\n",
      "marmoreus pyramidatus\n",
      "Araneus quadratus\n",
      "Araneus sturmi\n",
      "Araneus triguttatus\n",
      "as Atea\n",
      "Araniella cucurbitina\n",
      "as Araneus\n",
      "as Aranea\n",
      "Family Argiopidae\n",
      "Araniella displicata\n",
      "as Aranea\n",
      "Argiope bruennichi\n",
      "Cercidia prominens\n",
      "Cyclosa conica\n",
      "Cyclosa oculata\n",
      "Gibbaranea gibbosa\n",
      "Hypsosinga pygmaea\n",
      "as Singa\n",
      "Hypsosinga sanquinea\n",
      "Larinioides cornutus\n",
      "folium Fourcroy\n",
      "Larinioides ixobolus\n",
      "as Araneus\n",
      "Larinioides patagiatus\n",
      "ocellatus Clerck\n",
      "as Araneus\n",
      "duometorum Fourcroy\n",
      "as Araneus\n",
      "Larinioides sclopetarius\n",
      "sericatus Clerck\n",
      "as Araneus\n",
      "undatus Olivier\n",
      "as Araneus\n",
      "Mangora acalypha\n",
      "Neoscona adianta\n",
      "as Araneus\n",
      "Nuctenea silvicultrix\n",
      "as Aranea\n",
      "Nuctenea umbratica\n",
      "sexpunctatus Linnaeus\n",
      "as Araneus\n",
      "Singa albovittata\n",
      "as Hypsosinga\n",
      "Singa hamata\n",
      "Singa nitidula\n",
      "Stroemiellus stroemi\n",
      "as Zilla\n",
      "Family Metidae\n",
      "Zygiella atrica\n",
      "as Zilla\n",
      "Family Metidae\n",
      "Zygiella x\n",
      "as Zilla\n",
      "litterata Olivier\n",
      "as Zilla\n",
      "Acantholycoa lignaria\n",
      "Alopecosa aculeata\n",
      "as Tarentula\n",
      "Alopecosa barbipes\n",
      "Old records\n",
      "of T\n",
      "were wrongly\n",
      "assigned to\n",
      "Alopecosa cuneata\n",
      "as Tarentula\n",
      "Alopecosa cursor\n",
      "Alopecosa fabrilis\n",
      "as Tarentula\n",
      "Alopecosa inquilina\n",
      "Alopecosa mariae\n",
      "Alopecosa pinetorum\n",
      "fumigata Dahl\n",
      "as Tarentula\n",
      "Alopecosa pulverulenta\n",
      "as Tarentula\n",
      "Alopecosa trabalis\n",
      "Arctosa alpigena\n",
      "Arctosa alpigena\n",
      "Arctosa cinerea\n",
      "Arctosa leopardus\n",
      "Arctosa perita\n",
      "Arctosa stigmosa\n",
      "Aulonia albimana\n",
      "Hygrolycosa rubrofasciata\n",
      "Pardosa agrestis\n",
      "as Lycosa\n",
      "Pardosa agricola\n",
      "fluviatilis Blackwall\n",
      "as Lycosa\n",
      "Pardosa agricola\n",
      "arenicola fucicola\n",
      "as Lycosa\n",
      "agricola arenicola\n",
      "Pardosa amentata\n",
      "saccata Linnaeus\n",
      "as Lycosa\n",
      "Pardosa bifasciata\n",
      "Pardosa hyperborea\n",
      "hyperborea pusilla\n",
      "as Lycosa\n",
      "Pardosa lugubris\n",
      "chellata M\n",
      "as Lycosa\n",
      "Pardosa monticola\n",
      "Pardosa paludicola\n",
      "Pardosa palustris\n",
      "tarsalis Thorell\n",
      "as Lycosa\n",
      "Pardosa prativaga\n",
      "riparia O\n",
      "as Lycosa\n",
      "Pardosa pullata\n",
      "as Lycosa\n",
      "Pardosa riparia\n",
      "as Lycosa\n",
      "Pardosa schenkeli\n",
      "calida Dahl\n",
      "as Lycosa\n",
      "Pardosa sphagnicola\n",
      "riparia sphagnicola\n",
      "as Lycosa\n",
      "Pirata hygrophilus\n",
      "Pirata insularis\n",
      "Pirata latitans\n",
      "Pirata piraticus\n",
      "Pirata piscatorius\n",
      "Pirata tenuitarsis\n",
      "Pirata uliginosus\n",
      "Trochosa robusta\n",
      "Trochosa ruricola\n",
      "De Geer\n",
      "Trochosa spinipalpis\n",
      "Trochosa terricola\n",
      "Xerolycosa miniata\n",
      "Xerolycosa nemoralis\n",
      "Dolomedes fimbriatus\n",
      "Family Dolomedidae\n",
      "Dolomedes plantarius\n",
      "Family Dolomedidae\n",
      "Pisaura mirabilis\n",
      "Oxyopes ramosus\n",
      "Zora nemoralis\n",
      "Zora silvestris\n",
      "Zora spinimana\n",
      "Agelena labyrinthica\n",
      "Tegenaria agrestis\n",
      "Tegenaria atrica\n",
      "Tegenaria domestica\n",
      "derhami Scopoli\n",
      "Argyroneta aquatica\n",
      "Family Argyronetidae\n",
      "Antistea elegans\n",
      "Cryphoeca silvicola\n",
      "Hahnia nava\n",
      "Hahnia pusilla\n",
      "Hahnia ononidum\n",
      "Archaeodictyna consecuta\n",
      "sedilloti Simon\n",
      "as Dictyna\n",
      "Argenna patula\n",
      "Argenna subnigra\n",
      "Cicurina cicur\n",
      "Dictyna arundinacea\n",
      "Dictyna pusilla\n",
      "Dictyna uncinata\n",
      "Amaurobius fenestralis\n",
      "Cheiracanthium erraticum\n",
      "Cheiracanthium montanum\n",
      "Cheiracanthium punctorium\n",
      "Cheiracanthium virescens\n",
      "lapidicolens Simon\n",
      "Anyphaena accentuata\n",
      "Agraecina striata\n",
      "Agroeca brunnea\n",
      "Agroeca cuprea\n",
      "pullata Thorell\n",
      "Agroeca dentigera\n",
      "Agroeca lusatica\n",
      "Agroeca proxima\n",
      "Scotina palliardi\n",
      "Clubiona caerulescens\n",
      "Clubiona comta\n",
      "Clubiona diversa\n",
      "Clubiona frutetorum\n",
      "Clubiona germanica\n",
      "Clubiona juvenis\n",
      "Clubiona lutescens\n",
      "Clubiona marmorata\n",
      "Clubiona neglecta\n",
      "Clubiona norvegica\n",
      "Clubiona pallidula\n",
      "holosericea Hahn\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Clubiona phragmitis\n",
      "Clubiona reclusa\n",
      "Clubiona similis\n",
      "Clubiona stagnatilis\n",
      "Clubiona subsultans\n",
      "erratica C\n",
      "Clubiona trivialis\n",
      "Phrurolithus festivus\n",
      "Family Micariidae\n",
      "Family Liocranidae\n",
      "Phrurolithus minimus\n",
      "Family Liocranidae\n",
      "Berlandina cinerea\n",
      "Callilepis nocturna\n",
      "Drassodes hypocrita\n",
      "Drassodes pubescens\n",
      "Drassodes villosus\n",
      "Drassyllus lutetianus\n",
      "Gnaphosa bicolor\n",
      "Gnaphosa lugubris\n",
      "Gnaphosa microps\n",
      "Gnaphosa montana\n",
      "Gnaphosa muscorum\n",
      "Gnaphosa nigerrima\n",
      "Haplodrassus cognatus\n",
      "as Drassodes\n",
      "Haplodrassus dalmatensis\n",
      "Haplodrassus moderatus\n",
      "Haplodrassus signifer\n",
      "as Drassodes\n",
      "Haplodrassus silvestris\n",
      "Haplodrassus soerenseni\n",
      "Haplodrassus umbratilis\n",
      "as Drassodes\n",
      "Micaria fulgens\n",
      "Micaria lenzi\n",
      "Micaria pulicaria\n",
      "Family Micariidae\n",
      "Micaria silesiaca\n",
      "Micaria subopaca\n",
      "Family Micariidae\n",
      "albostriata L\n",
      "Scotophaeus blackwalli\n",
      "gotlandicus Thorell\n",
      "An old\n",
      "record as\n",
      "gotlandicus in\n",
      "was wrongly\n",
      "assigned to\n",
      "and was\n",
      "followed in\n",
      "Scotophaeus scutulatus\n",
      "Sosticus loricatus\n",
      "as Scotophaeus\n",
      "Zelotes aeneus\n",
      "Zelotes electus\n",
      "Family Drassidae\n",
      "Zelotes exiguus\n",
      "Zelotes clivicola\n",
      "Zelotes latreillei\n",
      "Zelotes longipes\n",
      "Zelotes petrensis\n",
      "Drassyllus praeficus\n",
      "Drassylus pusillus\n",
      "as Zelotes\n",
      "Zelotes subterraneus\n",
      "Micromata virescens\n",
      "Philodromus aureolus\n",
      "Philodromus cespitum\n",
      "Philodromus collinus\n",
      "Philodromus emarginatus\n",
      "Philodromus fuscomarginatus\n",
      "De Geer\n",
      "Philodromus histrio\n",
      "Philodromus margaritatus\n",
      "Philodromus poecilus\n",
      "Thanatus arenarius\n",
      "Thanatus formicinus\n",
      "Thanatus striatus\n",
      "Tibellus maritimus\n",
      "Tibellus oblongus\n",
      "Coriarachne depressa\n",
      "Diaea dorsata\n",
      "Misumena vatia\n",
      "Misumenops tricuspidatus\n",
      "as Misumena\n",
      "Ozyptila atomaria\n",
      "Ozyptila brevipes\n",
      "Ozyptila praticola\n",
      "Ozyptila scabricula\n",
      "Ozyptila simplex\n",
      "Ozyptila trux\n",
      "Thomisus onustus\n",
      "Xysticus audax\n",
      "pini Hahn\n",
      "Xysticus bifasciatus\n",
      "Xysticus cristatus\n",
      "viaticus Linnaeus\n",
      "Xysticus erraticus\n",
      "Xysticus kochi\n",
      "Xysticus lanio\n",
      "Xysticus lineatus\n",
      "Xysticus luctuosus\n",
      "Xysticus obscurus\n",
      "Xysticus sabulosus\n",
      "Xysticus ulmi\n",
      "FAMILY SALTICIDAE\n",
      "Aellurillus v\n",
      "Asianellus festivus\n",
      "Ballus chalybeius\n",
      "Dendryphantes hastatus\n",
      "Dendryphantes rudis\n",
      "Euophrys frontalis\n",
      "Evarcha arcuata\n",
      "marcgravii Scopoli\n",
      "Evarcha falcata\n",
      "Evarcha laetabunda\n",
      "Heliophanus auratus\n",
      "Heliophanus dampfi\n",
      "Heliophanus dubius\n",
      "Heliophanus flavipes\n",
      "Marpissa muscosa\n",
      "Marpissa radiata\n",
      "Neon reticulatus\n",
      "Neon robustus\n",
      "Neon valentulus\n",
      "Pelenes tripunctatus\n",
      "Philaeus chrysops\n",
      "Phlegra fasciata\n",
      "Pseudeuophrys erratica\n",
      "as Euophrys\n",
      "Salticus cingulatus\n",
      "Salticus scenicus\n",
      "Sibianor aurocinctus\n",
      "as Bianor\n",
      "Sitticus caricis\n",
      "Sitticus distinguendus\n",
      "Sitticus dzieduszyckii\n",
      "Sitticus floricola\n",
      "littoralis Hahn\n",
      "Sitticus rupicola\n",
      "Sitticus saltator\n",
      "as Attulus\n",
      "Sitticus terebratus\n",
      "Synageles hilarulus\n",
      "Synageles venator\n",
      "Talavera aequipes\n",
      "Talavera petrensis\n",
      "Talavera parvistyla\n",
      "westringi Miller\n",
      "as Euophrys\n",
      "Yllenus arenarius\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "grep -E '[[:alpha:]]+ [[:alpha:]]+ ' LT_spider_list.txt\n",
    "grep -E '[A-Za-z]+ [A-Za-z]+' LT_spider_list.txt\n",
    "\n",
    "# the following will only show the matches, one match per line (two matches in a single input line are in different output lines)\n",
    "grep -o -E '[A-Za-z]+ [A-Za-z]+' LT_spider_list.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c615efc3",
   "metadata": {},
   "source": [
    "The code above returns what is mostly \"genus species\" matches, but there are lines that correspond to words describing species in the original file. We need to remove those.\n",
    "\n",
    "One obvious pattern that emerges is matches like \"as speciesX\". One option to deal with this is using the **stream editor** `sed`.\n",
    "\n",
    "`sed` is an incredibly powerful command that does match replacement and removes lines by index number or pattern. In the first example, we will use sed to remove lines starting with a pattern `^as` (where `^` indicates that the following characters must be at the begining of the line)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "1509798d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "THE CHECKLIST\n",
      "OF SPECIES\n",
      "Pholcus phalangioides\n",
      "Segestria senoculata\n",
      "Harpactea rubicunda\n",
      "Ero cambridgei\n",
      "Ero furcata\n",
      "Eresus cinnaberinus\n",
      "Hyptiotes paradoxus\n",
      "Crustulina guttata\n",
      "Cryptachaea riparia\n",
      "Enoplognatha ovata\n",
      "redimitum Clerck\n",
      "Enoplognatha thoracica\n",
      "Episinus angulatus\n",
      "Episinus truncatus\n",
      "Euryopis flavomaculata\n",
      "Keijia tincta\n",
      "Lasaeola prona\n",
      "Lasaeola tristis\n",
      "Lessertia dentichelis\n",
      "Neottiura bimaculata\n",
      "Parasteatoda lunata\n",
      "Parasteatoda simulans\n",
      "Pholcomma gibbum\n",
      "Robertus arundineti\n",
      "Robertus lividus\n",
      "Robertus lyrifer\n",
      "Robertus neglectus\n",
      "Robertus scoticus\n",
      "Robertus ungulatus\n",
      "Simitidion simile\n",
      "Steatoda albomaculata\n",
      "De Geer\n",
      "Steatoda bipunctata\n",
      "Steatoda castanea\n",
      "Steatoda grossa\n",
      "Steatoda phalerata\n",
      "Theonoe minutissima\n",
      "Phyloneta impressa\n",
      "Phyloneta sisyphia\n",
      "notatum Linnaeus\n",
      "Theridion mystaceum\n",
      "Theridion pictum\n",
      "Theridion pinastri\n",
      "Theridion varians\n",
      "Abacoproeces saltum\n",
      "Agyneta cauta\n",
      "Agyneta conigera\n",
      "Agyneta decora\n",
      "Agyneta ramosa\n",
      "Agyneta subtilis\n",
      "Anguliphantes angulipalpis\n",
      "Aphileta misera\n",
      "Araeoncus humilis\n",
      "Bathyphantes approximatus\n",
      "Bathyphantes gracilis\n",
      "Bathyphantes nigrinus\n",
      "Bathyphantes parvulus\n",
      "Bathyphantes similis\n",
      "Bolyphantes alticeps\n",
      "Bolyphantes luteolus\n",
      "Centromerita bicolor\n",
      "Centromerus arcanus\n",
      "Centromerus brevivulvatus\n",
      "aequalis C\n",
      "Centromerus incilium\n",
      "Centromerus levitarsis\n",
      "Centromerus sylvaticus\n",
      "Centromerus unidentatus\n",
      "Ceratinella brevipes\n",
      "Ceratinella brevis\n",
      "Ceratinella scabrosa\n",
      "Cnephalocotes obscurus\n",
      "Dicymbium nigrum\n",
      "Dicymbium nigrum\n",
      "brevisetosum Locket\n",
      "Dicymbium tibiale\n",
      "Diplocentria bidentata\n",
      "Diplocephalus connatus\n",
      "Diplocephalus cristatus\n",
      "Diplocephalus dentatus\n",
      "Diplocephalus latifrons\n",
      "Diplocephalus permixtus\n",
      "Diplocephalus picinus\n",
      "Diplostyla concolor\n",
      "Dismodicus bifrons\n",
      "Dismodicus elevatus\n",
      "Drapetisca socialis\n",
      "Drepanotylus uncatus\n",
      "Erigone atra\n",
      "Erigone dentipalpis\n",
      "Erigonella hiemalis\n",
      "Erigonella ignobilis\n",
      "Erigone longipalpis\n",
      "Floronia bucculenta\n",
      "Formiphantes lepthyphantiformis\n",
      "Glyphesis cottonae\n",
      "La Touche\n",
      "Gnathonarium dentatum\n",
      "Gonatium rubellum\n",
      "Gonatium rubens\n",
      "Gongylidiellum latebricola\n",
      "Gongylidiellum murcidum\n",
      "Gongylidium rufipes\n",
      "Helophora insignis\n",
      "Hylyphantes graminicola\n",
      "Hypomma bituberculatum\n",
      "Hypomma cornutum\n",
      "Hypselistes jacksoni\n",
      "Kaestneria dorsalis\n",
      "Kaestneria pullata\n",
      "Lepthyphantes leprosus\n",
      "Linyphia triangularis\n",
      "Lophomma punctatum\n",
      "Macrargus carpenteri\n",
      "Macrargus rufus\n",
      "Mansuphantes mansuetus\n",
      "Maro minutus\n",
      "Maso sundevalli\n",
      "Megalepthyphantes nebulosus\n",
      "Mecynargus foveatus\n",
      "Meioneta affinis\n",
      "beata O\n",
      "Meioneta fuscipalpa\n",
      "Meioneta innotabilis\n",
      "Meioneta mossica\n",
      "Meioneta rurestris\n",
      "Metopobactrus prominulus\n",
      "Micrargus apertus\n",
      "Micrargus herbigradus\n",
      "Micrargus subaequalis\n",
      "Minicia marginella\n",
      "Minyriolus pusillus\n",
      "Microneta viaria\n",
      "Microlinyphia pusilla\n",
      "Mioxena blanda\n",
      "Neriene clathrata\n",
      "Neriene emphana\n",
      "Neriene marginata\n",
      "Neriene montana\n",
      "Neriene peltata\n",
      "Neriene radiata\n",
      "Notioscopus sarcinatus\n",
      "Oedothorax apicatus\n",
      "Oedothorax gibbosus\n",
      "Oedothorax retusus\n",
      "Oryphantes angulatus\n",
      "Palliduphantes insignis\n",
      "Palliduphantes pallidus\n",
      "Pelecopsis elongata\n",
      "Pelecopsis parallela\n",
      "Pityohyphantes phrygianus\n",
      "Pocadicnemis pumila\n",
      "Porrhomma microphthalmum\n",
      "Porrhomma oblitum\n",
      "Porrhomma pallidum\n",
      "Porrhomma pygmaeum\n",
      "Saaristoa abnormis\n",
      "Savignia frontata\n",
      "Silometopus reussi\n",
      "Sintula cornigera\n",
      "Stemonyphantes lineatus\n",
      "Tallusia experta\n",
      "Tapinocyba biscissa\n",
      "Tapinocyba insecta\n",
      "Tapinocyba pallens\n",
      "Tapinocyba praecox\n",
      "Tapinopa longidens\n",
      "Taranucnus setosus\n",
      "Thyreosthenius parasiticus\n",
      "Tiso vagans\n",
      "Tmeticus affinis\n",
      "Tenuiphantes alacris\n",
      "Tenuiphantes cristatus\n",
      "Tenuiphantes flavipes\n",
      "Tenuiphantes mengei\n",
      "Tenuiphantes tenebricola\n",
      "Troxochrus scabriculus\n",
      "Walckenaeria acuminata\n",
      "Walckenaeria alticeps\n",
      "Walckenaeria antica\n",
      "Walckenaeria atrotibialis\n",
      "Walckenaeria capito\n",
      "Walckenaeria cucullata\n",
      "Walckenaeria cuspidata\n",
      "Walckenaeria dysderoides\n",
      "Walckenaeria incisa\n",
      "Walckenaeria karpinskii\n",
      "Walckenaeria kochi\n",
      "Walckenaeria mitrata\n",
      "Walckeaneria nodosa\n",
      "Walckenaeria nudipalpis\n",
      "Walckenaeria obtusa\n",
      "Walckenaeria unicornis\n",
      "Walckenaeria vigilax\n",
      "Zornella cultrigera\n",
      "Metellina mengei\n",
      "Family Metidae\n",
      "reticulata mengei\n",
      "reticulata Wiehle\n",
      "Metellina merianae\n",
      "Family Metidae\n",
      "Metellina segmentata\n",
      "Family Metidae\n",
      "Pachygnatha clercki\n",
      "Pachygnatha degeeri\n",
      "Pachygnatha listeri\n",
      "Tetragnatha extensa\n",
      "solandrii Scopoli\n",
      "Tetragnatha montana\n",
      "The status\n",
      "of this\n",
      "species in\n",
      "Lithuania is\n",
      "not clear\n",
      "all records\n",
      "of T\n",
      "solandrii were\n",
      "wrongly assigned\n",
      "to T\n",
      "montana in\n",
      "and continued\n",
      "The recent\n",
      "not be\n",
      "Tetragnatha nigrita\n",
      "Tetragnatha obtusa\n",
      "Tetragnatha pinicola\n",
      "Tetragnatha striata\n",
      "Aculepeira ceropegia\n",
      "Agalenatea redii\n",
      "Araneus alsine\n",
      "Araneus angulatus\n",
      "Araneus diadematus\n",
      "diadema Linnaeus\n",
      "Araneus marmoreus\n",
      "raji Scopoli\n",
      "raji betulae\n",
      "marmoreus pyramidatus\n",
      "Araneus quadratus\n",
      "Araneus sturmi\n",
      "Araneus triguttatus\n",
      "Araniella cucurbitina\n",
      "Family Argiopidae\n",
      "Araniella displicata\n",
      "Argiope bruennichi\n",
      "Cercidia prominens\n",
      "Cyclosa conica\n",
      "Cyclosa oculata\n",
      "Gibbaranea gibbosa\n",
      "Hypsosinga pygmaea\n",
      "Hypsosinga sanquinea\n",
      "Larinioides cornutus\n",
      "folium Fourcroy\n",
      "Larinioides ixobolus\n",
      "Larinioides patagiatus\n",
      "ocellatus Clerck\n",
      "duometorum Fourcroy\n",
      "Larinioides sclopetarius\n",
      "sericatus Clerck\n",
      "undatus Olivier\n",
      "Mangora acalypha\n",
      "Neoscona adianta\n",
      "Nuctenea silvicultrix\n",
      "Nuctenea umbratica\n",
      "sexpunctatus Linnaeus\n",
      "Singa albovittata\n",
      "Singa hamata\n",
      "Singa nitidula\n",
      "Stroemiellus stroemi\n",
      "Family Metidae\n",
      "Zygiella atrica\n",
      "Family Metidae\n",
      "Zygiella x\n",
      "litterata Olivier\n",
      "Acantholycoa lignaria\n",
      "Alopecosa aculeata\n",
      "Alopecosa barbipes\n",
      "Old records\n",
      "of T\n",
      "were wrongly\n",
      "assigned to\n",
      "Alopecosa cuneata\n",
      "Alopecosa cursor\n",
      "Alopecosa fabrilis\n",
      "Alopecosa inquilina\n",
      "Alopecosa mariae\n",
      "Alopecosa pinetorum\n",
      "fumigata Dahl\n",
      "Alopecosa pulverulenta\n",
      "Alopecosa trabalis\n",
      "Arctosa alpigena\n",
      "Arctosa alpigena\n",
      "Arctosa cinerea\n",
      "Arctosa leopardus\n",
      "Arctosa perita\n",
      "Arctosa stigmosa\n",
      "Aulonia albimana\n",
      "Hygrolycosa rubrofasciata\n",
      "Pardosa agrestis\n",
      "Pardosa agricola\n",
      "fluviatilis Blackwall\n",
      "Pardosa agricola\n",
      "arenicola fucicola\n",
      "agricola arenicola\n",
      "Pardosa amentata\n",
      "saccata Linnaeus\n",
      "Pardosa bifasciata\n",
      "Pardosa hyperborea\n",
      "hyperborea pusilla\n",
      "Pardosa lugubris\n",
      "chellata M\n",
      "Pardosa monticola\n",
      "Pardosa paludicola\n",
      "Pardosa palustris\n",
      "tarsalis Thorell\n",
      "Pardosa prativaga\n",
      "riparia O\n",
      "Pardosa pullata\n",
      "Pardosa riparia\n",
      "Pardosa schenkeli\n",
      "calida Dahl\n",
      "Pardosa sphagnicola\n",
      "riparia sphagnicola\n",
      "Pirata hygrophilus\n",
      "Pirata insularis\n",
      "Pirata latitans\n",
      "Pirata piraticus\n",
      "Pirata piscatorius\n",
      "Pirata tenuitarsis\n",
      "Pirata uliginosus\n",
      "Trochosa robusta\n",
      "Trochosa ruricola\n",
      "De Geer\n",
      "Trochosa spinipalpis\n",
      "Trochosa terricola\n",
      "Xerolycosa miniata\n",
      "Xerolycosa nemoralis\n",
      "Dolomedes fimbriatus\n",
      "Family Dolomedidae\n",
      "Dolomedes plantarius\n",
      "Family Dolomedidae\n",
      "Pisaura mirabilis\n",
      "Oxyopes ramosus\n",
      "Zora nemoralis\n",
      "Zora silvestris\n",
      "Zora spinimana\n",
      "Agelena labyrinthica\n",
      "Tegenaria agrestis\n",
      "Tegenaria atrica\n",
      "Tegenaria domestica\n",
      "derhami Scopoli\n",
      "Argyroneta aquatica\n",
      "Family Argyronetidae\n",
      "Antistea elegans\n",
      "Cryphoeca silvicola\n",
      "Hahnia nava\n",
      "Hahnia pusilla\n",
      "Hahnia ononidum\n",
      "Archaeodictyna consecuta\n",
      "sedilloti Simon\n",
      "Argenna patula\n",
      "Argenna subnigra\n",
      "Cicurina cicur\n",
      "Dictyna arundinacea\n",
      "Dictyna pusilla\n",
      "Dictyna uncinata\n",
      "Amaurobius fenestralis\n",
      "Cheiracanthium erraticum\n",
      "Cheiracanthium montanum\n",
      "Cheiracanthium punctorium\n",
      "Cheiracanthium virescens\n",
      "lapidicolens Simon\n",
      "Anyphaena accentuata\n",
      "Agraecina striata\n",
      "Agroeca brunnea\n",
      "Agroeca cuprea\n",
      "pullata Thorell\n",
      "Agroeca dentigera\n",
      "Agroeca lusatica\n",
      "Agroeca proxima\n",
      "Scotina palliardi\n",
      "Clubiona caerulescens\n",
      "Clubiona comta\n",
      "Clubiona diversa\n",
      "Clubiona frutetorum\n",
      "Clubiona germanica\n",
      "Clubiona juvenis\n",
      "Clubiona lutescens\n",
      "Clubiona marmorata\n",
      "Clubiona neglecta\n",
      "Clubiona norvegica\n",
      "Clubiona pallidula\n",
      "holosericea Hahn\n",
      "Clubiona phragmitis\n",
      "Clubiona reclusa\n",
      "Clubiona similis\n",
      "Clubiona stagnatilis\n",
      "Clubiona subsultans\n",
      "erratica C\n",
      "Clubiona trivialis\n",
      "Phrurolithus festivus\n",
      "Family Micariidae\n",
      "Family Liocranidae\n",
      "Phrurolithus minimus\n",
      "Family Liocranidae\n",
      "Berlandina cinerea\n",
      "Callilepis nocturna\n",
      "Drassodes hypocrita\n",
      "Drassodes pubescens\n",
      "Drassodes villosus\n",
      "Drassyllus lutetianus\n",
      "Gnaphosa bicolor\n",
      "Gnaphosa lugubris\n",
      "Gnaphosa microps\n",
      "Gnaphosa montana\n",
      "Gnaphosa muscorum\n",
      "Gnaphosa nigerrima\n",
      "Haplodrassus cognatus\n",
      "Haplodrassus dalmatensis\n",
      "Haplodrassus moderatus\n",
      "Haplodrassus signifer\n",
      "Haplodrassus silvestris\n",
      "Haplodrassus soerenseni\n",
      "Haplodrassus umbratilis\n",
      "Micaria fulgens\n",
      "Micaria lenzi\n",
      "Micaria pulicaria\n",
      "Family Micariidae\n",
      "Micaria silesiaca\n",
      "Micaria subopaca\n",
      "Family Micariidae\n",
      "albostriata L\n",
      "Scotophaeus blackwalli\n",
      "gotlandicus Thorell\n",
      "An old\n",
      "record as\n",
      "gotlandicus in\n",
      "was wrongly\n",
      "assigned to\n",
      "and was\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "followed in\n",
      "Scotophaeus scutulatus\n",
      "Sosticus loricatus\n",
      "Zelotes aeneus\n",
      "Zelotes electus\n",
      "Family Drassidae\n",
      "Zelotes exiguus\n",
      "Zelotes clivicola\n",
      "Zelotes latreillei\n",
      "Zelotes longipes\n",
      "Zelotes petrensis\n",
      "Drassyllus praeficus\n",
      "Drassylus pusillus\n",
      "Zelotes subterraneus\n",
      "Micromata virescens\n",
      "Philodromus aureolus\n",
      "Philodromus cespitum\n",
      "Philodromus collinus\n",
      "Philodromus emarginatus\n",
      "Philodromus fuscomarginatus\n",
      "De Geer\n",
      "Philodromus histrio\n",
      "Philodromus margaritatus\n",
      "Philodromus poecilus\n",
      "Thanatus arenarius\n",
      "Thanatus formicinus\n",
      "Thanatus striatus\n",
      "Tibellus maritimus\n",
      "Tibellus oblongus\n",
      "Coriarachne depressa\n",
      "Diaea dorsata\n",
      "Misumena vatia\n",
      "Misumenops tricuspidatus\n",
      "Ozyptila atomaria\n",
      "Ozyptila brevipes\n",
      "Ozyptila praticola\n",
      "Ozyptila scabricula\n",
      "Ozyptila simplex\n",
      "Ozyptila trux\n",
      "Thomisus onustus\n",
      "Xysticus audax\n",
      "pini Hahn\n",
      "Xysticus bifasciatus\n",
      "Xysticus cristatus\n",
      "viaticus Linnaeus\n",
      "Xysticus erraticus\n",
      "Xysticus kochi\n",
      "Xysticus lanio\n",
      "Xysticus lineatus\n",
      "Xysticus luctuosus\n",
      "Xysticus obscurus\n",
      "Xysticus sabulosus\n",
      "Xysticus ulmi\n",
      "FAMILY SALTICIDAE\n",
      "Aellurillus v\n",
      "Asianellus festivus\n",
      "Ballus chalybeius\n",
      "Dendryphantes hastatus\n",
      "Dendryphantes rudis\n",
      "Euophrys frontalis\n",
      "Evarcha arcuata\n",
      "marcgravii Scopoli\n",
      "Evarcha falcata\n",
      "Evarcha laetabunda\n",
      "Heliophanus auratus\n",
      "Heliophanus dampfi\n",
      "Heliophanus dubius\n",
      "Heliophanus flavipes\n",
      "Marpissa muscosa\n",
      "Marpissa radiata\n",
      "Neon reticulatus\n",
      "Neon robustus\n",
      "Neon valentulus\n",
      "Pelenes tripunctatus\n",
      "Philaeus chrysops\n",
      "Phlegra fasciata\n",
      "Pseudeuophrys erratica\n",
      "Salticus cingulatus\n",
      "Salticus scenicus\n",
      "Sibianor aurocinctus\n",
      "Sitticus caricis\n",
      "Sitticus distinguendus\n",
      "Sitticus dzieduszyckii\n",
      "Sitticus floricola\n",
      "littoralis Hahn\n",
      "Sitticus rupicola\n",
      "Sitticus saltator\n",
      "Sitticus terebratus\n",
      "Synageles hilarulus\n",
      "Synageles venator\n",
      "Talavera aequipes\n",
      "Talavera petrensis\n",
      "Talavera parvistyla\n",
      "westringi Miller\n",
      "Yllenus arenarius\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "# sed pattern file, only in this case we don't specify the file - input is comming from the pipe\n",
    "grep -o -E '[A-Za-z]+ [A-Za-z]+' LT_spider_list.txt | sed -r '/^as /d'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e14734b6",
   "metadata": {},
   "source": [
    "However, that solution is not the best when lines have more than one matching text, i.e., more than one instance with two strings of alphabetic characters in the same line. Other solution can be instead to use `sed` and regex to match the first instance of the pattern and delete anything else afterwards (line by line)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "350303c2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "THE CHECKLIST\n",
      "Order  ARACHNIDA\n",
      "FAMILY  PHOLCIDAE\n",
      "Pholcus phalangioides\n",
      "FAMILY  SEGESTRIIDAE\n",
      "Segestria senoculata\n",
      "FAMILY  DYSDERIDAE\n",
      "Harpactea rubicunda\n",
      "[17] (as Harpactocrates).\n",
      "FAMILY  MIMETIDAE\n",
      "Ero cambridgei\n",
      "Ero furcata\n",
      "FAMILY  ERESIDAE\n",
      "Eresus cinnaberinus\n",
      "FAMILY  ULOBORIDAE\n",
      "Hyptiotes paradoxus\n",
      "FAMILY  THERIDIIDAE\n",
      "Crustulina guttata\n",
      "Cryptachaea riparia\n",
      "Enoplognatha ovata\n",
      "[2223] (as Theridium);\n",
      "= redimitum Clerck, 1757. [2426] (as Theridium).\n",
      "Enoplognatha thoracica\n",
      "Episinus angulatus\n",
      "Episinus truncatus\n",
      "Euryopis flavomaculata\n",
      "14].\n",
      "Keijia tincta\n",
      "30] (as Theridion).\n",
      "Lasaeola prona\n",
      "Lasaeola tristis\n",
      "Lessertia dentichelis\n",
      "Neottiura bimaculata\n",
      "(as Theridium).\n",
      "Parasteatoda lunata\n",
      "Parasteatoda simulans\n",
      "Pholcomma gibbum\n",
      "Robertus arundineti\n",
      "Robertus lividus\n",
      "1416].\n",
      "Robertus lyrifer\n",
      "Robertus neglectus\n",
      "Robertus scoticus\n",
      "Robertus ungulatus\n",
      "Simitidion simile\n",
      "Theridion).\n",
      "Steatoda albomaculata\n",
      "Steatoda bipunctata\n",
      "Asagena).\n",
      "Steatoda castanea\n",
      "[24] (as Teutana).\n",
      "Steatoda grossa\n",
      "Steatoda phalerata\n",
      "Theonoe minutissima\n",
      "15].\n",
      "Phyloneta impressa\n",
      "(as Theridion).\n",
      "Phyloneta sisyphia\n",
      "= notatum Linnaeus, 1758. [3, 2325] (as Theridion).\n",
      "Theridion mystaceum\n",
      "Theridion pictum\n",
      "2426].\n",
      "Theridion pinastri\n",
      "Theridion varians\n",
      "FAMILY  LINYPHIIDAE\n",
      "Abacoproeces saltum\n",
      "Agyneta cauta\n",
      "15; 31].\n",
      "Agyneta conigera\n",
      "Agyneta decora\n",
      "Agyneta ramosa\n",
      "Agyneta subtilis\n",
      "Anguliphantes angulipalpis\n",
      "(as Lepthyphanthes).\n",
      "Aphileta misera\n",
      "Araeoncus humilis\n",
      "Bathyphantes approximatus\n",
      "[6, 14].\n",
      "Bathyphantes gracilis\n",
      "Bathyphantes nigrinus\n",
      "Bathyphantes parvulus\n",
      "Bathyphantes similis\n",
      "Bolyphantes alticeps\n",
      "Bolyphantes luteolus\n",
      "Centromerita bicolor\n",
      "Centromerus arcanus\n",
      "11, 1415; 31].\n",
      "Centromerus brevivulvatus\n",
      "= aequalis C. L.Koch, 1841 [67; 9, 14].\n",
      "Centromerus incilium\n",
      "Centromerus levitarsis\n",
      "31].\n",
      "Centromerus sylvaticus\n",
      "14].\n",
      "Centromerus unidentatus\n",
      "Ceratinella brevipes\n",
      "Ceratinella brevis\n",
      "Ceratinella scabrosa\n",
      "Cnephalocotes obscurus\n",
      "Dicymbium nigrum\n",
      "Dicymbium nigrum\n",
      "Dicymbium tibiale\n",
      "Diplocentria bidentata\n",
      "Diplocephalus connatus\n",
      "Diplocephalus cristatus\n",
      "Diplocephalus dentatus\n",
      "Diplocephalus latifrons\n",
      "Diplocephalus permixtus\n",
      "Diplocephalus picinus\n",
      "Diplostyla concolor\n",
      "Dismodicus bifrons\n",
      "Dismodicus elevatus\n",
      "Drapetisca socialis\n",
      "Drepanotylus uncatus\n",
      "31].\n",
      "Erigone atra\n",
      "Erigone dentipalpis\n",
      "Erigonella hiemalis\n",
      "Erigonella ignobilis\n",
      "Erigone longipalpis\n",
      "Floronia bucculenta\n",
      "Formiphantes lepthyphantiformis\n",
      "(as Lepthyphanthes).\n",
      "Glyphesis cottonae\n",
      "Gnathonarium dentatum\n",
      "Gonatium rubellum\n",
      "Gonatium rubens\n",
      "15; 31].\n",
      "Gongylidiellum latebricola\n",
      "Gongylidiellum murcidum\n",
      "Gongylidium rufipes\n",
      "Helophora insignis\n",
      "Hylyphantes graminicola\n",
      "2223] (as Erigonidium); [2425] (as Erigone).\n",
      "Hypomma bituberculatum\n",
      "Hypomma cornutum\n",
      "Hypselistes jacksoni\n",
      "Kaestneria dorsalis\n",
      "Kaestneria pullata\n",
      "Lepthyphantes leprosus\n",
      "Linyphia triangularis\n",
      "26].\n",
      "Lophomma punctatum\n",
      "Macrargus carpenteri\n",
      "31].\n",
      "Macrargus rufus\n",
      "Mansuphantes mansuetus\n",
      "Maro minutus\n",
      "Maso sundevalli\n",
      "Megalepthyphantes nebulosus\n",
      "[34; 14, 22] (as Lepthyphanthes).\n",
      "Mecynargus foveatus\n",
      "Meioneta affinis\n",
      "= beata O. P.-Cambridge, 1906. [6, 14].\n",
      "Meioneta fuscipalpa\n",
      "Meioneta innotabilis\n",
      "Agyneta), [4, 26] (as Micryphantes).\n",
      "Meioneta mossica\n",
      "Meioneta rurestris\n",
      "Metopobactrus prominulus\n",
      "[67; 27].\n",
      "Micrargus apertus\n",
      "15].\n",
      "Micrargus herbigradus\n",
      "31].\n",
      "Micrargus subaequalis\n",
      "Minicia marginella\n",
      "Minyriolus pusillus\n",
      "Microneta viaria\n",
      "Microlinyphia pusilla\n",
      "24; 26] (as Linyphia).\n",
      "Mioxena blanda\n",
      "Neriene clathrata\n",
      "Linyphia).\n",
      "Neriene emphana\n",
      "Linyphia).\n",
      "Neriene marginata\n",
      "Linyphia).\n",
      "Neriene montana\n",
      "Neriene peltata\n",
      "Linyphia).\n",
      "Neriene radiata\n",
      "Notioscopus sarcinatus\n",
      "15; 31].\n",
      "Oedothorax apicatus\n",
      "Oedothorax gibbosus\n",
      "Oedothorax retusus\n",
      "Oryphantes angulatus\n",
      "1415; 31] (as Lepthyphantes).\n",
      "Palliduphantes insignis\n",
      "Lepthyphanthes).\n",
      "Palliduphantes pallidus\n",
      "(as Lepthyphanthes).\n",
      "Pelecopsis elongata\n",
      "Pelecopsis parallela\n",
      "Pityohyphantes phrygianus\n",
      "[3] (as Linyphia).\n",
      "Pocadicnemis pumila\n",
      "1415; 31].\n",
      "Porrhomma microphthalmum\n",
      "[6].\n",
      "Porrhomma oblitum\n",
      "Porrhomma pallidum\n",
      "Porrhomma pygmaeum\n",
      "Saaristoa abnormis\n",
      "Savignia frontata\n",
      "Silometopus reussi\n",
      "Sintula cornigera\n",
      "Stemonyphantes lineatus\n",
      "14, 27].\n",
      "Tallusia experta\n",
      "31].\n",
      "Tapinocyba biscissa\n",
      "Tapinocyba insecta\n",
      "Tapinocyba pallens\n",
      "14, 16, 31].\n",
      "Tapinocyba praecox\n",
      "23].\n",
      "Tapinopa longidens\n",
      "Taranucnus setosus\n",
      "31].\n",
      "Thyreosthenius parasiticus\n",
      "Tiso vagans\n",
      "Tmeticus affinis\n",
      "Tenuiphantes alacris\n",
      "Lepthyphantes).\n",
      "Tenuiphantes cristatus\n",
      "(as Lepthyphanthes).\n",
      "Tenuiphantes flavipes\n",
      "Lepthyphanthes).\n",
      "Tenuiphantes mengei\n",
      "(as Lepthyphanthes).\n",
      "Tenuiphantes tenebricola\n",
      "Troxochrus scabriculus\n",
      "Walckenaeria acuminata\n",
      "Walckenaeria alticeps\n",
      "27].\n",
      "Walckenaeria antica\n",
      "31].\n",
      "Walckenaeria atrotibialis\n",
      "15, 31].\n",
      "Walckenaeria capito\n",
      "Walckenaeria cucullata\n",
      "14]; [23] (as Wideria).\n",
      "Walckenaeria cuspidata\n",
      "Walckenaeria dysderoides\n",
      "Walckenaeria incisa\n",
      "Walckenaeria karpinskii\n",
      "Walckenaeria kochi\n",
      "Walckenaeria mitrata\n",
      "Walckeaneria nodosa\n",
      "Walckenaeria nudipalpis\n",
      "26, 31]; [3] (as Trachynella).\n",
      "Walckenaeria obtusa\n",
      "Walckenaeria unicornis\n",
      "Walckenaeria vigilax\n",
      "Zornella cultrigera\n",
      "FAMILY  TETRAGNATHIDAE\n",
      "Metellina mengei\n",
      "(Family Metidae).\n",
      "= reticulata mengei Wiehle, 1931. [32] (as Meta).\n",
      "= reticulata Wiehle, 1931. [26] (as Meta).\n",
      "Metellina merianae\n",
      "Meta); [4] (as Meta) (Family Metidae).\n",
      "Metellina segmentata\n",
      "32] (as Meta); [4] (as Meta) (Family Metidae).\n",
      "Pachygnatha clercki\n",
      "26, 32].\n",
      "Pachygnatha degeeri\n",
      "14, 24, 26, 32].\n",
      "Pachygnatha listeri\n",
      "26, 32].\n",
      "Tetragnatha extensa\n",
      "25; 32].\n",
      "= T. solandrii Scopoli 1763. [19, 2425; 32].\n",
      "Tetragnatha montana\n",
      "The status\n",
      "all records\n",
      "not be\n",
      "Tetragnatha nigrita\n",
      "Tetragnatha obtusa\n",
      "3233].\n",
      "Tetragnatha pinicola\n",
      "30, 32].\n",
      "Tetragnatha striata\n",
      "FAMILY  ARANEIDAE\n",
      "Aculepeira ceropegia\n",
      "Aranea).\n",
      "Agalenatea redii\n",
      "Araneus); [26] (as Aranea).\n",
      "Araneus alsine\n",
      "Araneus angulatus\n",
      "Araneus diadematus\n",
      "32].\n",
      "= diadema Linnaeus, 1758. [26].\n",
      "Araneus marmoreus\n",
      "= raji Scopoli, 1763. [2426; 32];\n",
      "= raji betulae Sulzer, 1776. [32];\n",
      "= marmoreus pyramidatus Clerck, 1757. [3].\n",
      "Araneus quadratus\n",
      "= A.reaumeri. [24, 32].\n",
      "Araneus sturmi\n",
      "Atea).\n",
      "Araneus triguttatus\n",
      "(as Atea).\n",
      "Araniella cucurbitina\n",
      "32] (as Araneus); [23, 26] (as Aranea) (Family Argiopidae).\n",
      "Araniella displicata\n",
      "Argiope bruennichi\n",
      "Cercidia prominens\n",
      "24, 32].\n",
      "Cyclosa conica\n",
      "Cyclosa oculata\n",
      "Gibbaranea gibbosa\n",
      "Hypsosinga pygmaea\n",
      "[19, 24, 32] (as Singa).\n",
      "Hypsosinga sanquinea\n",
      "Larinioides cornutus\n",
      "= folium Fourcroy, 1785. [3]; [19, 2425; 32] (as\n",
      "Araneus).\n",
      "Larinioides ixobolus\n",
      "(as Araneus).\n",
      "Larinioides patagiatus\n",
      "= ocellatus Clerck, 1757. [19, 2326] (as Araneus).\n",
      "= duometorum Fourcroy, 1785. [32] (as Araneus).\n",
      "Larinioides sclopetarius\n",
      "= sericatus Clerck, 1757. [19, 24] (as Araneus).\n",
      "= undatus Olivier, 1789. [32] (as Araneus).\n",
      "Mangora acalypha\n",
      "32].\n",
      "Neoscona adianta\n",
      "25] (as Araneus).\n",
      "Nuctenea silvicultrix\n",
      "Cyphepeira); [19] (as Aranea).\n",
      "Nuctenea umbratica\n",
      "Araneus).\n",
      "= sexpunctatus Linnaeus, 1758 [32] (as Araneus).\n",
      "Singa albovittata\n",
      "Singa hamata\n",
      "Singa nitidula\n",
      "Stroemiellus stroemi\n",
      "Zilla); [4] (as Zilla) (Family Metidae).\n",
      "Zygiella atrica\n",
      "(Family Metidae).\n",
      "Zygiella x\n",
      "Metidae).\n",
      "= litterata Olivier, 1789. [2326] (as Zilla).\n",
      "FAMILY  LYCOSIDAE\n",
      "Acantholycoa lignaria\n",
      "Alopecosa aculeata\n",
      "35] (as Tarentula).\n",
      "Alopecosa barbipes\n",
      "Tarentula).\n",
      "= accentuata (Latreille, 1817). [3].\n",
      "Old records\n",
      "were wrongly\n",
      "1817) in [3].\n",
      "Alopecosa cuneata\n",
      "35] (as Tarentula).\n",
      "Alopecosa cursor\n",
      "Tarentula).\n",
      "Alopecosa fabrilis\n",
      "Alopecosa inquilina\n",
      "Tarentula).\n",
      "Alopecosa mariae\n",
      "Tarentula).\n",
      "Alopecosa pinetorum\n",
      "= fumigata Dahl, 1908. [26] (as Tarentula).\n",
      "Alopecosa pulverulenta\n",
      "1415; 31]; [3435] (as Tarentula).\n",
      "Alopecosa trabalis\n",
      "Tarentula).\n",
      "Arctosa alpigena\n",
      "Arctosa alpigena\n",
      "Arctosa cinerea\n",
      "Arctosa leopardus\n",
      "Arctosa perita\n",
      "Arctosa stigmosa\n",
      "Aulonia albimana\n",
      "31].\n",
      "Hygrolycosa rubrofasciata\n",
      "1415; 20, 31, 3435].\n",
      "Pardosa agrestis\n",
      "3435] (as Lycosa).\n",
      "Pardosa agricola\n",
      "= fluviatilis Blackwall, 1861. [3435] (as Lycosa).\n",
      "Pardosa agricola\n",
      "= arenicola fucicola Dahl, 1908. [34] (as Lycosa).\n",
      "= agricola arenicola var. fucicola (Dahl, 1908). [3].\n",
      "Pardosa amentata\n",
      "= saccata Linnaeus, 1758. [3435] (as Lycosa).\n",
      "Pardosa bifasciata\n",
      "Pardosa hyperborea\n",
      "31].\n",
      "= hyperborea pusilla (Thorell, 1872). [3435] (as Lycosa).\n",
      "Pardosa lugubris\n",
      "14, 16].\n",
      "= chellata Mller, 1764. [3435] (as Lycosa).\n",
      "Pardosa monticola\n",
      "Lycosa).\n",
      "Pardosa paludicola\n",
      "Lycosa).\n",
      "Pardosa palustris\n",
      "= tarsalis Thorell, 1856. [3435] (as Lycosa).\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pardosa prativaga\n",
      "16, 20, 31].\n",
      "= riparia O. P.-Cambridge, 1875. [19, 2324, 26, 3435]\n",
      "(as Lycosa).\n",
      "Pardosa pullata\n",
      "20, 31]; [3435] (as Lycosa).\n",
      "Pardosa riparia\n",
      "= cursoria (C. L.Koch, 1847). [3435] (as Lycosa).\n",
      "Pardosa schenkeli\n",
      "= calida Dahl, 1908. [3435] (as Lycosa).\n",
      "Pardosa sphagnicola\n",
      "15, 20, 31].\n",
      "= riparia sphagnicola Dahl, 1908. [34] (as Lycosa).\n",
      "Pirata hygrophilus\n",
      "20, 26, 31, 3435].\n",
      "Pirata insularis\n",
      "Pirata latitans\n",
      "Pirata piraticus\n",
      "35].\n",
      "Pirata piscatorius\n",
      "3435].\n",
      "Pirata tenuitarsis\n",
      "Pirata uliginosus\n",
      "20, 31, 3435].\n",
      "Trochosa robusta\n",
      "= lapidicola (Dahl, 1927). [26, 3435].\n",
      "Trochosa ruricola\n",
      "19, 24, 26, 3435].\n",
      "Trochosa spinipalpis\n",
      "67; 9, 11, 1415; 26, 31, 3435].\n",
      "Trochosa terricola\n",
      "2324; 26, 31, 3435].\n",
      "Xerolycosa miniata\n",
      "Xerolycosa nemoralis\n",
      "24, 3435].\n",
      "FAMILY  PISAURIDAE\n",
      "Dolomedes fimbriatus\n",
      "16, 1921; 2324; 31, 3435]; (Family Dolomedidae.\n",
      "[3]).\n",
      "Dolomedes plantarius\n",
      "(Family Dolomedidae. [3]).\n",
      "Pisaura mirabilis\n",
      "FAMILY  OXYOPIDAE\n",
      "Oxyopes ramosus\n",
      "1920; 2425].\n",
      "FAMILY  ZORIDAE\n",
      "Zora nemoralis\n",
      "Zora silvestris\n",
      "Zora spinimana\n",
      "19, 31, 3435].\n",
      "FAMILY  AGELENIDAE\n",
      "Agelena labyrinthica\n",
      "Tegenaria agrestis\n",
      "Tegenaria atrica\n",
      "Tegenaria domestica\n",
      "= derhami Scopoli 1763. [24].\n",
      "FAMILY  CYBAEIDAE\n",
      "Argyroneta aquatica\n",
      "FAMILY  HAHNIIDAE\n",
      "Antistea elegans\n",
      "Cryphoeca silvicola\n",
      "Hahnia nava\n",
      "Hahnia pusilla\n",
      "Hahnia ononidum\n",
      "FAMILY  DICTYNIDAE\n",
      "Archaeodictyna consecuta\n",
      "= sedilloti Simon, 1875. [3, 26] (as Dictyna).\n",
      "Argenna patula\n",
      "Argenna subnigra\n",
      "Cicurina cicur\n",
      "Dictyna arundinacea\n",
      "2326].\n",
      "Dictyna pusilla\n",
      "Dictyna uncinata\n",
      "FAMILY  AMAUROBIIDAE\n",
      "Amaurobius fenestralis\n",
      "FAMILY  MITURGIDAE\n",
      "Cheiracanthium erraticum\n",
      "Clubionidae. [6]).\n",
      "Cheiracanthium montanum\n",
      "Clubionidae. [34; 17]).\n",
      "Cheiracanthium punctorium\n",
      "Clubionidae. [34; 17]).\n",
      "Cheiracanthium virescens\n",
      "Clubionidae, [34; 10]).\n",
      "= lapidicolens Simon, 1878. [17].\n",
      "FAMILY  ANYPHAENIDAE\n",
      "Anyphaena accentuata\n",
      "FAMILY  LIOCRANIDAE\n",
      "Agraecina striata\n",
      "Agroeca brunnea\n",
      "31].\n",
      "Agroeca cuprea\n",
      "= pullata Thorell, 1875 [4].\n",
      "Agroeca dentigera\n",
      "Agroeca lusatica\n",
      "Agroeca proxima\n",
      "31].\n",
      "Scotina palliardi\n",
      "FAMILY  CLUBIONIDAE\n",
      "Clubiona caerulescens\n",
      "Clubiona comta\n",
      "Clubiona diversa\n",
      "26].\n",
      "Clubiona frutetorum\n",
      "Clubiona germanica\n",
      "Clubiona juvenis\n",
      "Clubiona lutescens\n",
      "Clubiona marmorata\n",
      "Clubiona neglecta\n",
      "17].\n",
      "Clubiona norvegica\n",
      "Clubiona pallidula\n",
      "= C. holosericea Hahn, 1829 [17].\n",
      "Clubiona phragmitis\n",
      "1920; 22, 2425].\n",
      "Clubiona reclusa\n",
      "Clubiona similis\n",
      "Clubiona stagnatilis\n",
      "Clubiona subsultans\n",
      "= erratica C. L.Koch, 1843 [17, 19].\n",
      "Clubiona trivialis\n",
      "30].\n",
      "FAMILY  CORINNIDAE\n",
      "Phrurolithus festivus\n",
      "17, 31]; (Family Micariidae. [3]); (Family Liocranidae. [3, 6, 8].)\n",
      "Phrurolithus minimus\n",
      "(Family Liocranidae. [8].)\n",
      "FAMILY  GNAPHOSIDAE\n",
      "Berlandina cinerea\n",
      "Pterotricha).\n",
      "Callilepis nocturna\n",
      "Drassodes hypocrita\n",
      "Drassodes pubescens\n",
      "31].\n",
      "Drassodes villosus\n",
      "Drassyllus lutetianus\n",
      "Gnaphosa bicolor\n",
      "Gnaphosa lugubris\n",
      "Gnaphosa microps\n",
      "Gnaphosa montana\n",
      "Gnaphosa muscorum\n",
      "Gnaphosa nigerrima\n",
      "Haplodrassus cognatus\n",
      "[17] (as Drassodes).\n",
      "Haplodrassus dalmatensis\n",
      "Haplodrassus moderatus\n",
      "Haplodrassus signifer\n",
      "14, 31]; [17] (as Drassodes).\n",
      "Haplodrassus silvestris\n",
      "Haplodrassus soerenseni\n",
      "Haplodrassus umbratilis\n",
      "(as Drassodes).\n",
      "Micaria fulgens\n",
      "Micariidae. [3]).\n",
      "Micaria lenzi\n",
      "Micaria pulicaria\n",
      "(Family Micariidae. [3]).\n",
      "Micaria silesiaca\n",
      "Micaria subopaca\n",
      "= albostriata L.Koch, 1877. [17].\n",
      "Scotophaeus blackwalli\n",
      "= gotlandicus Thorell, 1871 [17].\n",
      "An old\n",
      "and was\n",
      "Scotophaeus scutulatus\n",
      "Sosticus loricatus\n",
      "Zelotes aeneus\n",
      "Zelotes electus\n",
      "Zelotes exiguus\n",
      "Zelotes clivicola\n",
      "Zelotes latreillei\n",
      "Zelotes longipes\n",
      "Zelotes petrensis\n",
      "Drassyllus praeficus\n",
      "Drassylus pusillus\n",
      "(as Zelotes).\n",
      "Zelotes subterraneus\n",
      "FAMILY  SPARASSIDAE\n",
      "Micromata virescens\n",
      "FAMILY  PHILODROMIDAE\n",
      "Philodromus aureolus\n",
      "Philodromus cespitum\n",
      "Philodromus collinus\n",
      "Philodromus emarginatus\n",
      "2325].\n",
      "Philodromus fuscomarginatus\n",
      "30].\n",
      "Philodromus histrio\n",
      "Philodromus margaritatus\n",
      "Philodromus poecilus\n",
      "Thanatus arenarius\n",
      "Thanatus formicinus\n",
      "Thanatus striatus\n",
      "Tibellus maritimus\n",
      "Tibellus oblongus\n",
      "24].\n",
      "FAMILY  THOMISIDAE\n",
      "Coriarachne depressa\n",
      "Diaea dorsata\n",
      "Misumena vatia\n",
      "Misumenops tricuspidatus\n",
      "[26] (as Misumena).\n",
      "Ozyptila atomaria\n",
      "Ozyptila brevipes\n",
      "Ozyptila praticola\n",
      "Ozyptila scabricula\n",
      "Ozyptila simplex\n",
      "Ozyptila trux\n",
      "Thomisus onustus\n",
      "Xysticus audax\n",
      "= pini Hahn, 1831. [19, 23].\n",
      "Xysticus bifasciatus\n",
      "Xysticus cristatus\n",
      "16, 20, 26].\n",
      "= viaticus Linnaeus, 1758. [2325].\n",
      "Xysticus erraticus\n",
      "Xysticus kochi\n",
      "Xysticus lanio\n",
      "Xysticus lineatus\n",
      "Xysticus luctuosus\n",
      "Xysticus obscurus\n",
      "Xysticus sabulosus\n",
      "Xysticus ulmi\n",
      "FAMILY SALTICIDAE\n",
      "Aellurillus v\n",
      "Asianellus festivus\n",
      "Ballus chalybeius\n",
      "Dendryphantes hastatus\n",
      "Dendryphantes rudis\n",
      "Euophrys frontalis\n",
      "27].\n",
      "Evarcha arcuata\n",
      "26].\n",
      "= marcgravii Scopoli, 1763. [26].\n",
      "Evarcha falcata\n",
      "25].\n",
      "Evarcha laetabunda\n",
      "Heliophanus auratus\n",
      "Heliophanus dampfi\n",
      "Heliophanus dubius\n",
      "25].\n",
      "Heliophanus flavipes\n",
      "Marpissa muscosa\n",
      "Marpissa radiata\n",
      "Neon reticulatus\n",
      "Neon robustus\n",
      "Neon valentulus\n",
      "Pelenes tripunctatus\n",
      "Philaeus chrysops\n",
      "Phlegra fasciata\n",
      "Pseudeuophrys erratica\n",
      "Salticus cingulatus\n",
      "Salticus scenicus\n",
      "Sibianor aurocinctus\n",
      "Sitticus caricis\n",
      "Sitticus distinguendus\n",
      "Sitticus dzieduszyckii\n",
      "Sitticus floricola\n",
      "= littoralis Hahn, 1836. [19, 24].\n",
      "Sitticus rupicola\n",
      "Sitticus saltator\n",
      "Sitticus terebratus\n",
      "Synageles hilarulus\n",
      "Synageles venator\n",
      "Talavera aequipes\n",
      "Euophrys).\n",
      "Talavera petrensis\n",
      "Euophrys).\n",
      "Talavera parvistyla\n",
      "= westringi Miller, 1971. [7, 9] (as Euophrys).\n",
      "Yllenus arenariusTHE CHECKLIST\n",
      "Pholcus phalangioides\n",
      "Segestria senoculata\n",
      "Harpactea rubicunda\n",
      "Ero cambridgei\n",
      "Ero furcata\n",
      "Eresus cinnaberinus\n",
      "Hyptiotes paradoxus\n",
      "Crustulina guttata\n",
      "Cryptachaea riparia\n",
      "Enoplognatha ovata\n",
      "Enoplognatha thoracica\n",
      "Episinus angulatus\n",
      "Episinus truncatus\n",
      "Euryopis flavomaculata\n",
      "Keijia tincta\n",
      "Lasaeola prona\n",
      "Lasaeola tristis\n",
      "Lessertia dentichelis\n",
      "Neottiura bimaculata\n",
      "Parasteatoda lunata\n",
      "Parasteatoda simulans\n",
      "Pholcomma gibbum\n",
      "Robertus arundineti\n",
      "Robertus lividus\n",
      "Robertus lyrifer\n",
      "Robertus neglectus\n",
      "Robertus scoticus\n",
      "Robertus ungulatus\n",
      "Simitidion simile\n",
      "Steatoda albomaculata\n",
      "Steatoda bipunctata\n",
      "Steatoda castanea\n",
      "Steatoda grossa\n",
      "Steatoda phalerata\n",
      "Theonoe minutissima\n",
      "Phyloneta impressa\n",
      "Phyloneta sisyphia\n",
      "Theridion mystaceum\n",
      "Theridion pictum\n",
      "Theridion pinastri\n",
      "Theridion varians\n",
      "Abacoproeces saltum\n",
      "Agyneta cauta\n",
      "Agyneta conigera\n",
      "Agyneta decora\n",
      "Agyneta ramosa\n",
      "Agyneta subtilis\n",
      "Anguliphantes angulipalpis\n",
      "Aphileta misera\n",
      "Araeoncus humilis\n",
      "Bathyphantes approximatus\n",
      "Bathyphantes gracilis\n",
      "Bathyphantes nigrinus\n",
      "Bathyphantes parvulus\n",
      "Bathyphantes similis\n",
      "Bolyphantes alticeps\n",
      "Bolyphantes luteolus\n",
      "Centromerita bicolor\n",
      "Centromerus arcanus\n",
      "Centromerus brevivulvatus\n",
      "Centromerus incilium\n",
      "Centromerus levitarsis\n",
      "Centromerus sylvaticus\n",
      "Centromerus unidentatus\n",
      "Ceratinella brevipes\n",
      "Ceratinella brevis\n",
      "Ceratinella scabrosa\n",
      "Cnephalocotes obscurus\n",
      "Dicymbium nigrum\n",
      "Dicymbium nigrum\n",
      "Dicymbium tibiale\n",
      "Diplocentria bidentata\n",
      "Diplocephalus connatus\n",
      "Diplocephalus cristatus\n",
      "Diplocephalus dentatus\n",
      "Diplocephalus latifrons\n",
      "Diplocephalus permixtus\n",
      "Diplocephalus picinus\n",
      "Diplostyla concolor\n",
      "Dismodicus bifrons\n",
      "Dismodicus elevatus\n",
      "Drapetisca socialis\n",
      "Drepanotylus uncatus\n",
      "Erigone atra\n",
      "Erigone dentipalpis\n",
      "Erigonella hiemalis\n",
      "Erigonella ignobilis\n",
      "Erigone longipalpis\n",
      "Floronia bucculenta\n",
      "Formiphantes lepthyphantiformis\n",
      "Glyphesis cottonae\n",
      "Gnathonarium dentatum\n",
      "Gonatium rubellum\n",
      "Gonatium rubens\n",
      "Gongylidiellum latebricola\n",
      "Gongylidiellum murcidum\n",
      "Gongylidium rufipes\n",
      "Helophora insignis\n",
      "Hylyphantes graminicola\n",
      "Hypomma bituberculatum\n",
      "Hypomma cornutum\n",
      "Hypselistes jacksoni\n",
      "Kaestneria dorsalis\n",
      "Kaestneria pullata\n",
      "Lepthyphantes leprosus\n",
      "Linyphia triangularis\n",
      "Lophomma punctatum\n",
      "Macrargus carpenteri\n",
      "Macrargus rufus\n",
      "Mansuphantes mansuetus\n",
      "Maro minutus\n",
      "Maso sundevalli\n",
      "Megalepthyphantes nebulosus\n",
      "Mecynargus foveatus\n",
      "Meioneta affinis\n",
      "Meioneta fuscipalpa\n",
      "Meioneta innotabilis\n",
      "Meioneta mossica\n",
      "Meioneta rurestris\n",
      "Metopobactrus prominulus\n",
      "Micrargus apertus\n",
      "Micrargus herbigradus\n",
      "Micrargus subaequalis\n",
      "Minicia marginella\n",
      "Minyriolus pusillus\n",
      "Microneta viaria\n",
      "Microlinyphia pusilla\n",
      "Mioxena blanda\n",
      "Neriene clathrata\n",
      "Neriene emphana\n",
      "Neriene marginata\n",
      "Neriene montana\n",
      "Neriene peltata\n",
      "Neriene radiata\n",
      "Notioscopus sarcinatus\n",
      "Oedothorax apicatus\n",
      "Oedothorax gibbosus\n",
      "Oedothorax retusus\n",
      "Oryphantes angulatus\n",
      "Palliduphantes insignis\n",
      "Palliduphantes pallidus\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pelecopsis elongata\n",
      "Pelecopsis parallela\n",
      "Pityohyphantes phrygianus\n",
      "Pocadicnemis pumila\n",
      "Porrhomma microphthalmum\n",
      "Porrhomma oblitum\n",
      "Porrhomma pallidum\n",
      "Porrhomma pygmaeum\n",
      "Saaristoa abnormis\n",
      "Savignia frontata\n",
      "Silometopus reussi\n",
      "Sintula cornigera\n",
      "Stemonyphantes lineatus\n",
      "Tallusia experta\n",
      "Tapinocyba biscissa\n",
      "Tapinocyba insecta\n",
      "Tapinocyba pallens\n",
      "Tapinocyba praecox\n",
      "Tapinopa longidens\n",
      "Taranucnus setosus\n",
      "Thyreosthenius parasiticus\n",
      "Tiso vagans\n",
      "Tmeticus affinis\n",
      "Tenuiphantes alacris\n",
      "Tenuiphantes cristatus\n",
      "Tenuiphantes flavipes\n",
      "Tenuiphantes mengei\n",
      "Tenuiphantes tenebricola\n",
      "Troxochrus scabriculus\n",
      "Walckenaeria acuminata\n",
      "Walckenaeria alticeps\n",
      "Walckenaeria antica\n",
      "Walckenaeria atrotibialis\n",
      "Walckenaeria capito\n",
      "Walckenaeria cucullata\n",
      "Walckenaeria cuspidata\n",
      "Walckenaeria dysderoides\n",
      "Walckenaeria incisa\n",
      "Walckenaeria karpinskii\n",
      "Walckenaeria kochi\n",
      "Walckenaeria mitrata\n",
      "Walckeaneria nodosa\n",
      "Walckenaeria nudipalpis\n",
      "Walckenaeria obtusa\n",
      "Walckenaeria unicornis\n",
      "Walckenaeria vigilax\n",
      "Zornella cultrigera\n",
      "Metellina mengei\n",
      "Metellina merianae\n",
      "Metellina segmentata\n",
      "Pachygnatha clercki\n",
      "Pachygnatha degeeri\n",
      "Pachygnatha listeri\n",
      "Tetragnatha extensa\n",
      "Tetragnatha montana\n",
      "The status\n",
      "all records\n",
      "not be\n",
      "Tetragnatha nigrita\n",
      "Tetragnatha obtusa\n",
      "Tetragnatha pinicola\n",
      "Tetragnatha striata\n",
      "Aculepeira ceropegia\n",
      "Agalenatea redii\n",
      "Araneus alsine\n",
      "Araneus angulatus\n",
      "Araneus diadematus\n",
      "Araneus marmoreus\n",
      "Araneus quadratus\n",
      "Araneus sturmi\n",
      "Araneus triguttatus\n",
      "Araniella cucurbitina\n",
      "Araniella displicata\n",
      "Argiope bruennichi\n",
      "Cercidia prominens\n",
      "Cyclosa conica\n",
      "Cyclosa oculata\n",
      "Gibbaranea gibbosa\n",
      "Hypsosinga pygmaea\n",
      "Hypsosinga sanquinea\n",
      "Larinioides cornutus\n",
      "Larinioides ixobolus\n",
      "Larinioides patagiatus\n",
      "Larinioides sclopetarius\n",
      "Mangora acalypha\n",
      "Neoscona adianta\n",
      "Nuctenea silvicultrix\n",
      "Nuctenea umbratica\n",
      "Singa albovittata\n",
      "Singa hamata\n",
      "Singa nitidula\n",
      "Stroemiellus stroemi\n",
      "Zygiella atrica\n",
      "Zygiella x\n",
      "Acantholycoa lignaria\n",
      "Alopecosa aculeata\n",
      "Alopecosa barbipes\n",
      "Old records\n",
      "were wrongly\n",
      "Alopecosa cuneata\n",
      "Alopecosa cursor\n",
      "Alopecosa fabrilis\n",
      "Alopecosa inquilina\n",
      "Alopecosa mariae\n",
      "Alopecosa pinetorum\n",
      "Alopecosa pulverulenta\n",
      "Alopecosa trabalis\n",
      "Arctosa alpigena\n",
      "Arctosa alpigena\n",
      "Arctosa cinerea\n",
      "Arctosa leopardus\n",
      "Arctosa perita\n",
      "Arctosa stigmosa\n",
      "Aulonia albimana\n",
      "Hygrolycosa rubrofasciata\n",
      "Pardosa agrestis\n",
      "Pardosa agricola\n",
      "Pardosa agricola\n",
      "Pardosa amentata\n",
      "Pardosa bifasciata\n",
      "Pardosa hyperborea\n",
      "Pardosa lugubris\n",
      "Pardosa monticola\n",
      "Pardosa paludicola\n",
      "Pardosa palustris\n",
      "Pardosa prativaga\n",
      "Pardosa pullata\n",
      "Pardosa riparia\n",
      "Pardosa schenkeli\n",
      "Pardosa sphagnicola\n",
      "Pirata hygrophilus\n",
      "Pirata insularis\n",
      "Pirata latitans\n",
      "Pirata piraticus\n",
      "Pirata piscatorius\n",
      "Pirata tenuitarsis\n",
      "Pirata uliginosus\n",
      "Trochosa robusta\n",
      "Trochosa ruricola\n",
      "Trochosa spinipalpis\n",
      "Trochosa terricola\n",
      "Xerolycosa miniata\n",
      "Xerolycosa nemoralis\n",
      "Dolomedes fimbriatus\n",
      "Dolomedes plantarius\n",
      "Pisaura mirabilis\n",
      "Oxyopes ramosus\n",
      "Zora nemoralis\n",
      "Zora silvestris\n",
      "Zora spinimana\n",
      "Agelena labyrinthica\n",
      "Tegenaria agrestis\n",
      "Tegenaria atrica\n",
      "Tegenaria domestica\n",
      "Argyroneta aquatica\n",
      "Antistea elegans\n",
      "Cryphoeca silvicola\n",
      "Hahnia nava\n",
      "Hahnia pusilla\n",
      "Hahnia ononidum\n",
      "Archaeodictyna consecuta\n",
      "Argenna patula\n",
      "Argenna subnigra\n",
      "Cicurina cicur\n",
      "Dictyna arundinacea\n",
      "Dictyna pusilla\n",
      "Dictyna uncinata\n",
      "Amaurobius fenestralis\n",
      "Cheiracanthium erraticum\n",
      "Cheiracanthium montanum\n",
      "Cheiracanthium punctorium\n",
      "Cheiracanthium virescens\n",
      "Anyphaena accentuata\n",
      "Agraecina striata\n",
      "Agroeca brunnea\n",
      "Agroeca cuprea\n",
      "Agroeca dentigera\n",
      "Agroeca lusatica\n",
      "Agroeca proxima\n",
      "Scotina palliardi\n",
      "Clubiona caerulescens\n",
      "Clubiona comta\n",
      "Clubiona diversa\n",
      "Clubiona frutetorum\n",
      "Clubiona germanica\n",
      "Clubiona juvenis\n",
      "Clubiona lutescens\n",
      "Clubiona marmorata\n",
      "Clubiona neglecta\n",
      "Clubiona norvegica\n",
      "Clubiona pallidula\n",
      "Clubiona phragmitis\n",
      "Clubiona reclusa\n",
      "Clubiona similis\n",
      "Clubiona stagnatilis\n",
      "Clubiona subsultans\n",
      "Clubiona trivialis\n",
      "Phrurolithus festivus\n",
      "Phrurolithus minimus\n",
      "Berlandina cinerea\n",
      "Callilepis nocturna\n",
      "Drassodes hypocrita\n",
      "Drassodes pubescens\n",
      "Drassodes villosus\n",
      "Drassyllus lutetianus\n",
      "Gnaphosa bicolor\n",
      "Gnaphosa lugubris\n",
      "Gnaphosa microps\n",
      "Gnaphosa montana\n",
      "Gnaphosa muscorum\n",
      "Gnaphosa nigerrima\n",
      "Haplodrassus cognatus\n",
      "Haplodrassus dalmatensis\n",
      "Haplodrassus moderatus\n",
      "Haplodrassus signifer\n",
      "Haplodrassus silvestris\n",
      "Haplodrassus soerenseni\n",
      "Haplodrassus umbratilis\n",
      "Micaria fulgens\n",
      "Micaria lenzi\n",
      "Micaria pulicaria\n",
      "Micaria silesiaca\n",
      "Micaria subopaca\n",
      "Scotophaeus blackwalli\n",
      "An old\n",
      "and was\n",
      "Scotophaeus scutulatus\n",
      "Sosticus loricatus\n",
      "Zelotes aeneus\n",
      "Zelotes electus\n",
      "Zelotes exiguus\n",
      "Zelotes clivicola\n",
      "Zelotes latreillei\n",
      "Zelotes longipes\n",
      "Zelotes petrensis\n",
      "Drassyllus praeficus\n",
      "Drassylus pusillus\n",
      "Zelotes subterraneus\n",
      "Micromata virescens\n",
      "Philodromus aureolus\n",
      "Philodromus cespitum\n",
      "Philodromus collinus\n",
      "Philodromus emarginatus\n",
      "Philodromus fuscomarginatus\n",
      "Philodromus histrio\n",
      "Philodromus margaritatus\n",
      "Philodromus poecilus\n",
      "Thanatus arenarius\n",
      "Thanatus formicinus\n",
      "Thanatus striatus\n",
      "Tibellus maritimus\n",
      "Tibellus oblongus\n",
      "Coriarachne depressa\n",
      "Diaea dorsata\n",
      "Misumena vatia\n",
      "Misumenops tricuspidatus\n",
      "Ozyptila atomaria\n",
      "Ozyptila brevipes\n",
      "Ozyptila praticola\n",
      "Ozyptila scabricula\n",
      "Ozyptila simplex\n",
      "Ozyptila trux\n",
      "Thomisus onustus\n",
      "Xysticus audax\n",
      "Xysticus bifasciatus\n",
      "Xysticus cristatus\n",
      "Xysticus erraticus\n",
      "Xysticus kochi\n",
      "Xysticus lanio\n",
      "Xysticus lineatus\n",
      "Xysticus luctuosus\n",
      "Xysticus obscurus\n",
      "Xysticus sabulosus\n",
      "Xysticus ulmi\n",
      "Aellurillus v\n",
      "Asianellus festivus\n",
      "Ballus chalybeius\n",
      "Dendryphantes hastatus\n",
      "Dendryphantes rudis\n",
      "Euophrys frontalis\n",
      "Evarcha arcuata\n",
      "Evarcha falcata\n",
      "Evarcha laetabunda\n",
      "Heliophanus auratus\n",
      "Heliophanus dampfi\n",
      "Heliophanus dubius\n",
      "Heliophanus flavipes\n",
      "Marpissa muscosa\n",
      "Marpissa radiata\n",
      "Neon reticulatus\n",
      "Neon robustus\n",
      "Neon valentulus\n",
      "Pelenes tripunctatus\n",
      "Philaeus chrysops\n",
      "Phlegra fasciata\n",
      "Pseudeuophrys erratica\n",
      "Salticus cingulatus\n",
      "Salticus scenicus\n",
      "Sibianor aurocinctus\n",
      "Sitticus caricis\n",
      "Sitticus distinguendus\n",
      "Sitticus dzieduszyckii\n",
      "Sitticus floricola\n",
      "Sitticus rupicola\n",
      "Sitticus saltator\n",
      "Sitticus terebratus\n",
      "Synageles hilarulus\n",
      "Synageles venator\n",
      "Talavera aequipes\n",
      "Talavera petrensis\n",
      "Talavera parvistyla\n",
      "Yllenus arenarius"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "# s stands for substitute (the command within sed)\n",
    "# we define groups or variables of patterns using the parentheses\n",
    "# we call back the groups by index (order in which they ocurr)\n",
    "# anything other character (symbolised by .)\n",
    "# g stands for global - replace all matches, not just the first\n",
    "sed -r 's/^([A-Za-z]+) ([A-Za-z]+).+/\\1 \\2/g' LT_spider_list.txt\n",
    "\n",
    "# better yet - we can delete lines matching \"FAMILY\" first, then removing anything after \"genus species\" matches\n",
    "# then remove lines with any punctuation symbol\n",
    "sed '/FAMILY/d' LT_spider_list.txt | sed -r 's/^([A-Za-z]+) ([A-Za-z]+).+/\\1 \\2/g' | sed '/[[:punct:]]/d'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32ecd681",
   "metadata": {},
   "source": [
    "We can confirm how good the processing of the file is by sorting the lines and looking at unique strings. In other words, sort the lines to look for things that do not look like \"genus species\".\n",
    "\n",
    "In bioinformatics (and in life in general) is always good to **check and double check your results/output.** Particularly if you use match/replace, since there will be instances that you did not expect."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "ca17c0cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "# sort and count unique, then pass output to a file\n",
    "sed '/FAMILY/d' LT_spider_list.txt | sed -r 's/^([A-Za-z]+) ([A-Za-z]+).+/\\1 \\2/g' | sed '/[[:punct:]]/d' | sort | uniq -c > species.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "64a51b00",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      1 Abacoproeces saltum\n",
      "      1 Acantholycoa lignaria\n",
      "      1 Aculepeira ceropegia\n",
      "      1 Aellurillus v\n",
      "      1 Agalenatea redii\n",
      "      1 Agelena labyrinthica\n",
      "      1 Agraecina striata\n",
      "      1 Agroeca brunnea\n",
      "      1 Agroeca cuprea\n",
      "      1 Agroeca dentigera\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "head species.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2cfdae35",
   "metadata": {},
   "source": [
    "You will notice that `uniq -c` produces two columns, the first column is the frequency of a string and the second column is the string itself.\n",
    "\n",
    "You can **cut** the columns in a file and look to specific ones using `cut` and specifiying a field delimiter (like in Excel) using `cut -d\" \"`. In this case, the delimiter is any space (generally includes tabs, spaces, etc).\n",
    "\n",
    "You will also notice that `uniq -c` adds a TAB (or four spaces) before the first column, you can replace multiple spaces with a single one by **translating** the strings using `tr`.\n",
    "\n",
    "`cat` simply prints the file contents to the screen."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "415f9b2a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Abacoproeces saltum\n",
      "Acantholycoa lignaria\n",
      "Aculepeira ceropegia\n",
      "Aellurillus v\n",
      "Agalenatea redii\n",
      "Agelena labyrinthica\n",
      "Agraecina striata\n",
      "Agroeca brunnea\n",
      "Agroeca cuprea\n",
      "Agroeca dentigera\n",
      "Agroeca lusatica\n",
      "Agroeca proxima\n",
      "Agyneta cauta\n",
      "Agyneta conigera\n",
      "Agyneta decora\n",
      "Agyneta ramosa\n",
      "Agyneta subtilis\n",
      "Alopecosa aculeata\n",
      "Alopecosa barbipes\n",
      "Alopecosa cuneata\n",
      "Alopecosa cursor\n",
      "Alopecosa fabrilis\n",
      "Alopecosa inquilina\n",
      "Alopecosa mariae\n",
      "Alopecosa pinetorum\n",
      "Alopecosa pulverulenta\n",
      "Alopecosa trabalis\n",
      "Amaurobius fenestralis\n",
      "An old\n",
      "Anguliphantes angulipalpis\n",
      "Antistea elegans\n",
      "Anyphaena accentuata\n",
      "Aphileta misera\n",
      "Araeoncus humilis\n",
      "Araneus alsine\n",
      "Araneus angulatus\n",
      "Araneus diadematus\n",
      "Araneus marmoreus\n",
      "Araneus quadratus\n",
      "Araneus sturmi\n",
      "Araneus triguttatus\n",
      "Araniella cucurbitina\n",
      "Araniella displicata\n",
      "Archaeodictyna consecuta\n",
      "Arctosa alpigena\n",
      "Arctosa cinerea\n",
      "Arctosa leopardus\n",
      "Arctosa perita\n",
      "Arctosa stigmosa\n",
      "Argenna patula\n",
      "Argenna subnigra\n",
      "Argiope bruennichi\n",
      "Argyroneta aquatica\n",
      "Asianellus festivus\n",
      "Aulonia albimana\n",
      "Ballus chalybeius\n",
      "Bathyphantes approximatus\n",
      "Bathyphantes gracilis\n",
      "Bathyphantes nigrinus\n",
      "Bathyphantes parvulus\n",
      "Bathyphantes similis\n",
      "Berlandina cinerea\n",
      "Bolyphantes alticeps\n",
      "Bolyphantes luteolus\n",
      "Callilepis nocturna\n",
      "Centromerita bicolor\n",
      "Centromerus arcanus\n",
      "Centromerus brevivulvatus\n",
      "Centromerus incilium\n",
      "Centromerus levitarsis\n",
      "Centromerus sylvaticus\n",
      "Centromerus unidentatus\n",
      "Ceratinella brevipes\n",
      "Ceratinella brevis\n",
      "Ceratinella scabrosa\n",
      "Cercidia prominens\n",
      "Cheiracanthium erraticum\n",
      "Cheiracanthium montanum\n",
      "Cheiracanthium punctorium\n",
      "Cheiracanthium virescens\n",
      "Cicurina cicur\n",
      "Clubiona caerulescens\n",
      "Clubiona comta\n",
      "Clubiona diversa\n",
      "Clubiona frutetorum\n",
      "Clubiona germanica\n",
      "Clubiona juvenis\n",
      "Clubiona lutescens\n",
      "Clubiona marmorata\n",
      "Clubiona neglecta\n",
      "Clubiona norvegica\n",
      "Clubiona pallidula\n",
      "Clubiona phragmitis\n",
      "Clubiona reclusa\n",
      "Clubiona similis\n",
      "Clubiona stagnatilis\n",
      "Clubiona subsultans\n",
      "Clubiona trivialis\n",
      "Cnephalocotes obscurus\n",
      "Coriarachne depressa\n",
      "Crustulina guttata\n",
      "Cryphoeca silvicola\n",
      "Cryptachaea riparia\n",
      "Cyclosa conica\n",
      "Cyclosa oculata\n",
      "Dendryphantes hastatus\n",
      "Dendryphantes rudis\n",
      "Diaea dorsata\n",
      "Dictyna arundinacea\n",
      "Dictyna pusilla\n",
      "Dictyna uncinata\n",
      "Dicymbium nigrum\n",
      "Dicymbium tibiale\n",
      "Diplocentria bidentata\n",
      "Diplocephalus connatus\n",
      "Diplocephalus cristatus\n",
      "Diplocephalus dentatus\n",
      "Diplocephalus latifrons\n",
      "Diplocephalus permixtus\n",
      "Diplocephalus picinus\n",
      "Diplostyla concolor\n",
      "Dismodicus bifrons\n",
      "Dismodicus elevatus\n",
      "Dolomedes fimbriatus\n",
      "Dolomedes plantarius\n",
      "Drapetisca socialis\n",
      "Drassodes hypocrita\n",
      "Drassodes pubescens\n",
      "Drassodes villosus\n",
      "Drassyllus lutetianus\n",
      "Drassyllus praeficus\n",
      "Drassylus pusillus\n",
      "Drepanotylus uncatus\n",
      "Enoplognatha ovata\n",
      "Enoplognatha thoracica\n",
      "Episinus angulatus\n",
      "Episinus truncatus\n",
      "Eresus cinnaberinus\n",
      "Erigone atra\n",
      "Erigone dentipalpis\n",
      "Erigone longipalpis\n",
      "Erigonella hiemalis\n",
      "Erigonella ignobilis\n",
      "Ero cambridgei\n",
      "Ero furcata\n",
      "Euophrys frontalis\n",
      "Euryopis flavomaculata\n",
      "Evarcha arcuata\n",
      "Evarcha falcata\n",
      "Evarcha laetabunda\n",
      "Floronia bucculenta\n",
      "Formiphantes lepthyphantiformis\n",
      "Gibbaranea gibbosa\n",
      "Glyphesis cottonae\n",
      "Gnaphosa bicolor\n",
      "Gnaphosa lugubris\n",
      "Gnaphosa microps\n",
      "Gnaphosa montana\n",
      "Gnaphosa muscorum\n",
      "Gnaphosa nigerrima\n",
      "Gnathonarium dentatum\n",
      "Gonatium rubellum\n",
      "Gonatium rubens\n",
      "Gongylidiellum latebricola\n",
      "Gongylidiellum murcidum\n",
      "Gongylidium rufipes\n",
      "Hahnia nava\n",
      "Hahnia ononidum\n",
      "Hahnia pusilla\n",
      "Haplodrassus cognatus\n",
      "Haplodrassus dalmatensis\n",
      "Haplodrassus moderatus\n",
      "Haplodrassus signifer\n",
      "Haplodrassus silvestris\n",
      "Haplodrassus soerenseni\n",
      "Haplodrassus umbratilis\n",
      "Harpactea rubicunda\n",
      "Heliophanus auratus\n",
      "Heliophanus dampfi\n",
      "Heliophanus dubius\n",
      "Heliophanus flavipes\n",
      "Helophora insignis\n",
      "Hygrolycosa rubrofasciata\n",
      "Hylyphantes graminicola\n",
      "Hypomma bituberculatum\n",
      "Hypomma cornutum\n",
      "Hypselistes jacksoni\n",
      "Hypsosinga pygmaea\n",
      "Hypsosinga sanquinea\n",
      "Hyptiotes paradoxus\n",
      "Kaestneria dorsalis\n",
      "Kaestneria pullata\n",
      "Keijia tincta\n",
      "Larinioides cornutus\n",
      "Larinioides ixobolus\n",
      "Larinioides patagiatus\n",
      "Larinioides sclopetarius\n",
      "Lasaeola prona\n",
      "Lasaeola tristis\n",
      "Lepthyphantes leprosus\n",
      "Lessertia dentichelis\n",
      "Linyphia triangularis\n",
      "Lophomma punctatum\n",
      "Macrargus carpenteri\n",
      "Macrargus rufus\n",
      "Mangora acalypha\n",
      "Mansuphantes mansuetus\n",
      "Maro minutus\n",
      "Marpissa muscosa\n",
      "Marpissa radiata\n",
      "Maso sundevalli\n",
      "Mecynargus foveatus\n",
      "Megalepthyphantes nebulosus\n",
      "Meioneta affinis\n",
      "Meioneta fuscipalpa\n",
      "Meioneta innotabilis\n",
      "Meioneta mossica\n",
      "Meioneta rurestris\n",
      "Metellina mengei\n",
      "Metellina merianae\n",
      "Metellina segmentata\n",
      "Metopobactrus prominulus\n",
      "Micaria fulgens\n",
      "Micaria lenzi\n",
      "Micaria pulicaria\n",
      "Micaria silesiaca\n",
      "Micaria subopaca\n",
      "Micrargus apertus\n",
      "Micrargus herbigradus\n",
      "Micrargus subaequalis\n",
      "Microlinyphia pusilla\n",
      "Micromata virescens\n",
      "Microneta viaria\n",
      "Minicia marginella\n",
      "Minyriolus pusillus\n",
      "Mioxena blanda\n",
      "Misumena vatia\n",
      "Misumenops tricuspidatus\n",
      "Neon reticulatus\n",
      "Neon robustus\n",
      "Neon valentulus\n",
      "Neoscona adianta\n",
      "Neottiura bimaculata\n",
      "Neriene clathrata\n",
      "Neriene emphana\n",
      "Neriene marginata\n",
      "Neriene montana\n",
      "Neriene peltata\n",
      "Neriene radiata\n",
      "Notioscopus sarcinatus\n",
      "Nuctenea silvicultrix\n",
      "Nuctenea umbratica\n",
      "Oedothorax apicatus\n",
      "Oedothorax gibbosus\n",
      "Oedothorax retusus\n",
      "Old records\n",
      "Oryphantes angulatus\n",
      "Oxyopes ramosus\n",
      "Ozyptila atomaria\n",
      "Ozyptila brevipes\n",
      "Ozyptila praticola\n",
      "Ozyptila scabricula\n",
      "Ozyptila simplex\n",
      "Ozyptila trux\n",
      "Pachygnatha clercki\n",
      "Pachygnatha degeeri\n",
      "Pachygnatha listeri\n",
      "Palliduphantes insignis\n",
      "Palliduphantes pallidus\n",
      "Parasteatoda lunata\n",
      "Parasteatoda simulans\n",
      "Pardosa agrestis\n",
      "Pardosa agricola\n",
      "Pardosa amentata\n",
      "Pardosa bifasciata\n",
      "Pardosa hyperborea\n",
      "Pardosa lugubris\n",
      "Pardosa monticola\n",
      "Pardosa paludicola\n",
      "Pardosa palustris\n",
      "Pardosa prativaga\n",
      "Pardosa pullata\n",
      "Pardosa riparia\n",
      "Pardosa schenkeli\n",
      "Pardosa sphagnicola\n",
      "Pelecopsis elongata\n",
      "Pelecopsis parallela\n",
      "Pelenes tripunctatus\n",
      "Philaeus chrysops\n",
      "Philodromus aureolus\n",
      "Philodromus cespitum\n",
      "Philodromus collinus\n",
      "Philodromus emarginatus\n",
      "Philodromus fuscomarginatus\n",
      "Philodromus histrio\n",
      "Philodromus margaritatus\n",
      "Philodromus poecilus\n",
      "Phlegra fasciata\n",
      "Pholcomma gibbum\n",
      "Pholcus phalangioides\n",
      "Phrurolithus festivus\n",
      "Phrurolithus minimus\n",
      "Phyloneta impressa\n",
      "Phyloneta sisyphia\n",
      "Pirata hygrophilus\n",
      "Pirata insularis\n",
      "Pirata latitans\n",
      "Pirata piraticus\n",
      "Pirata piscatorius\n",
      "Pirata tenuitarsis\n",
      "Pirata uliginosus\n",
      "Pisaura mirabilis\n",
      "Pityohyphantes phrygianus\n",
      "Pocadicnemis pumila\n",
      "Porrhomma microphthalmum\n",
      "Porrhomma oblitum\n",
      "Porrhomma pallidum\n",
      "Porrhomma pygmaeum\n",
      "Pseudeuophrys erratica\n",
      "Robertus arundineti\n",
      "Robertus lividus\n",
      "Robertus lyrifer\n",
      "Robertus neglectus\n",
      "Robertus scoticus\n",
      "Robertus ungulatus\n",
      "Saaristoa abnormis\n",
      "Salticus cingulatus\n",
      "Salticus scenicus\n",
      "Savignia frontata\n",
      "Scotina palliardi\n",
      "Scotophaeus blackwalli\n",
      "Scotophaeus scutulatus\n",
      "Segestria senoculata\n",
      "Sibianor aurocinctus\n",
      "Silometopus reussi\n",
      "Simitidion simile\n",
      "Singa albovittata\n",
      "Singa hamata\n",
      "Singa nitidula\n",
      "Sintula cornigera\n",
      "Sitticus caricis\n",
      "Sitticus distinguendus\n",
      "Sitticus dzieduszyckii\n",
      "Sitticus floricola\n",
      "Sitticus rupicola\n",
      "Sitticus saltator\n",
      "Sitticus terebratus\n",
      "Sosticus loricatus\n",
      "Steatoda albomaculata\n",
      "Steatoda bipunctata\n",
      "Steatoda castanea\n",
      "Steatoda grossa\n",
      "Steatoda phalerata\n",
      "Stemonyphantes lineatus\n",
      "Stroemiellus stroemi\n",
      "Synageles hilarulus\n",
      "Synageles venator\n",
      "THE CHECKLIST\n",
      "Talavera aequipes\n",
      "Talavera parvistyla\n",
      "Talavera petrensis\n",
      "Tallusia experta\n",
      "Tapinocyba biscissa\n",
      "Tapinocyba insecta\n",
      "Tapinocyba pallens\n",
      "Tapinocyba praecox\n",
      "Tapinopa longidens\n",
      "Taranucnus setosus\n",
      "Tegenaria agrestis\n",
      "Tegenaria atrica\n",
      "Tegenaria domestica\n",
      "Tenuiphantes alacris\n",
      "Tenuiphantes cristatus\n",
      "Tenuiphantes flavipes\n",
      "Tenuiphantes mengei\n",
      "Tenuiphantes tenebricola\n",
      "Tetragnatha extensa\n",
      "Tetragnatha montana\n",
      "Tetragnatha nigrita\n",
      "Tetragnatha obtusa\n",
      "Tetragnatha pinicola\n",
      "Tetragnatha striata\n",
      "Thanatus arenarius\n",
      "Thanatus formicinus\n",
      "Thanatus striatus\n",
      "The status\n",
      "Theonoe minutissima\n",
      "Theridion mystaceum\n",
      "Theridion pictum\n",
      "Theridion pinastri\n",
      "Theridion varians\n",
      "Thomisus onustus\n",
      "Thyreosthenius parasiticus\n",
      "Tibellus maritimus\n",
      "Tibellus oblongus\n",
      "Tiso vagans\n",
      "Tmeticus affinis\n",
      "Trochosa robusta\n",
      "Trochosa ruricola\n",
      "Trochosa spinipalpis\n",
      "Trochosa terricola\n",
      "Troxochrus scabriculus\n",
      "Walckeaneria nodosa\n",
      "Walckenaeria acuminata\n",
      "Walckenaeria alticeps\n",
      "Walckenaeria antica\n",
      "Walckenaeria atrotibialis\n",
      "Walckenaeria capito\n",
      "Walckenaeria cucullata\n",
      "Walckenaeria cuspidata\n",
      "Walckenaeria dysderoides\n",
      "Walckenaeria incisa\n",
      "Walckenaeria karpinskii\n",
      "Walckenaeria kochi\n",
      "Walckenaeria mitrata\n",
      "Walckenaeria nudipalpis\n",
      "Walckenaeria obtusa\n",
      "Walckenaeria unicornis\n",
      "Walckenaeria vigilax\n",
      "Xerolycosa miniata\n",
      "Xerolycosa nemoralis\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Xysticus audax\n",
      "Xysticus bifasciatus\n",
      "Xysticus cristatus\n",
      "Xysticus erraticus\n",
      "Xysticus kochi\n",
      "Xysticus lanio\n",
      "Xysticus lineatus\n",
      "Xysticus luctuosus\n",
      "Xysticus obscurus\n",
      "Xysticus sabulosus\n",
      "Xysticus ulmi\n",
      "Yllenus arenarius\n",
      "Zelotes aeneus\n",
      "Zelotes clivicola\n",
      "Zelotes electus\n",
      "Zelotes exiguus\n",
      "Zelotes latreillei\n",
      "Zelotes longipes\n",
      "Zelotes petrensis\n",
      "Zelotes subterraneus\n",
      "Zora nemoralis\n",
      "Zora silvestris\n",
      "Zora spinimana\n",
      "Zornella cultrigera\n",
      "Zygiella atrica\n",
      "Zygiella x\n",
      "all records\n",
      "and was\n",
      "not be\n",
      "were wrongly\n",
      "Abacoproeces saltum\n",
      "Acantholycoa lignaria\n",
      "Aculepeira ceropegia\n",
      "Aellurillus v\n",
      "Agalenatea redii\n",
      "Agelena labyrinthica\n",
      "Agraecina striata\n",
      "Agroeca brunnea\n",
      "Agroeca cuprea\n",
      "Agroeca dentigera\n",
      "Agroeca lusatica\n",
      "Agroeca proxima\n",
      "Agyneta cauta\n",
      "Agyneta conigera\n",
      "Agyneta decora\n",
      "Agyneta ramosa\n",
      "Agyneta subtilis\n",
      "Alopecosa aculeata\n",
      "Alopecosa barbipes\n",
      "Alopecosa cuneata\n",
      "Alopecosa cursor\n",
      "Alopecosa fabrilis\n",
      "Alopecosa inquilina\n",
      "Alopecosa mariae\n",
      "Alopecosa pinetorum\n",
      "Alopecosa pulverulenta\n",
      "Alopecosa trabalis\n",
      "Amaurobius fenestralis\n",
      "An old\n",
      "Anguliphantes angulipalpis\n",
      "Antistea elegans\n",
      "Anyphaena accentuata\n",
      "Aphileta misera\n",
      "Araeoncus humilis\n",
      "Araneus alsine\n",
      "Araneus angulatus\n",
      "Araneus diadematus\n",
      "Araneus marmoreus\n",
      "Araneus quadratus\n",
      "Araneus sturmi\n",
      "Araneus triguttatus\n",
      "Araniella cucurbitina\n",
      "Araniella displicata\n",
      "Archaeodictyna consecuta\n",
      "Arctosa alpigena\n",
      "Arctosa cinerea\n",
      "Arctosa leopardus\n",
      "Arctosa perita\n",
      "Arctosa stigmosa\n",
      "Argenna patula\n",
      "Argenna subnigra\n",
      "Argiope bruennichi\n",
      "Argyroneta aquatica\n",
      "Asianellus festivus\n",
      "Aulonia albimana\n",
      "Ballus chalybeius\n",
      "Bathyphantes approximatus\n",
      "Bathyphantes gracilis\n",
      "Bathyphantes nigrinus\n",
      "Bathyphantes parvulus\n",
      "Bathyphantes similis\n",
      "Berlandina cinerea\n",
      "Bolyphantes alticeps\n",
      "Bolyphantes luteolus\n",
      "Callilepis nocturna\n",
      "Centromerita bicolor\n",
      "Centromerus arcanus\n",
      "Centromerus brevivulvatus\n",
      "Centromerus incilium\n",
      "Centromerus levitarsis\n",
      "Centromerus sylvaticus\n",
      "Centromerus unidentatus\n",
      "Ceratinella brevipes\n",
      "Ceratinella brevis\n",
      "Ceratinella scabrosa\n",
      "Cercidia prominens\n",
      "Cheiracanthium erraticum\n",
      "Cheiracanthium montanum\n",
      "Cheiracanthium punctorium\n",
      "Cheiracanthium virescens\n",
      "Cicurina cicur\n",
      "Clubiona caerulescens\n",
      "Clubiona comta\n",
      "Clubiona diversa\n",
      "Clubiona frutetorum\n",
      "Clubiona germanica\n",
      "Clubiona juvenis\n",
      "Clubiona lutescens\n",
      "Clubiona marmorata\n",
      "Clubiona neglecta\n",
      "Clubiona norvegica\n",
      "Clubiona pallidula\n",
      "Clubiona phragmitis\n",
      "Clubiona reclusa\n",
      "Clubiona similis\n",
      "Clubiona stagnatilis\n",
      "Clubiona subsultans\n",
      "Clubiona trivialis\n",
      "Cnephalocotes obscurus\n",
      "Coriarachne depressa\n",
      "Crustulina guttata\n",
      "Cryphoeca silvicola\n",
      "Cryptachaea riparia\n",
      "Cyclosa conica\n",
      "Cyclosa oculata\n",
      "Dendryphantes hastatus\n",
      "Dendryphantes rudis\n",
      "Diaea dorsata\n",
      "Dictyna arundinacea\n",
      "Dictyna pusilla\n",
      "Dictyna uncinata\n",
      "Dicymbium nigrum\n",
      "Dicymbium tibiale\n",
      "Diplocentria bidentata\n",
      "Diplocephalus connatus\n",
      "Diplocephalus cristatus\n",
      "Diplocephalus dentatus\n",
      "Diplocephalus latifrons\n",
      "Diplocephalus permixtus\n",
      "Diplocephalus picinus\n",
      "Diplostyla concolor\n",
      "Dismodicus bifrons\n",
      "Dismodicus elevatus\n",
      "Dolomedes fimbriatus\n",
      "Dolomedes plantarius\n",
      "Drapetisca socialis\n",
      "Drassodes hypocrita\n",
      "Drassodes pubescens\n",
      "Drassodes villosus\n",
      "Drassyllus lutetianus\n",
      "Drassyllus praeficus\n",
      "Drassylus pusillus\n",
      "Drepanotylus uncatus\n",
      "Enoplognatha ovata\n",
      "Enoplognatha thoracica\n",
      "Episinus angulatus\n",
      "Episinus truncatus\n",
      "Eresus cinnaberinus\n",
      "Erigone atra\n",
      "Erigone dentipalpis\n",
      "Erigone longipalpis\n",
      "Erigonella hiemalis\n",
      "Erigonella ignobilis\n",
      "Ero cambridgei\n",
      "Ero furcata\n",
      "Euophrys frontalis\n",
      "Euryopis flavomaculata\n",
      "Evarcha arcuata\n",
      "Evarcha falcata\n",
      "Evarcha laetabunda\n",
      "Floronia bucculenta\n",
      "Formiphantes lepthyphantiformis\n",
      "Gibbaranea gibbosa\n",
      "Glyphesis cottonae\n",
      "Gnaphosa bicolor\n",
      "Gnaphosa lugubris\n",
      "Gnaphosa microps\n",
      "Gnaphosa montana\n",
      "Gnaphosa muscorum\n",
      "Gnaphosa nigerrima\n",
      "Gnathonarium dentatum\n",
      "Gonatium rubellum\n",
      "Gonatium rubens\n",
      "Gongylidiellum latebricola\n",
      "Gongylidiellum murcidum\n",
      "Gongylidium rufipes\n",
      "Hahnia nava\n",
      "Hahnia ononidum\n",
      "Hahnia pusilla\n",
      "Haplodrassus cognatus\n",
      "Haplodrassus dalmatensis\n",
      "Haplodrassus moderatus\n",
      "Haplodrassus signifer\n",
      "Haplodrassus silvestris\n",
      "Haplodrassus soerenseni\n",
      "Haplodrassus umbratilis\n",
      "Harpactea rubicunda\n",
      "Heliophanus auratus\n",
      "Heliophanus dampfi\n",
      "Heliophanus dubius\n",
      "Heliophanus flavipes\n",
      "Helophora insignis\n",
      "Hygrolycosa rubrofasciata\n",
      "Hylyphantes graminicola\n",
      "Hypomma bituberculatum\n",
      "Hypomma cornutum\n",
      "Hypselistes jacksoni\n",
      "Hypsosinga pygmaea\n",
      "Hypsosinga sanquinea\n",
      "Hyptiotes paradoxus\n",
      "Kaestneria dorsalis\n",
      "Kaestneria pullata\n",
      "Keijia tincta\n",
      "Larinioides cornutus\n",
      "Larinioides ixobolus\n",
      "Larinioides patagiatus\n",
      "Larinioides sclopetarius\n",
      "Lasaeola prona\n",
      "Lasaeola tristis\n",
      "Lepthyphantes leprosus\n",
      "Lessertia dentichelis\n",
      "Linyphia triangularis\n",
      "Lophomma punctatum\n",
      "Macrargus carpenteri\n",
      "Macrargus rufus\n",
      "Mangora acalypha\n",
      "Mansuphantes mansuetus\n",
      "Maro minutus\n",
      "Marpissa muscosa\n",
      "Marpissa radiata\n",
      "Maso sundevalli\n",
      "Mecynargus foveatus\n",
      "Megalepthyphantes nebulosus\n",
      "Meioneta affinis\n",
      "Meioneta fuscipalpa\n",
      "Meioneta innotabilis\n",
      "Meioneta mossica\n",
      "Meioneta rurestris\n",
      "Metellina mengei\n",
      "Metellina merianae\n",
      "Metellina segmentata\n",
      "Metopobactrus prominulus\n",
      "Micaria fulgens\n",
      "Micaria lenzi\n",
      "Micaria pulicaria\n",
      "Micaria silesiaca\n",
      "Micaria subopaca\n",
      "Micrargus apertus\n",
      "Micrargus herbigradus\n",
      "Micrargus subaequalis\n",
      "Microlinyphia pusilla\n",
      "Micromata virescens\n",
      "Microneta viaria\n",
      "Minicia marginella\n",
      "Minyriolus pusillus\n",
      "Mioxena blanda\n",
      "Misumena vatia\n",
      "Misumenops tricuspidatus\n",
      "Neon reticulatus\n",
      "Neon robustus\n",
      "Neon valentulus\n",
      "Neoscona adianta\n",
      "Neottiura bimaculata\n",
      "Neriene clathrata\n",
      "Neriene emphana\n",
      "Neriene marginata\n",
      "Neriene montana\n",
      "Neriene peltata\n",
      "Neriene radiata\n",
      "Notioscopus sarcinatus\n",
      "Nuctenea silvicultrix\n",
      "Nuctenea umbratica\n",
      "Oedothorax apicatus\n",
      "Oedothorax gibbosus\n",
      "Oedothorax retusus\n",
      "Old records\n",
      "Oryphantes angulatus\n",
      "Oxyopes ramosus\n",
      "Ozyptila atomaria\n",
      "Ozyptila brevipes\n",
      "Ozyptila praticola\n",
      "Ozyptila scabricula\n",
      "Ozyptila simplex\n",
      "Ozyptila trux\n",
      "Pachygnatha clercki\n",
      "Pachygnatha degeeri\n",
      "Pachygnatha listeri\n",
      "Palliduphantes insignis\n",
      "Palliduphantes pallidus\n",
      "Parasteatoda lunata\n",
      "Parasteatoda simulans\n",
      "Pardosa agrestis\n",
      "Pardosa agricola\n",
      "Pardosa amentata\n",
      "Pardosa bifasciata\n",
      "Pardosa hyperborea\n",
      "Pardosa lugubris\n",
      "Pardosa monticola\n",
      "Pardosa paludicola\n",
      "Pardosa palustris\n",
      "Pardosa prativaga\n",
      "Pardosa pullata\n",
      "Pardosa riparia\n",
      "Pardosa schenkeli\n",
      "Pardosa sphagnicola\n",
      "Pelecopsis elongata\n",
      "Pelecopsis parallela\n",
      "Pelenes tripunctatus\n",
      "Philaeus chrysops\n",
      "Philodromus aureolus\n",
      "Philodromus cespitum\n",
      "Philodromus collinus\n",
      "Philodromus emarginatus\n",
      "Philodromus fuscomarginatus\n",
      "Philodromus histrio\n",
      "Philodromus margaritatus\n",
      "Philodromus poecilus\n",
      "Phlegra fasciata\n",
      "Pholcomma gibbum\n",
      "Pholcus phalangioides\n",
      "Phrurolithus festivus\n",
      "Phrurolithus minimus\n",
      "Phyloneta impressa\n",
      "Phyloneta sisyphia\n",
      "Pirata hygrophilus\n",
      "Pirata insularis\n",
      "Pirata latitans\n",
      "Pirata piraticus\n",
      "Pirata piscatorius\n",
      "Pirata tenuitarsis\n",
      "Pirata uliginosus\n",
      "Pisaura mirabilis\n",
      "Pityohyphantes phrygianus\n",
      "Pocadicnemis pumila\n",
      "Porrhomma microphthalmum\n",
      "Porrhomma oblitum\n",
      "Porrhomma pallidum\n",
      "Porrhomma pygmaeum\n",
      "Pseudeuophrys erratica\n",
      "Robertus arundineti\n",
      "Robertus lividus\n",
      "Robertus lyrifer\n",
      "Robertus neglectus\n",
      "Robertus scoticus\n",
      "Robertus ungulatus\n",
      "Saaristoa abnormis\n",
      "Salticus cingulatus\n",
      "Salticus scenicus\n",
      "Savignia frontata\n",
      "Scotina palliardi\n",
      "Scotophaeus blackwalli\n",
      "Scotophaeus scutulatus\n",
      "Segestria senoculata\n",
      "Sibianor aurocinctus\n",
      "Silometopus reussi\n",
      "Simitidion simile\n",
      "Singa albovittata\n",
      "Singa hamata\n",
      "Singa nitidula\n",
      "Sintula cornigera\n",
      "Sitticus caricis\n",
      "Sitticus distinguendus\n",
      "Sitticus dzieduszyckii\n",
      "Sitticus floricola\n",
      "Sitticus rupicola\n",
      "Sitticus saltator\n",
      "Sitticus terebratus\n",
      "Sosticus loricatus\n",
      "Steatoda albomaculata\n",
      "Steatoda bipunctata\n",
      "Steatoda castanea\n",
      "Steatoda grossa\n",
      "Steatoda phalerata\n",
      "Stemonyphantes lineatus\n",
      "Stroemiellus stroemi\n",
      "Synageles hilarulus\n",
      "Synageles venator\n",
      "THE CHECKLIST\n",
      "Talavera aequipes\n",
      "Talavera parvistyla\n",
      "Talavera petrensis\n",
      "Tallusia experta\n",
      "Tapinocyba biscissa\n",
      "Tapinocyba insecta\n",
      "Tapinocyba pallens\n",
      "Tapinocyba praecox\n",
      "Tapinopa longidens\n",
      "Taranucnus setosus\n",
      "Tegenaria agrestis\n",
      "Tegenaria atrica\n",
      "Tegenaria domestica\n",
      "Tenuiphantes alacris\n",
      "Tenuiphantes cristatus\n",
      "Tenuiphantes flavipes\n",
      "Tenuiphantes mengei\n",
      "Tenuiphantes tenebricola\n",
      "Tetragnatha extensa\n",
      "Tetragnatha montana\n",
      "Tetragnatha nigrita\n",
      "Tetragnatha obtusa\n",
      "Tetragnatha pinicola\n",
      "Tetragnatha striata\n",
      "Thanatus arenarius\n",
      "Thanatus formicinus\n",
      "Thanatus striatus\n",
      "The status\n",
      "Theonoe minutissima\n",
      "Theridion mystaceum\n",
      "Theridion pictum\n",
      "Theridion pinastri\n",
      "Theridion varians\n",
      "Thomisus onustus\n",
      "Thyreosthenius parasiticus\n",
      "Tibellus maritimus\n",
      "Tibellus oblongus\n",
      "Tiso vagans\n",
      "Tmeticus affinis\n",
      "Trochosa robusta\n",
      "Trochosa ruricola\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Trochosa spinipalpis\n",
      "Trochosa terricola\n",
      "Troxochrus scabriculus\n",
      "Walckeaneria nodosa\n",
      "Walckenaeria acuminata\n",
      "Walckenaeria alticeps\n",
      "Walckenaeria antica\n",
      "Walckenaeria atrotibialis\n",
      "Walckenaeria capito\n",
      "Walckenaeria cucullata\n",
      "Walckenaeria cuspidata\n",
      "Walckenaeria dysderoides\n",
      "Walckenaeria incisa\n",
      "Walckenaeria karpinskii\n",
      "Walckenaeria kochi\n",
      "Walckenaeria mitrata\n",
      "Walckenaeria nudipalpis\n",
      "Walckenaeria obtusa\n",
      "Walckenaeria unicornis\n",
      "Walckenaeria vigilax\n",
      "Xerolycosa miniata\n",
      "Xerolycosa nemoralis\n",
      "Xysticus audax\n",
      "Xysticus bifasciatus\n",
      "Xysticus cristatus\n",
      "Xysticus erraticus\n",
      "Xysticus kochi\n",
      "Xysticus lanio\n",
      "Xysticus lineatus\n",
      "Xysticus luctuosus\n",
      "Xysticus obscurus\n",
      "Xysticus sabulosus\n",
      "Xysticus ulmi\n",
      "Yllenus arenarius\n",
      "Zelotes aeneus\n",
      "Zelotes clivicola\n",
      "Zelotes electus\n",
      "Zelotes exiguus\n",
      "Zelotes latreillei\n",
      "Zelotes longipes\n",
      "Zelotes petrensis\n",
      "Zelotes subterraneus\n",
      "Zora nemoralis\n",
      "Zora silvestris\n",
      "Zora spinimana\n",
      "Zornella cultrigera\n",
      "Zygiella atrica\n",
      "Zygiella x\n",
      "all records\n",
      "and was\n",
      "not be\n",
      "were wrongly\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "# print second column if delimiter is a single space\n",
    "# cut shows the fields 3 and 4.\n",
    "cat species.txt | tr -s ' ' | cut -d' ' -f3,4\n",
    "\n",
    "# the above transformation results in the same output as the code below\n",
    "cut -d' ' -f8,9 species.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df4a1795",
   "metadata": {},
   "source": [
    "Similarly, to check the unique genera in the list, we can combine `cat`, `tr`, `sort`, and `unique`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "209a02b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cat species.txt | tr -s ' ' | cut -d' ' -f3 | sort | uniq > genera.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "67917e13",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Abacoproeces\n",
      "Acantholycoa\n",
      "Aculepeira\n",
      "Aellurillus\n",
      "Agalenatea\n",
      "Agelena\n",
      "Agraecina\n",
      "Agroeca\n",
      "Agyneta\n",
      "Alopecosa\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "head genera.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2c067f8",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-success\">\n",
    "    <b>TO DO: </b>What is happening? Is it working perfectly or not? How many genera is there? How could you process this file more efficiently? Manually add your answer to <i>answersL0_name.txt</i> using nano.</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7bdaeefa",
   "metadata": {},
   "source": [
    "To speed up the process of sequence downloads, we will select the first 10 genera."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5beacfd8",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-success\">\n",
    "    <b>TO DO: </b>Create a file called <i>genera_filtered.txt</i> with the first 10 genera from <i>genera.txt</i>. We will search for COI sequences for the genera in the list by processing one line at a time. We will use loops and variables to achieve it.</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7cb4d41",
   "metadata": {},
   "source": [
    "## 3.4. Variables, loops, and conditionals  \n",
    "\n",
    "A variable is a symbolic name that representes a value that can be asigned and re-assigned depending on the context they are used. Variables are useful for generalising programs, creating pipelines where only one imput/argument changes, or for making loops more efficiently.\n",
    "\n",
    "You can assign a value to a the variable NAME llike this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "ae031f72",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "my_name\n",
      "15\n",
      "NAME\n",
      "AGE\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "NAME='my_name'\n",
    "AGE=15\n",
    "\n",
    "# check the variable values using the echo function\n",
    "echo $NAME\n",
    "echo $AGE\n",
    "\n",
    "# what happens if you miss the $ symbol?\n",
    "echo NAME\n",
    "echo AGE"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76ff9444",
   "metadata": {},
   "source": [
    "A variable can also be the output of a command, for example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "7aef3be3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Agroeca\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "GENUS=$(head -n 10 species.txt | tail -n 1 | sed -r 's/([A-Za-z]+) .+/\\1/g' | cut -d' ' -f8)\n",
    "echo $GENUS"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a837b30",
   "metadata": {},
   "source": [
    "A **loop** is a structure that repeat a set of commands multiple times based on a certain condition. It could be a array of items being exhausted or a variable being equal to a defined value.\n",
    "\n",
    "Two common loops in bioinformatics are `for` and `while`. `for` runs a command in a array of items (variables, files, numbers, etc) until the array is empty. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "3a8f64e5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Abacoproeces\n",
      "Agelena\n",
      "Alopecosa\n",
      "Gnaphosa\n",
      "Haplodrassus\n",
      "Abacoproeces\n",
      "Agelena\n",
      "Alopecosa\n",
      "Gnaphosa\n",
      "Haplodrassus\n",
      "Genus: Abacoproeces\n",
      "Genus: Agelena\n",
      "Genus: Alopecosa\n",
      "Genus: Gnaphosa\n",
      "Genus: Haplodrassus\n",
      "Count is 1\n",
      "Count is 2\n",
      "Count is 3\n",
      "Count is 4\n",
      "Count is 5\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "for genus in Abacoproeces Agelena Alopecosa Gnaphosa Haplodrassus; do\n",
    "  echo $genus;\n",
    "done\n",
    "\n",
    "# or \n",
    "\n",
    "for genus in Abacoproeces Agelena Alopecosa Gnaphosa Haplodrassus; do echo $genus; done\n",
    "for genus in Abacoproeces Agelena Alopecosa Gnaphosa Haplodrassus; do echo \"Genus: $genus\"; done\n",
    "# note above that passing variables and text or passing variables to grep sed, needs double quotes\n",
    "\n",
    "count=1\n",
    "while [ $count -le 5 ]; do\n",
    "    echo \"Count is $count\";\n",
    "    ((count++));\n",
    "done\n",
    "\n",
    "# above, the -le argument stands for \"less than or equal to\"\n",
    "# count++ adds count to count. as count == 1, then it adds 1 at every loop"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "728205d0",
   "metadata": {},
   "source": [
    "Now, we will use the `genera_filtered.txt` file to search for all [COI - cytochrome c oxidase I](https://en.wikipedia.org/wiki/Cytochrome_c_oxidase_subunit_I) sequences for each genus available on NCBI.\n",
    "\n",
    "We will use a couple of commands from the Blast+ suit, `esearch` and `efetch`. We will learn more about these in more depth in another lecture but for now, `esearch` creates a search query and returns an NCBI URL with the results of the search. `efetch` then takes the URL and retrieves the results in the format chosen by the user, in our case, fasta format.\n",
    "\n",
    "Note that we are reading `genera_filtered.txt` line by line and asigning the line value to the `$GENUS` variable. Then, we pass that variable to `esearch` and `efetch` to retrieve the sequences in fasta format (we will learn more about sequence formats later)."
   ]
  },
  {
   "cell_type": "raw",
   "id": "dc35a5c5",
   "metadata": {},
   "source": [
    "%%bash\n",
    "\n",
    "while IFS='' read -u 9 -r line; do\n",
    "    GENUS=$(echo \"$line\" | cut -f1);\n",
    "    echo $GENUS;\n",
    "    esearch -db nucleotide -query \"${GENUS}[Organism] AND cytochrome[Titl]\" | efetch -format fasta >> spiders.fasta;\n",
    "    esearch -db nucleotide -query \"${GENUS}[Organism] AND cytochrome[Titl]\" | efetch -format gb >> spiders.gb;\n",
    "done 9< genera_filtered.txt\n",
    "\n",
    "grep \"$GENUS\" spiders.fasta | wc -l"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6534221",
   "metadata": {},
   "source": [
    "**Important note:** Keep in mind that different search terms will output different results. For example, searching for \"cytochrome\" will return sequences with either *cytochrome* or *cytochrome II* or *cytochrome I* in the title and will not return sequences without *cytochrome* but with *COI* in the title.\n",
    "\n",
    "The code above creates the URL with the results from searching sequences from each genus and with *cytochrome* in the sequence title. Then, one line downloads the results in fasta format (only sequences) and in GeneBank (gb) format.\n",
    "\n",
    "In fasta files, the line containing the information about the sequence starts with the '\\>' symbol whereas the lines containing the sequences will not have any symbol at all. Thus, you can count the number of sequences in the fasta file by using `grep` to find the '\\>' symbol in the `spiders.fasta` file, then pipe the result to `wc` and count the number of lines."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96162bac",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-success\">\n",
    "    <b>TO DO: </b>How many sequences are there in the <i>spiders.fasta</i> file? Append the answer to your <i>answersL0_name.txt</i> using <i>>></i>.</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b08835ae",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-success\">\n",
    "    <b>TO DO: </b>Use a <i>for</i> loop to grep each genera from the <i>genera_filtered.txt</i> file and count how many sequences were downloaded.</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0d8ddfc",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-success\">\n",
    "    <b>TO DO: </b>Which spider genera do NOT have sequences available? Manually add you answer to your <i>answersL0_name.txt</i> using <i>nano</i>.</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f9976df",
   "metadata": {},
   "source": [
    "To check the publications associated with the sequence submissions, you can use `grep` and include in the results one extra line for a bit of context.\n",
    "\n",
    "Check for the country and `lat_lon` fields associated with every sequence in the `gb` (or GenBank) file. Where are the vouchers (the source of the sequence) from? Where were they collected? Is any voucher collected in Lithuania?"
   ]
  },
  {
   "cell_type": "raw",
   "id": "0aebc8bf",
   "metadata": {},
   "source": [
    "%%bash\n",
    "grep -A 1 'TITLE' spiders.gb\n",
    "grep 'country' spiders.gb\n",
    "grep 'lat_lon' spiders.gb"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87cadfa7",
   "metadata": {},
   "source": [
    "Finally, but not less importantly, we want to understand the quality state of our data.\n",
    "\n",
    "We downloaded COI Sequences from spiders. COI is a universaly used DNA barcode that is sequenced using primers, oligonucleotides that match a conserved region near the sequence of interest that *prime* the taq polymerase to start the synthesis of the complementary strand (after denaturation and annealing).\n",
    "\n",
    "Primers are often sequenced together with the rest of the barcode fragment and it is advisable to remove their sequences from the data before proceeding to do phylogenetic analyses. However, sequences submited to NCBI are not always curated and clean and we must identify which sequences still have the primers. This is easily done with `grep`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "a52945cd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATAGATATTGGGACTTTATATTTAATTTTTGGGGCTTGGTCTGCTATAGTGGGGACAGCTATAAGAGTAT\n",
      "ATAGATATTGGGACTTTATATTTAATTTTTGGGGCTTGGTCTGCTATAGTGGGGACAGCTATAAGAGTAT\n",
      "TTTTCAACTAATCATAAAGATATTGGAAGTTTATATTTTATTTTTGGAGCTTGAGCGGCTATAGTTGGAA\n",
      "AATCATAAAGATATTGGAACTTTATATTTAATGTTAGGTGTTTGATCGGCTATAATAGGGACTGCTATAT\n",
      "ATAGATATTGGGACTTTATATTTACTGTTAGGTGTTTGATCGGCAATAATAGGTACTGCTATATCAGTGT\n",
      "ATATCTTTA\n",
      "TTATTATTGTTTATTTCTTCTATAGCTGAAATAGGAGTTGGTGCTGGTTGAACTGTTTATCCTCCTTTAG\n",
      "CATCTAGAGTTGGTCATGTAGGTAGTGCTATGGATTTTGCTATTTTTTCTTTACATTTAGCAGGGGCATC\n",
      "TTCAATTATAGGTGCTGTAAATTTCATTTCTACTATTATAAATATACGTTCTACTGGAATATCTATAGAG\n",
      "AAGGTATCTTTGTTTGTTTGATCGGTATTTATTACTGCAGTATTATGATTAATATCTTTACCTGTATTAG\n",
      "CAGGAGCAATTACCATATTATTGACTGATCGGAATTTTAATACTTCTTTTTTTGAACCTGCAAGGGGGGG\n",
      "AGATCCTATTTGACATCAACATT\n",
      ">MW996910.1 Agroeca inopina voucher UB-MD615 cytochrome oxidase subunit 1 (COI) gene, partial cds; mitochondrial\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "grep 'AGATATTGG' spiders.fasta # LCO 1490 (Miller et. al., 2013)\n",
    "\n",
    "# it is possible to search for the complementary and reversed sequence of a primer in the sequences downloaded\n",
    "# primer Lepidoptera Forward (Hebert et. al., 2004)\n",
    "# note that the code is not using the pipe but giving three separate instructions\n",
    "# each instruction is separated by a ;\n",
    "# this is rarely done because reading the code gets difficult.\n",
    "# -C provides n lines of context\n",
    "\n",
    "REVSEQ=$(echo 'TAAAGATAT' | tr ACGTacgt TGCAtgca | rev ); echo $REVSEQ; grep -C 3 \"$REVSEQ\" spiders.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3dbf08fa",
   "metadata": {},
   "source": [
    "Think about this. What implications exist if the primer sequence is at the begining, at the end, or in the middle of the sequence?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff0319be",
   "metadata": {},
   "source": [
    "## 3.5. Compress and dicompress\n",
    "\n",
    "Compressing and decompressing files is a common task in bioinformatics and it [consists in reducing the size of a file by reducing the bits needed to encode the same information](https://en.wikipedia.org/wiki/Data_compression).\n",
    "\n",
    "Compressing files is useful because pipelines (a series of interconnected programs/analyses to process data, from raw to final, often generate large files that fill the disk space pretty quickly. It is also useful as smaller files are easier to transfer.\n",
    "\n",
    "Common compression and decompression pairs of programs are (respectively) `gzip` and `gunzip`, and the parallelizable `pigz` and `pigz -dc`. Another popular program is `tar`, which stores multiple files in a single one and has compression options (is not limited to just compress/decompress single files)\n",
    "\n",
    "Explore the options for each of those programs, particularly how to deal with the output. Which one creates a file automatically? Which one sends the output to *standard output* ([stdout](https://en.wikipedia.org/wiki/Standard_streams)) instead?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b10d5fb3",
   "metadata": {},
   "source": [
    "## 3.6. Connecting with other machines\n",
    "\n",
    "Sometimes we need to connect to other machines and send files through the command line. Personal computers do not always have enough computational capabilities to handle data but we can connect from our machines to other [High-Performance Computing systems (HPC)](https://en.wikipedia.org/wiki/High-performance_computing) that can handle larger data and more memory-intensive processes. If that is the case, then we will need to transfer large files between local (yours) and remote (the other), something that is not possible through email or simple cloud services.\n",
    "\n",
    "`ssh` is the most common command for securely login into a remote computer and executing commands on that machine. It generally looks like this:"
   ]
  },
  {
   "cell_type": "raw",
   "id": "369487d0",
   "metadata": {},
   "source": [
    "%%bash\n",
    "ssh username@remote_host\n",
    "\n",
    "# if we want to specify a specific port, usually port 22\n",
    "ssh -p 22 username@remote_host"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f283479",
   "metadata": {},
   "source": [
    "Once you are connected, you can run commands just like if you were on your machine, minding the folder structure of the remote machine you are connected to.\n",
    "\n",
    "`scp` allows you to securely transfer files between your local machine and the remote computer and viceversa. You need to keep in mind where you are (local machine? or already logged into the remote machine?) and the source and destination paths for the files."
   ]
  },
  {
   "cell_type": "raw",
   "id": "d43d114e",
   "metadata": {},
   "source": [
    "%%bash\n",
    "# the command below assumes you are sitting inside the machine from which you are transfering files\n",
    "# it is usually easier to transfer files from your local machine to remote machines when sitting inside your local machine (password and address)\n",
    "# copy local file to remote\n",
    "\n",
    "scp local_file username@remote_host:destination_path/\n",
    "\n",
    "# to transfer a file from a remote machine to the local machine in which you are sitting\n",
    "# copy remote file to local\n",
    "scp username@remote_host:remote_path/file destination_local_folder/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e7ab2de",
   "metadata": {},
   "source": [
    "`scp` is good enough when files are small to medium files and internet connections are robust. However, `scp` is not enough for transfering larger files over internet connections that might fail and when having to check the integrity of the file after transfer. \n",
    "\n",
    "For that, it is better to use `rsync`. Moreover, if the files exist in both the local and remote machines `rsyn` only transfers the changes between file copies and skips the rest, making the transfer faster."
   ]
  },
  {
   "cell_type": "raw",
   "id": "e1aaf231",
   "metadata": {},
   "source": [
    "%%bash\n",
    "# simple transfer from local machine to remote machine\n",
    "# -a keeps file permissions as they are in the source local machine\n",
    "# -v verbose, print all output and errors\n",
    "# -z compress data for transfer (faster)\n",
    "# after transfer, rsync checks that source and destination files are the same\n",
    "\n",
    "rsync -avz local_folder/ username@remote_host:destination_path"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c58fb7a0",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# [**4.  AWK**](#awk)  \n",
    "\n",
    "AWK is a programming language design to process text efficiently, particularly tabular text. AWK takes as inputs a pattern, an action, and a file or text from the *standard input* ([stdin](https://en.wikipedia.org/wiki/Standard_streams)) to operate upon. The pattern is a regex expression that indicates a text match or a field in the input file. The action is a command that will be executed on the match."
   ]
  },
  {
   "cell_type": "raw",
   "id": "8482f6c8",
   "metadata": {},
   "source": [
    "%%bash\n",
    "awk 'pattern { action }' input_file"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ddc8cbab",
   "metadata": {},
   "source": [
    "AWK has pre-defined variables that can be passed as patterns. The most common ones are variables that indicate the entire line of the file `$0` or a particular field (think column) of the file, e.g. `$1` and `$2` for printing the first and second fields/columns of the file. Very similar to the command `cut` would do. As AWK considers fields as variables, it is possible to indicate the delimiter with `-F`.\n",
    "\n",
    "We will apply AWK commands to `Zelote_coordinates.csv`, a tabular text file that contains geographic coordinates for the spider genera [*Zelotes*](https://en.wikipedia.org/wiki/Zelotes_subterraneus). The data was downloaded on July 26th, 2023 from GBIF and the DOI associated with the search is [here](https://doi.org/10.15468/dl.4ke7tf)\n",
    "\n",
    "First, use `head`, `cut`, or other commands that you have learnt so far to explore the data within the file."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fdbd00ce",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-success\">\n",
    "    <b>TO DO: </b>How many lines are in the file (each line represents a coordinate record)? How many different species are there in the file? Hint: you can use a combination of <i>head</i> and <i>cut</i> to figure out which file corresponds to the species ephitet. Append your answer to <i>answersL0_name.txt</i> using <i>>></i>.</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "04a1e72c",
   "metadata": {},
   "source": [
    "## 4.1. Basic filtering\n",
    "\n",
    "\n",
    "Now that you understand what is the structure of the file, the header and name of the columns, and the kind of information that it has, we will use AWK to extract information and clean records with flagged coordinate records. It is possible to carry out the same actions on the file using e.g. Excel, however, there's a limit to the size of files that Excel can process and command-line programs/languages like AWK are faster and more efficient, allowing you to process files that are [gigabytes](https://en.wikipedia.org/wiki/File_size) in size.\n",
    "\n",
    "First, we will filter out the columns that contain information that is not relevant for us right now. We can use AWK's `print` and field variables to do it. Then, we can add another a filter to only print the lines with non-empty `decimalLatitude` fields, i.e. printing records that are not missing the geolographic coordinates."
   ]
  },
  {
   "cell_type": "raw",
   "id": "a715c84e",
   "metadata": {},
   "source": [
    "%%bash\n",
    "# the new columns are \"order family genus species occurrenceStatus individualCount decimalLatitude decimalLongitude issue\"\n",
    "# \\t is the regular expression for tab\n",
    "awk -F '\\t' '{print $7, $8, $9, $10, $19, $20, $22, $23, $50}' Zelote_coordinates.csv > Zelote_coordinates_sel.csv\n",
    "\n",
    "# we are using the \\ symbol to introduce a line break without breaking the code (for easy reading)\n",
    "# OFS sets the output field separator\n",
    "# in the second part of the pipeline we need to ask awk to BEGIN the program to execute the first action before reading the input file\n",
    "# the filter is the if conditional and the symbols != mean is different than\n",
    "# printf prints formated strings\n",
    "# %s refers to a string match that will be replaced by the variable, in the same order that %s and $var appear\n",
    "# every %s is separated by a tab symbol and the line ends with the line-end regular expression \\n\n",
    "# we need to do it this way because the column species has a space\n",
    "\n",
    "awk -F '\\t' '{printf (\"%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t\\n\", $7, $8, $9, $10, $19, $20, $22, $23, $50)}' Zelote_coordinates.csv | \\\n",
    "awk -F '\\t' 'BEGIN {OFS = FS} {if ($7 != \"\") print $0;}' > Zelote_coordinates_sel.csv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5ddab3d",
   "metadata": {},
   "source": [
    "## 4.2. Operations\n",
    "\n",
    "It is possible to caryr out matematichal operations using AWK. The `Zelote_coordinates_sel.csv` column *individualCount* contains the number of spider observations per record. It is common that insects are recorded in counts instead of creating a single record for each insect with the same metadata. Thus, counting lines as records is not enough to count the total number of individuals per species that have been observed. But we can do that with AWK.\n",
    "\n",
    "AWK has associative arrays, something like a dictionary where the word is a key and the \"meaning\" is the value. Assuming that we will sum up the individual counts across all the records within a species, then the value of the species column is the key and the individual counts will be the values.\n",
    "\n",
    "First, familiarise yourself with the *species* and *individualCount* columns, then you can do the sums."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "d97aa17a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Zelotes anglo 246\n",
      "Zelotes tuobus 27\n",
      "Zelotes zhaoi 1\n",
      "Zelotes occidentalis 60\n",
      "Zelotes laccus 75\n",
      "Zelotes funestus 2\n",
      "Zelotes tenuis 163\n",
      "Zelotes gertschi 87\n",
      "Zelotes discens 74\n",
      "Zelotes electus 3433\n",
      "Zelotes mundus 1\n",
      "Zelotes exiguoides 133\n",
      "Zelotes scrutatus 0\n",
      "Zelotes perditus 8\n",
      "Zelotes anthereus 9\n",
      "Zelotes asiaticus 2\n",
      "Zelotes azsheganovae 178\n",
      "Zelotes gynethus 19\n",
      "Zelotes fulvopilosus 19\n",
      "Zelotes anatolyi 1\n",
      "Zelotes duplex 86\n",
      "Zelotes moestus 5\n",
      "Zelotes bajo 3\n",
      "Zelotes latreillei 3975\n",
      "Zelotes sarawakensis 7\n",
      "Zelotes gabriel 23\n",
      "Zelotes hentzi 388\n",
      "Zelotes monachus 47\n",
      "Zelotes captator 1\n",
      "Zelotes egregioides 21\n",
      "Zelotes segrex 10\n",
      "Zelotes flagellans 16\n",
      "Zelotes hermani 37\n",
      "Zelotes kimwha 1\n",
      "Zelotes similis 37\n",
      "Zelotes pseudogallicus 39\n",
      "Zelotes plumiger 4\n",
      "Zelotes pullus 37\n",
      "Zelotes hardwar 1\n",
      "Zelotes laetus 5\n",
      "Zelotes potani 5\n",
      "Zelotes foresta 3\n",
      "Zelotes sula 87\n",
      "Zelotes spadix 1\n",
      "Zelotes aurantiacus 25\n",
      "Zelotes pinos 2\n",
      "Zelotes subterraneus 3136\n",
      "Zelotes nannodes 173\n",
      "Zelotes clivicola 1290\n",
      "Zelotes gallicus 42\n",
      "Zelotes longipes 4055\n",
      "Zelotes metellus 1\n",
      "Zelotes erebeus 106\n",
      "Zelotes manius 9\n",
      "Zelotes cordubensis 2\n",
      "Zelotes wuchangensis 2\n",
      "Zelotes ocala 1\n",
      "Zelotes petrensis 4679\n",
      "Zelotes puritanus 628\n",
      "Zelotes lasalanus 679\n",
      "Zelotes fuzeta 19\n",
      "Zelotes criniger 3\n",
      "Zelotes aiken 5\n",
      "Zelotes fuligineus 1\n",
      "Zelotes griswoldi 5\n",
      "Zelotes talpinus 40\n",
      "Zelotes devotus 1\n",
      "Zelotes pseustes 10\n",
      "Zelotes fuscus 86\n",
      "Zelotes lymnophilus 1\n",
      "Zelotes catholicus 1\n",
      "Zelotes josephine 2\n",
      "Zelotes exiguus 98\n",
      "Zelotes thorelli 15\n",
      "Zelotes atrocaeruleus 1\n",
      "Zelotes limnatis 1\n",
      "Zelotes callidus 35\n",
      "Zelotes grovus 11\n",
      "Zelotes florodes 12\n",
      "Zelotes khatlonicus 2\n",
      "Zelotes aeneus 404\n",
      "Zelotes viola 39\n",
      "Zelotes oblongus 4\n",
      "Zelotes fratris 2019\n",
      "Zelotes apricorum 80\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# we need to skip the header line by asking Number Record higher than one --> NR>1\n",
    "# && means AND as a conditional such that if species is not empty AND individualCount is not empty\n",
    "# the information passed to the second pipe only has two columns, species and individualCount\n",
    "# the last part creates an array called sum from the species column and adds the values in column 2 as a list of values\n",
    "\n",
    "# AWK ends the first action and starts iterating through the array sum, adding all the individual counts in the list of values\n",
    "\n",
    "awk '(NR>1)' Zelote_coordinates_sel.csv \\\n",
    "| awk -F '\\t' '{if ($4 != \"\" && $6 != \"\") printf\"%s\\t%s\\n\", $4, $6;}' \\\n",
    "| awk -F '\\t' '{ sum[$1] += $2 } END { for (group in sum) print group, sum[group] }'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4372b4a2",
   "metadata": {},
   "source": [
    "The code above then takes a tabular list of geographic records from GBIF, then removes lines with no coordinates or species information, and finally counts how many individual counts in total there are per species."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4bac3b1f",
   "metadata": {},
   "source": [
    "# [**5. Summary**](#summary) \n",
    "\n",
    "**Navigating the folders:**  \n",
    "`cd`: Change directory  \n",
    "`ls`: List files and directories  \n",
    "`pwd`: Print working directory  \n",
    "\n",
    "**File and folder manipulation:**  \n",
    "`mkdir`: Create a new directory\n",
    "`touch`: Create a new file  \n",
    "`cp`: Copy files or directories  \n",
    "`mv`: Move or rename files or directories  \n",
    "`rm`: Remove files or directories  \n",
    "`grep`: Search for a specific pattern in files  \n",
    "`find`: Search for files and directories  \n",
    "\n",
    "**Permissions:**  \n",
    "`chmod`: Change file permissions  \n",
    "`chown`: Change file ownership  \n",
    "\n",
    "**redirecting information and piping:**  \n",
    "`>`: Redirect output to a file (overwrite)  \n",
    "`>>` Redirect output to a file (append)  \n",
    "`|`: Pipe output of one command as input to another  \n",
    "\n",
    "**Exploring and editing files:**  \n",
    "`cat`: Concatenate and display file content  \n",
    "`less`: View file content with pagination  \n",
    "`head`: Display the beginning of a file  \n",
    "`tail`: Display the end of a file  \n",
    "`nano`: Text editor for modifying files  \n",
    "\n",
    "**Data Manipulation:**  \n",
    "`awk`: Text processing and data extraction  \n",
    "`sed`: Stream editor for modifying text  \n",
    "\n",
    "**Process management:**  \n",
    "`ps`: View running processes  \n",
    "`kill`: Terminate a process  \n",
    "\n",
    "**File compression:**  \n",
    "`tar`: Archive files and directories  \n",
    "`gzip` or `gunzip`: Compress or decompress files  \n",
    "\n",
    "**Remote connections and transfers:**  \n",
    "`ssh`: Securely connect to a remote server  \n",
    "`scp`: Securely copy files between local and remote machines  \n",
    "`rsync`: Securely and efficiently transfer files, more flexible  \n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
